<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Changelog • radiator</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-news">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/radiator" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href="https://github.com/thierrygosselin/radiator/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.5" id="radiator-125-2023-01-20">radiator 1.2.5 2023-01-20<a class="anchor" aria-label="anchor" href="#radiator-125-2023-01-20"></a></h2>
<ul><li>huge work on <code>filter_ma</code> that now incorporate Minor Allele Frequency (MAF), Minor Allele Count (MAC) and Minor Allele Depth (MAD).</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.4" id="radiator-124-2022-11-17">radiator 1.2.4 2022-11-17<a class="anchor" aria-label="anchor" href="#radiator-124-2022-11-17"></a></h2>
<ul><li>bug fix when using some tidy function from GDS, the connection was not closing properly</li>
<li>work on genind, genlight, genepop, hierfstat, arlequin functions</li>
<li>updated the vignette</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.3" id="radiator-123-2022-10-30">radiator 1.2.3 2022-10-30<a class="anchor" aria-label="anchor" href="#radiator-123-2022-10-30"></a></h2>
<ul><li>bug fix with reading VCF, the new update works with SeqArray (&gt;= 1.36.0)</li>
<li>work on Tidy VCF.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.2" id="radiator-122-2021-12-20">radiator 1.2.2 2021-12-20<a class="anchor" aria-label="anchor" href="#radiator-122-2021-12-20"></a></h2>
<ul><li>bug fix when using <em>arlequin</em> output file</li>
<li>bug fix using DArT data</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.1" id="radiator-121-2021-10-28">radiator 1.2.1 2021-10-28<a class="anchor" aria-label="anchor" href="#radiator-121-2021-10-28"></a></h2>
<ul><li>bug fix when using genepop file and <code>genomic_converter</code>
</li>
<li>bug fix in <code>genomic_converter</code> when using genlight as input and PLINK as output file</li>
<li>More GATK VCF problems detections and warnings</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.0" id="radiator-120-2021-06-16">radiator 1.2.0 2021-06-16<a class="anchor" aria-label="anchor" href="#radiator-120-2021-06-16"></a></h2>
<ul><li>updated to work with R 4.1.0</li>
<li>bug fix with some format during conversion when necessary genotype format wasn’t found</li>
<li>more safe results when importing VCF files with windows and parallel processing</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.9" id="radiator-119-2020-12-14">radiator 1.1.9 2020-12-14<a class="anchor" aria-label="anchor" href="#radiator-119-2020-12-14"></a></h2>
<ul><li>bug fix using plink files and future</li>
<li>continue to test future backend and carrier</li>
<li>bug fix when using microsatellites as input and some specific output format</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.8" id="radiator-118-2020-10-17">radiator 1.1.8 2020-10-17<a class="anchor" aria-label="anchor" href="#radiator-118-2020-10-17"></a></h2>
<ul><li>No longer using Travis CI and AppVeyor to test the package</li>
<li>R-CMD-check: now using GitHub actions and rhub that test on the 3 OS.</li>
<li>Taking advantage of future, furrr and carrier packages</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.7" id="radiator-117-2020-08-21">radiator 1.1.7 2020-08-21<a class="anchor" aria-label="anchor" href="#radiator-117-2020-08-21"></a></h2>
<ul><li>PLINK files: fix a couple of bugs reading the tped</li>
<li>removed <code><a href="https://tidyr.tidyverse.org/reference/gather.html" class="external-link">tidyr::gather</a></code> and <code><a href="https://tidyr.tidyverse.org/reference/spread.html" class="external-link">tidyr::spread</a></code> dependencies (they are deprecated)</li>
<li>DArT data in 1-row format was not working properly with latest <code>data.table</code> melt function. Changed to <code>tidyr::pivot_long</code>.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.6" id="radiator-116-2020-06-23">radiator 1.1.6 2020-06-23<a class="anchor" aria-label="anchor" href="#radiator-116-2020-06-23"></a></h2>
<ul><li>updated radiator so that it work with latest release of SeqArray (v.1.28.1), that introduced breaking changes.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.5" id="radiator-115-2020-03-07">radiator 1.1.5 2020-03-07<a class="anchor" aria-label="anchor" href="#radiator-115-2020-03-07"></a></h2>
<ul><li>
<code>fineRADstructure</code> rules have changed, I’m no longer doing gymnastic to make the population work with <code>genomic_converter</code> or <code>write_fineradstructure</code>, it’s not up to the user to make sure the pop id starts with a letter. Some codes might now be broken because of this.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.4" id="radiator-114-2020-02-04">radiator 1.1.4 2020-02-04<a class="anchor" aria-label="anchor" href="#radiator-114-2020-02-04"></a></h2>
<ul><li>
<code>read_plink</code> and <code>tidy_plink</code>: new rules to get the best out of PLINK <strong>tped</strong> and <strong>bed</strong> files.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.3" id="radiator-113-2020-01-21">radiator 1.1.3 2020-01-21<a class="anchor" aria-label="anchor" href="#radiator-113-2020-01-21"></a></h2>
<ul><li>
<code>detect_microsatellites</code>: new radiator function that detect microsatellites using GMATA.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.0" id="radiator-110-2019-05-31">radiator 1.1.0 2019-05-31<a class="anchor" aria-label="anchor" href="#radiator-110-2019-05-31"></a></h2>
<ul><li>2 new output formats: <code>genepopedit</code> and <code>rubias</code>
</li>
<li>1 new detect function: <code>detect_paralogs</code> that copy the method described in McKinney et al. 2017. This function is the logical step to make available for users after <code>detect_mixed_genomes</code>, <code>filter_hwe</code> and <code>sexy_markers</code>.</li>
<li>
<code>sexy_markers</code>: couple of bug fixed. More testing with the different genomic format.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.0.0" id="radiator-100-2019-05-01">radiator 1.0.0 2019-05-01<a class="anchor" aria-label="anchor" href="#radiator-100-2019-05-01"></a></h2>
<ul><li>There is still documentation and vignette to fix, but this is the release that will be submitted to CRAN</li>
<li>Major <code>SeqArray</code> and <code>GDS</code> integration</li>
<li>Imputation module was removed from <code>radiator</code> and now lives exclusively in package <code>grur</code>
</li>
<li>
<code>filter_dart</code> is deprecated. Please use <code>filter_rad</code>, the ONE function to rule them all;)</li>
<li>Worked on travis</li>
<li>pkgdown website</li>
<li>vignettes</li>
<li>better function doc</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.21" id="radiator-0021-2018-11-14">radiator 0.0.21 2018-11-14<a class="anchor" aria-label="anchor" href="#radiator-0021-2018-11-14"></a></h2>
<ul><li>
<code>snp.ld</code>: using missing data now works by chromosome/scaffold.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.20" id="radiator-0020-2018-11-12">radiator 0.0.20 2018-11-12<a class="anchor" aria-label="anchor" href="#radiator-0020-2018-11-12"></a></h2>
<ul><li>
<code>tidy_vcf</code>, <code>tidy_genomic_data</code> and <code>genomic_converter</code>: works better with ipyrad vcf’s</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.19" id="radiator-0019-2018-11-07">radiator 0.0.19 2018-11-07<a class="anchor" aria-label="anchor" href="#radiator-0019-2018-11-07"></a></h2>
<ul><li>
<code>tidy_vcf</code>, <code>tidy_genomic_data</code> and <code>genomic_converter</code>: works better with freebayes and stacks vcf</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.18" id="radiator-0018-2018-10-23">radiator 0.0.18 2018-10-23<a class="anchor" aria-label="anchor" href="#radiator-0018-2018-10-23"></a></h2>
<ul><li>
<code>tidy_vcf</code>, <code>tidy_genomic_data</code> and <code>genomic_converter</code>: work without strata/pop grouping</li>
<li>
<code>snp_ld</code>: new argument <code>ld.threshold</code> for long.distance linkage disequilibrium.</li>
<li>
<code>write_vcf</code>: will now output ID as LOCUS_COL or LOCUS_(POS-1) if COL info is not provided.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.17" id="radiator-0017-2018-10-04">radiator 0.0.17 2018-10-04<a class="anchor" aria-label="anchor" href="#radiator-0017-2018-10-04"></a></h2>
<ul><li>
<code>tidy_vcf</code>, <code>tidy_genomic_data</code> and <code>genomic_converter</code>: way faster with huge VCF</li>
<li>
<code>tidy_vcf</code> and <code>write_seqarray</code>: work better with vcf generated by Stacks</li>
<li>
<code>write_ldna</code>: new function that generates a LDna input file from a tidy data frame.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.16" id="radiator-0016-2018-09-04">radiator 0.0.16 2018-09-04<a class="anchor" aria-label="anchor" href="#radiator-0016-2018-09-04"></a></h2>
<ul><li>
<code>tidy_vcf</code>, <code>tidy_genomic_data</code> and <code>genomic_converter</code>: way faster with huge VCF</li>
<li>
<code>write_fineradstructure</code>: fix bug when data was from DArT</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.15" id="radiator-0015-2018-08-17">radiator 0.0.15 2018-08-17<a class="anchor" aria-label="anchor" href="#radiator-0015-2018-08-17"></a></h2>
<ul><li>
<code>genomic_converter</code>, <code>tidy_genomic_data</code>: bug fix when individuals are integers</li>
<li>
<code>fis_summary</code>: arguments updated</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.14" id="radiator-0014-2018-08-04">radiator 0.0.14 2018-08-04<a class="anchor" aria-label="anchor" href="#radiator-0014-2018-08-04"></a></h2>
<ul><li>
<code>filter_dart</code> and <code>filter_rad</code>: can now opt out of HWE filtering.</li>
<li>
<code>filter_hwe</code>: user can opt to see figures but skip filtering</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.13" id="radiator-0013-2018-07-09">radiator 0.0.13 2018-07-09<a class="anchor" aria-label="anchor" href="#radiator-0013-2018-07-09"></a></h2>
<ul><li>working to make radiator work correctly with ggplot2 v.3.0.0</li>
<li>radiator ready for R 3.5.1 “Feather Spray” released on 2018/07/05</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.12" id="radiator-0012-2018-06-21">radiator 0.0.12 2018-06-21<a class="anchor" aria-label="anchor" href="#radiator-0012-2018-06-21"></a></h2>
<ul><li>transferred <code>write_gsi_sim</code> from assigner to radiator</li>
<li>2 <code>write</code> functions: <code>write_snprelate</code> and <code>write_seqarray</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.11" id="radiator-0011-2018-04-26">radiator 0.0.11 2018-04-26<a class="anchor" aria-label="anchor" href="#radiator-0011-2018-04-26"></a></h2>
<ul><li>when individuals in strata file/object and data don’t match, an error is generated</li>
<li>to reduce <code>radiator</code> dependencies, several packages were moved in the <strong>Suggests</strong> field.</li>
<li>
<code>filter_dart</code>: lots of new stuff. More appropriate filter arguments:
<ul><li>
<code>filter.coverage</code> is deprecated in favour of: filter.markers.coverage</li>
<li>
<code>filter.ind.missing.geno</code> is deprecated in favour of: <code>filter.markers.missing</code>
</li>
<li>
<code>erase.genotypes</code>: new argument tailored to handle coverage for DArT counts data.</li>
<li>
<code>filter.individuals.missing</code>: allows to blacklist sample early in the filtering pipeline.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.10" id="radiator-0010-2018-03-06">radiator 0.0.10 2018-03-06<a class="anchor" aria-label="anchor" href="#radiator-0010-2018-03-06"></a></h2>
<ul><li>New output file: stockr This output file format enables to run the data in the stockR package from Scott Fisher at CSIRO in Hobart.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.9" id="radiator-009-2018-03-01">radiator 0.0.9 2018-03-01<a class="anchor" aria-label="anchor" href="#radiator-009-2018-03-01"></a></h2>
<ul><li>New output file: related This output file format enables to run the data in the related R package, which is essantially the R version of <em>COANCESTRY</em> fortran program developed by Jinliang Wang.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.8" id="radiator-008-2018-02-08">radiator 0.0.8 2018-02-08<a class="anchor" aria-label="anchor" href="#radiator-008-2018-02-08"></a></h2>
<ul><li>New output file: fineRADstructure</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.7" id="radiator-007-2018-02-06">radiator 0.0.7 2018-02-06<a class="anchor" aria-label="anchor" href="#radiator-007-2018-02-06"></a></h2>
<ul><li>Better VCF parsing</li>
<li>
<code>radiator</code> is ready for stacks v.2 beta8</li>
<li>starting to re-introduce <code>data.table</code> to increase speed during melting or casting data frames</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.6" id="radiator-006-2017-10-02">radiator 0.0.6 2017-10-02<a class="anchor" aria-label="anchor" href="#radiator-006-2017-10-02"></a></h2>
<ul><li>
<code>tidy_genomic_data</code> and <code>genomic_converter</code>: bug fix and improvements.</li>
<li>
<code>write_snprelate</code> and it’s associated output from <code>genomic_converter</code> is no longer available because of difficulties students are having in installing the package, it’s dependencies and/or using the output. The code remains available <a href="https://www.dropbox.com/s/7xujizkvpi0ddac/write_snprelate.R?dl=0" class="external-link">here</a>, without support.</li>
<li>
<code>radiator::snp.ld</code>: the module inside <code>tidy_genomic_data</code> and <code>genomic_converter</code> to minimize short linkage disequilibrium as a new option <code>snp.ld = "middle"</code>. For locus with &gt; 2 SNPs/read the option allows to select at random one SNP between the first and the last SNP on the read. If the locus as &lt;= 2 SNPs on the read, the first one is selected. Note that for that last option, the numbers are reported. Thanks to Ido Bar for the idea.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.5" id="radiator-005-2017-09-20">radiator 0.0.5 2017-09-20<a class="anchor" aria-label="anchor" href="#radiator-005-2017-09-20"></a></h2>
<ul><li>
<code>tidy_dart</code> and <code>filter_dart</code>:
<ul><li>big overhaul.</li>
<li>importing 1 and 2 row genotypes (called sometimes binary format) is easier and no longer require to prep the DArT file (the function parse DArT data with <code>*</code> at the beginning of certain lines).</li>
<li>filtering is easier and now interactive for those who want.</li>
<li>plots are generated automatically.</li>
</ul></li>
<li>
<code>tidy_genomic_data</code> and <code>genomic_converter</code> can now accept DArT data directly.</li>
<li>several typos where fix. Thanks to <a href="https://github.com/IdoBar" class="external-link">@IdoBar</a> for this.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.4" id="radiator-004-2017-09-07">radiator 0.0.4 2017-09-07<a class="anchor" aria-label="anchor" href="#radiator-004-2017-09-07"></a></h2>
<ul><li>new function: <code>run_bayescan</code> to run <em>BayeScan</em> … with <em>radiator</em>
</li>
<li>new separate <code>write</code> functions also included in <code>genomic_converter</code>: <code>write_bayescan</code>, <code>write_pcadapt</code> and <code>write_hzar</code>.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.3" id="radiator-003-2017-07-18">radiator 0.0.3 2017-07-18<a class="anchor" aria-label="anchor" href="#radiator-003-2017-07-18"></a></h2>
<ul><li>small mod</li>
<li>writting in specific folder</li>
<li>writting specific plot</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.2" id="radiator-002-2017-07-15">radiator 0.0.2 2017-07-15<a class="anchor" aria-label="anchor" href="#radiator-002-2017-07-15"></a></h2>
<ul><li>Work on different functions to prep for official package launch</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.1" id="radiator-001-2017-06-18">radiator 0.0.1 2017-06-18<a class="anchor" aria-label="anchor" href="#radiator-001-2017-06-18"></a></h2>
<ul><li>First commit</li>
</ul></div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

