<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Get Started • radiator</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="48x48" href="../favicon-48x48.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Get Started">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/thierrygosselin/radiator" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Get Started</h1>
                        <h4 data-toc-skip class="author">Thierry
Gosselin</h4>
            
            <h4 data-toc-skip class="date">2025-07-04</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/radiator/blob/HEAD/vignettes/get_started.Rmd" class="external-link"><code>vignettes/get_started.Rmd</code></a></small>
      <div class="hidden name"><code>get_started.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="learning-radiator">Learning radiator<a class="anchor" aria-label="anchor" href="#learning-radiator"></a>
</h2>
<p>To see if radiator is the right tool for you, you can start with the
basics.</p>
<div class="section level3">
<h3 id="prepare-a-strata-file">1. Prepare a strata file<a class="anchor" aria-label="anchor" href="#prepare-a-strata-file"></a>
</h3>
<ul>
<li>It’s a tab separated file, e.g. <a href="https://www.dropbox.com/s/g0vsek0dmtpxntt/example.strata.tsv?dl=0" class="external-link">example.strata.tsv</a>.</li>
<li>A minimum of 2 columns: <code>INDIVIDUALS</code> and
<code>STRATA</code> is required.</li>
<li>The <code>STRATA</code> column identifies the individuals
stratification, the hierarchical groupings: populations, sampling sites
or any grouping you want.</li>
<li>It’s like <em>stacks</em> population map file with header…</li>
<li>DArT users: the strata requires 3 columns and is described in
<code>??radiator::readr_dart</code> <a href="https://www.dropbox.com/s/utq2h6o00v55kep/example.dart.strata.tsv?dl=0" class="external-link">example.dart.strata.tsv</a>.</li>
</ul>
<p>To make sure it’s going to work properly, try:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">radiator</span><span class="fu">::</span><span class="fu"><a href="../reference/summary_strata.html">summary_strata</a></span><span class="op">(</span><span class="st">"my.strata.tsv"</span><span class="op">)</span></span>
<span><span class="co"># more details in with `??radiator::summary_strata`</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filter-your-radseq-data">2. Filter your RADseq data<a class="anchor" aria-label="anchor" href="#filter-your-radseq-data"></a>
</h3>
<ul>
<li>
<code>filter_rad</code> is the ONE FUNCTION TO RULE THEM ALL.</li>
<li>There’s a built-in interactive mode that requires users to visualize
the data before choosing thresholds.</li>
<li>The function is made of filtering modules that user’s can also
access separately or in combination.</li>
<li>For 95% of users, <code>filter_rad</code> will be enough to start
exploring the biology!</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">radiator</span><span class="fu">::</span><span class="fu"><a href="../reference/filter_rad.html">filter_rad</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="st">"my.vcf"</span>,</span>
<span>    strata <span class="op">=</span> <span class="st">"my.strata.tsv"</span>, </span>
<span>    output <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"genind"</span>, <span class="st">"stockr"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p><strong>To cite radiator</strong>, inside R type
<code>citation("radiator")</code></p>
</div>
</div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<table class="table">
<colgroup>
<col width="25%">
<col width="74%">
</colgroup>
<thead><tr class="header">
<th align="left">Caracteristics</th>
<th align="left">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><strong>Import</strong></td>
<td align="left">List of the 14 supported input genomic file formats
(<strong>diploid data only</strong>) and their variations:<br><a href="https://samtools.github.io/hts-specs/" class="external-link">VCF: SNPs and
haplotypes</a> (Danecek et al., 2011)<br><a href="http://www.diversityarrays.com" class="external-link">DArT files (5): genotypes in 1row,
alleles counts and coverage in 2 rows, SilicoDArT genotypes and
counts</a><br><a href="https://www.cog-genomics.org/plink2" class="external-link">PLINK:
bed/tped/tfam</a> (Purcell et al., 2007)<br><a href="https://github.com/thibautjombart/adegenet" class="external-link">genind</a> (Jombart et
al., 2010; Jombart and Ahmed, 2011)<br><a href="https://github.com/thibautjombart/adegenet" class="external-link">genlight</a> (Jombart
et al., 2010; Jombart and Ahmed, 2011)<br><a href="https://github.com/EricArcher/strataG" class="external-link">strataG gtypes</a> (Archer
et al., 2016)<br><a href="http://genepop.curtin.edu.au" class="external-link">Genepop</a>
(Raymond and Rousset, 1995; Rousset, 2008)<br><a href="http://catchenlab.life.illinois.edu/stacks/" class="external-link">STACKS haplotype
file</a> (Catchen et al., 2011, 2013)<br><a href="https://github.com/jgx65/hierfstat" class="external-link">hierfstat</a> (Goudet,
2005)<br><a href="https://github.com/zhengxwen/SeqArray" class="external-link">SeqArray</a>
(Zheng et al., 2017)<br><a href="https://github.com/zhengxwen/SNPRelate" class="external-link">SNPRelate</a> (Zheng et
al., 2012)<br>Dataframes of genotypes in wide or long/tidy
format<br><br><strong>Reading and tidying is found inside:
<code>genomic_converter</code>, <code>tidy_</code> and
<code>read_</code> functions</strong>
</td>
</tr>
<tr class="even">
<td align="left"><strong>Output</strong></td>
<td align="left">29 genomic data formats can be exported out of
<strong>radiator</strong>. The module responsible for this is
<code>genomic_converter</code>. Separate modules handles the different
formats and are also available:<br><code>write_vcf</code>: <a href="https://samtools.github.io/hts-specs/" class="external-link">VCF</a> (Danecek et al.,
2011)<br><code>write_plink</code>: <a href="https://www.cog-genomics.org/plink2" class="external-link">PLINK tped/tfam</a> (Purcell
et al., 2007)<br><code>write_genind</code>: <a href="https://github.com/thibautjombart/adegenet" class="external-link">adegenet genind and
genlight</a> (Jombart et al., 2010; Jombart and Ahmed,
2011)<br><code>write_genlight</code>: <a href="https://github.com/thibautjombart/adegenet" class="external-link">genlight</a> (Jombart
et al., 2010; Jombart and Ahmed, 2011)<br><code>write_gsi_sim</code>: <a href="https://github.com/eriqande/gsi_sim" class="external-link">gsi_sim</a> (Anderson et
al. 2008)<br><code>write_rubias</code>: <a href="https://github.com/eriqande/rubias" class="external-link">rubias</a> (Moran and
Anderson, 2018)<br><code>write_gtypes</code>: <a href="https://github.com/EricArcher/strataG" class="external-link">strataG gtypes</a> (Archer
et al. 2016)<br><code>write_colony</code>: <a href="https://www.zsl.org/science/software/colony" class="external-link">COLONY</a> (Jones and
Wang, 2010; Wang, 2012)<br><code>write_genepop</code>: <a href="http://genepop.curtin.edu.au" class="external-link">Genepop</a> (Raymond and Rousset,
1995; Rousset, 2008)<br><code>write_genepopedit</code>: <a href="https://github.com/rystanley/genepopedit" class="external-link">genepopedit</a> (Stanley
et al., 2017)<br><a href="http://catchenlab.life.illinois.edu/stacks/" class="external-link">STACKS haplotype
file</a> (Catchen et al., 2011, 2013)<br><code>write_betadiv</code>: <a href="http://adn.biol.umontreal.ca/~numericalecology/Rcode/" class="external-link">betadiv</a>
(Lamy, 2015)<br><code>write_dadi</code>: <a href="http://gutengroup.mcb.arizona.edu/software/" class="external-link">δaδi</a> (Gutenkunst
et al., 2009)<br><code>write_structure</code>: <a href="https://web.stanford.edu/group/pritchardlab/structure.html" class="external-link">structure</a>
(Pritchard et al., 2000)<br><code>write_faststructure</code>: <a href="https://github.com/rajanil/fastStructure" class="external-link">faststructure</a> (Raj
&amp; Pritchard, 2014)<br><code>write_arlequin</code>: <a href="http://cmpg.unibe.ch/software/arlequin35/" class="external-link">Arlequin</a> (Excoffier
et al. 2005)<br><code>write_hierfstat</code>: <a href="https://github.com/jgx65/hierfstat" class="external-link">hierfstat</a> (Goudet,
2005)<br><code>write_snprelate</code>: <a href="https://github.com/zhengxwen/SNPRelate" class="external-link">SNPRelate</a> (Zheng et
al. 2012)<br><code>write_seqarray</code>: <a href="https://github.com/zhengxwen/SeqArray" class="external-link">SeqArray</a> (Zheng et
al. 2017)<br><code>write_bayescan</code>: <a href="http://cmpg.unibe.ch/software/BayeScan" class="external-link">BayeScan</a> (Foll and
Gaggiotti, 2008)<br><code>write_pcadapt</code>: <a href="https://github.com/bcm-uga/pcadapt" class="external-link">pcadapt</a> (Luu et
al. 2017)<br><code>write_hzar</code> (Derryberry et al. 2013)
<br><code>write_fineradstructure</code> (Malinsky et al., 2018)
<br><code>write_related</code> <a href="https://github.com/timothyfrasier/related" class="external-link">related</a> (Pew et
al., 2015)<br><code>write_stockr</code> for stockR package (Foster el
al., submitted)<br><code>write_maverick</code> <a href="http://www.bobverity.com/home/maverick/what-is-maverick/" class="external-link">MavericK</a>
(Verity &amp; Nichols, 2016)<br><code>write_ldna</code> <a href="https://github.com/petrikemppainen/LDna" class="external-link">LDna</a> (Kemppainen et
al. 2015)<br><code>write_hapmap</code> <a href="https://www.ncbi.nlm.nih.gov/variation/news/NCBI_retiring_HapMap/" class="external-link">HapMap</a><br>Dataframes
of genotypes in wide or long/tidy format</td>
</tr>
<tr class="odd">
<td align="left"><strong>Conversion function</strong></td>
<td align="left">
<code>genomic_converter</code> import/export genomic
formats mentioned above. The function is also integrated with usefull
filters, blacklist and whitelist.</td>
</tr>
<tr class="even">
<td align="left"><strong>Outliers detection</strong></td>
<td align="left">
<code>detect_duplicate_genomes</code>: detect and
remove duplicate individuals from your dataset
<br><code>detect_mixed_genomes</code>: detect and remove potentially
mixed individuals<br><code>stackr::summary_haplotype</code> and
<code>filter_snp_number</code>: Discard of outlier markers with <em>de
novo</em> assembly artifact (e.g. markers with an extreme number of SNP
per haplotype or with irregular number of alleles)</td>
</tr>
<tr class="odd">
<td align="left"><strong>Filters</strong></td>
<td align="left">Targets of filters: alleles, genotypes, markers,
individuals and populations and associated metrics and statistics can be
filtered and/or selected in several ways inside the main filtering
function <code>filter_rad</code> and/or the underlying
modules:<br><br><code>filter_rad</code>: designed for RADseq data, it’s
the <em>one function to rule them all</em>. Best used with unfiltered or
very low filtered VCF (or listed input) file. The function can handle
very large VCF files (e.g. no problem with &gt;2M SNPs, &gt; 30GB
files), all within R!!<br><code>filter_dart_reproducibility</code>:
blaclist markers under a certain threshold of DArT reproducibility
metric.<br><code>filter_monomorphic</code>: blacklist markers with only
1 morph.<br><code>filter_common_markers</code>: keep only markers common
between strata.<br><code>filter_individuals</code>: blacklist
individuals based on missingness, heterozygosity and/or total
coverage.<br><code>filter_mac</code>: blacklist markers based on
minor/alternate allele count.<br><code>filter_coverage</code>: blacklist
markers based on mean read depth
(coverage).<br><code>filter_genotype_likelihood</code>: Discard markers
based on genotype likelihood<br><code>filter_genotyping</code>:
blacklist markers based on genotyping/call
rate.<br><code>filter_snp_position_read</code>: blacklist markers based
based on the SNP position on the
read/locus.<br><code>filter_snp_number</code>: blacklist locus with too
many SNPs.<br><code>filter_ld</code>: blacklist markers based on short
and/or long distance linkage disequilibrium.<br><code>filter_hwe</code>:
blacklist markers based on Hardy-Weinberg Equilibrium expectations
(HWE).<br><code>filter_het</code>: blacklist markers based on the
observed heterozygosity (Het obs).<br><code>filter_fis</code>: blacklist
markers based on the inbreeding coefficient
(Fis).<br><code>filter_whitelist</code>: keep only markers present in a
whitelist</td>
</tr>
<tr class="even">
<td align="left"><strong><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a>-based
plotting</strong></td>
<td align="left">Visualize distribution of important metric and
statistics and create publication-ready figures</td>
</tr>
<tr class="odd">
<td align="left"><strong>Parallel</strong></td>
<td align="left">Codes designed and optimized for fast computations
using Genomic Data Structure <a href="https://github.com/zhengxwen/gdsfmt" class="external-link">GDS</a> file format and data
science packages in <a href="https://www.tidyverse.org" class="external-link">tiverse</a>.
Works with all OS: Linux, Mac and now PC!</td>
</tr>
</tbody>
</table>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
