<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Detect alternate allele problems — detect_allele_problems • radiator</title><!-- favicons --><link rel="icon" type="image/png" sizes="48x48" href="../favicon-48x48.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Detect alternate allele problems — detect_allele_problems"><meta property="og:description" content="Detect alternate allele problems.
Used internally in radiator
and might be of interest for users.
The function computes alternate allele
counts and looks for alternate alleles bellow a certain threshold.
Keeping in mind the overall dataset QC, this function
helps to understand potential problems with
Allele DropOut and Null Alleles.
The summary statistics for the
markers with problematic allele is computed based on coverage and
genotype likelihood. This function is very fast to highlight: i) bias in
representation of allelic copies and unequal coverage and ii) type I error
during genotyping of heterozygote with low coverage data."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/radiator" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Detect alternate allele problems</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/radiator/blob/HEAD/R/detect_allele_problems.R" class="external-link"><code>R/detect_allele_problems.R</code></a></small>
    <div class="hidden name"><code>detect_allele_problems.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Detect alternate allele problems.
Used internally in <a href="https://github.com/thierrygosselin/radiator" class="external-link">radiator</a>
and might be of interest for users.</p>
<p>The function computes alternate allele
counts and looks for alternate alleles bellow a certain threshold.</p>
<p>Keeping in mind the overall dataset QC, this function
helps to understand <em>potential problems</em> with
<strong>Allele DropOut and Null Alleles</strong>.
The summary statistics for the
markers with problematic allele is computed based on coverage and
genotype likelihood. This function is very fast to highlight: i) bias in
representation of allelic copies and unequal coverage and ii) type I error
during genotyping of heterozygote with low coverage data.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">detect_allele_problems</span><span class="op">(</span><span class="va">data</span>, allele.threshold <span class="op">=</span> <span class="fl">3</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>A tidy data frame object in the global environment or
a tidy data frame in wide or long format in the working directory.
<em>How to get a tidy data frame ?</em>
Look into <span class="pkg">radiator</span> <code><a href="tidy_genomic_data.html">tidy_genomic_data</a></code>.</p></dd>


<dt id="arg-allele-threshold">allele.threshold<a class="anchor" aria-label="anchor" href="#arg-allele-threshold"></a></dt>
<dd><p>(integer) Threshold of alternate allele copies.
Below this threshold markers are blacklisted. Choose this threshold based
on your tolerance for uninformative or unreliable information and
minor allele frequency. Summary statistics for this blacklist will be calculated.
See details for more info.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>(optional, logical) When <code>verbose = TRUE</code>
the function is a little more chatty during execution.
Default: <code>verbose = TRUE</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>(optional) Advance mode that allows to pass further arguments
for fine-tuning the function (see details).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list with summary, plot and blacklist.
Summary statistics for the blacklisted markers are generated and include
coverage and genotype likelihood information.</p>
<p><code>$alt.allele.count.plot</code>: Distribution of marker's alternate allele number.
The shape is usually skewed right.</p>
<p><code>$alt.depth.count.distribution.plot</code>: Distribution of markers alternate
allele depth (in number of read). The range showed on the plot is
between 1 and 10, and you decide your tolerance for low coverage
and good genotype calls.</p>
<p>This plot is associated with <code>$number.markers</code>, a table that shows the
number of markers &lt;= <code>allele.threshold </code>. Also in the table, the
range of alternate allele read depth (1 to 10), same as plot but showing here
the the cumulative number of markers. Again, would you trust a marker with only
1 alternate allele (an heterozygote individual) with a read depth of 1 or 2 ?</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><strong>Under developement, use with caution</strong></p>

<p><strong>Input files:</strong> see <span class="pkg">radiator</span> <code><a href="tidy_genomic_data.html">tidy_genomic_data</a></code>
for detailed information about supported file format.</p>

<p><strong>allele.threshold</strong>:</p>
<p>An <code>allele.threshold = 1</code>, says that you want to check the statistics
for markers with only 1 copy of the alternate allele. Said differently,
<code>allele.threshold = 1</code> = only 1 heterozygote individual is making this
marker polymorphic, very thin line if this allele is backed by less than 3 reads.</p>
<p>Similarly, a <code>allele.threshold = 2</code> can represent markers with
2 heterozygote individuals calling the polymorphic markers or 1 individual
homozygote for the alternate allele.</p>
<p>Look for problem with :</p><ul><li><p>read depth/coverage problem</p></li>
<li><p>high imbalance between the alt and ref depth coverage</p></li>
<li><p>genotype likelihood abnormally lower than normal</p></li>
<li><p>alternate allele with NA for read depth information</p></li>
</ul><p><strong>Alternate allele with no depth information:</strong>
You can highlight those problematic genotypes with
alt.no.depth &lt;- dplyr::filter(your.object$allele.summary, is.na(ALLELE_ALT_DEPTH))</p>

<p>Use the summary statistics of the blacklisted markers to refine update
the blacklist of markers. You can decide to discard the markers or
blacklist the problematic genotypes.
<span class="pkg">radiator</span> <code><a href="tidy_genomic_data.html">tidy_genomic_data</a></code> and
<code><a href="genomic_converter.html">genomic_converter</a></code> allows to erase problematic genotypes
using a blacklist of genotypes...</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Thierry Gosselin <a href="mailto:thierrygosselin@icloud.com">thierrygosselin@icloud.com</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">problem</span> <span class="op">&lt;-</span> <span class="fu">detect_allele_problems</span><span class="op">(</span></span></span>
<span class="r-in"><span>        data <span class="op">=</span> <span class="va">salamander</span>,</span></span>
<span class="r-in"><span>        strata <span class="op">=</span> <span class="st">"strata.salmon.tsv"</span>,</span></span>
<span class="r-in"><span>        allele.threshold <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># The default with this function:</span></span></span>
<span class="r-in"><span><span class="co"># filter.monomorphic = TRUE # = discarded</span></span></span>
<span class="r-in"><span><span class="co"># filter.common.markers = TRUE # markers not in common between pop are discarded</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

