<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Detect mixed genomes — detect_mixed_genomes • radiator</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Detect mixed genomes — detect_mixed_genomes"><meta property="og:description" content="Highlight outliers individual's observed heterozygosity for a quick
diagnostic of mixed samples or poor polymorphism discovery due to DNA quality,
sequencing effort, etc."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/radiator" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Detect mixed genomes</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/radiator/blob/HEAD/R/detect_mixed_genomes.R" class="external-link"><code>R/detect_mixed_genomes.R</code></a></small>
    <div class="hidden name"><code>detect_mixed_genomes.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Highlight outliers individual's observed heterozygosity for a quick
diagnostic of mixed samples or poor polymorphism discovery due to DNA quality,
sequencing effort, etc.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">detect_mixed_genomes</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  interactive.filter <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  detect.mixed.genomes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ind.heterozygosity.threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  by.strata <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>(4 options) A file or object generated by radiator:</p><ul><li><p>tidy data</p></li>
<li><p>Genomic Data Structure (GDS)</p></li>
</ul><p><em>How to get GDS and tidy data ?</em>
Look into <code><a href="tidy_genomic_data.html">tidy_genomic_data</a></code>,
<code><a href="read_vcf.html">read_vcf</a></code> or
<code><a href="tidy_vcf.html">tidy_vcf</a></code>.</p></dd>


<dt id="arg-interactive-filter">interactive.filter<a class="anchor" aria-label="anchor" href="#arg-interactive-filter"></a></dt>
<dd><p>(optional, logical) Do you want the filtering session to
be interactive. Figures of distribution are shown before asking for filtering
thresholds.
Default: <code>interactive.filter = TRUE</code>.</p></dd>


<dt id="arg-detect-mixed-genomes">detect.mixed.genomes<a class="anchor" aria-label="anchor" href="#arg-detect-mixed-genomes"></a></dt>
<dd><p>(optional, logical) For use inside radiator pipelines.
Default: <code>detect.mixed.genomes = TRUE</code>.</p></dd>


<dt id="arg-ind-heterozygosity-threshold">ind.heterozygosity.threshold<a class="anchor" aria-label="anchor" href="#arg-ind-heterozygosity-threshold"></a></dt>
<dd><p>(string, double, optional)
Blacklist individuals based on observed heterozygosity (averaged across markers).</p>

<p>The string contains 2 thresholds values (min and max).
The values are proportions (0 to 1), where 0 turns off the min threshold and
1 turns off the max threshold.
Individuals with mean observed heterozygosity higher (&gt;) or lower (&lt;)
than the thresholds will be blacklisted.</p>
<p>Default: <code>ind.heterozygosity.threshold = NULL</code> will turn off completely
the filter and the function will only output the plots and table of heterozygosity.</p></dd>


<dt id="arg-by-strata">by.strata<a class="anchor" aria-label="anchor" href="#arg-by-strata"></a></dt>
<dd><p>(optional, logical)
Can only be used when <code>interactive.filter = TRUE</code>, allows to enter
heterozygosity thresholds by strata, instead of overall. This is not recommended
to use inside <code><a href="filter_rad.html">filter_rad</a></code> or when doing population discovery.
This is a good use when dealing with different species.
Default: <code>by.strata = FALSE</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>(optional, logical) When <code>verbose = TRUE</code>
the function is a little more chatty during execution.
Default: <code>verbose = TRUE</code>.</p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>(optional) The number of core used for parallel
execution during import.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>(optional) Advance mode that allows to pass further arguments
for fine-tuning the function. Also used for legacy arguments (see details or
special section)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>The function returns inside the global environment a list with
5 objects:</p>
<ol><li><p>the individual's heterozigosity (<code>$individual.heterozygosity</code>)
a dataframe containing for each individual, the population id, the number of
genotyped markers, the number of missing genotypes (based on the number of
markers of the population and overall), the number of markers genotyped as heterozygote
and it's proportion based on the number of genotyped markers.</p></li>
<li><p>the heterozygosity statistics per populations and overall:<code>$heterozygosity.statistics</code></p></li>
<li><p>the blacklisted individuals if <code>ind.heterozygosity.threshold</code> was selected: <code>$blacklist.ind.het</code></p></li>
<li><p>the boxplot of individual heterozygosity:<code>$individual.heterozygosity.boxplot</code></p></li>
<li><p>the manhattan plot of individual heterozygosity (<code>$individual.heterozygosity.manhattan.plot</code>)
contrasted with missingness proportion based on the number of markers (population or overall).
The 2 facets will be identical when the dataset as common markers between
the populations. The dotted lines are the mean hetegozygosities.</p></li>
</ol></div>
    <div id="details">
    <h2>Details</h2>
    <p>To help discard an individual based on his observed heterozygosity
(averaged across markers),
use the manhanttan plot to:</p><ol><li><p>contrast the individual with population and overall samples.</p></li>
<li><p>visualize the impact of missigness information (based on population or
overall number of markers) and the individual observed heterozygosity. The
larger the point, the more missing genotypes.</p></li>
</ol><p><strong>Outlier above average:</strong></p><ul><li><p>potentially represent two samples mixed together (action: blacklist), or...</p></li>
<li><p>a sample with more sequecing effort (point size small): did you merge your replicates fq files ? (action : keep and monitor)</p></li>
<li><p>a sample with poor sequencing effort (point size large) where the genotyped markers are
all heterozygotes, verify this with missingness (action: discard)</p></li>
</ul><p>In all cases, if there is no bias in the population sequencing effort,
the size of the point will usually be "average" based on the population or
overall number of markers.</p>

<p>You can visualize individual observed heterozygosity, choose thresholds and
then visualize, choose thresholds and filter markers based on observed
heterozygosity in one run with: <span class="pkg">radiator</span> <code><a href="filter_het.html">filter_het</a></code>.</p>
<p><strong>Outlier below average:</strong></p><ul><li><p>A point with a size larger than the population or overall average (= lots of missing):
the poor polymorphism discovery of the sample is probably the result of bad
DNA quality, a bias in sequencing effort, etc. (action: blacklist)</p></li>
<li><p>A point with a size that looks average (not much missing):
this sample requires more attention (action: blacklist) and conduct more tests.
e.g. for biallelic data, look for coverage imbalance between ALT/REF allele.
At this point you need to distinguish between an artifact of poor polymorphism discovery
or a biological reason (highly inbred individual, etc.).</p></li>
</ul><p><strong>heterozygosity and missing data:</strong>
If you see a pattern with the heterozygosity and missing data,
try changing the genotyping rate required to keep markers and/or individuals.</p>
    </div>
    <div id="note">
    <h2>Note</h2>
    <p>Thanks to Peter Grewe for very useful comments
on previous version of this function.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Thierry Gosselin <a href="mailto:thierrygosselin@icloud.com">thierrygosselin@icloud.com</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co">#Step1: highlight outlier individuals, the simplest way to run:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">radiator</span><span class="fu">::</span><span class="fu">detect_mixed_genomes</span><span class="op">(</span>data <span class="op">=</span> <span class="st">"wombat_tidy.rad"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># where the .rad file is a tidy data set generated by radiator.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Or if you don't have a tidy df:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">radiator</span><span class="fu">::</span><span class="fu"><a href="tidy_genomic_data.html">tidy_genomic_data</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    data <span class="op">=</span> <span class="st">"wombat.vcf"</span>,</span></span>
<span class="r-in"><span>    strata <span class="op">=</span> <span class="st">"strata.wombat.tsv"</span></span></span>
<span class="r-in"><span>    <span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span><span class="fu">radiator</span><span class="fu">::</span><span class="fu">detect_mixed_genomes</span><span class="op">(</span><span class="va">.</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

