<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Detect paralogs — detect_paralogs • radiator</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Detect paralogs — detect_paralogs"><meta property="og:description" content="The method uses the proportion of heterozygotes
and deviations in read ratios described in McKinney et al. 2017 to highlight
potential paralogous markers. Additionally, the function produces the HD
figure with marker's missingness accounted for.
Note: Violating Assumptions/Prerequisites (see below) can lead to
false positive or the absence of detection of paralogous markers."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/radiator" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Detect paralogs</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/radiator/blob/HEAD/R/detect_paralogs.R" class="external-link"><code>R/detect_paralogs.R</code></a></small>
    <div class="hidden name"><code>detect_paralogs.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The method uses the proportion of heterozygotes
and deviations in read ratios described in McKinney et al. 2017 to highlight
potential paralogous markers. Additionally, the function produces the HD
figure with marker's missingness accounted for.</p>
<p><em>Note:</em> Violating Assumptions/Prerequisites (see below) can lead to
false positive or the absence of detection of paralogous markers.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">detect_paralogs</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>(2 options) A Genomic Data Structure (GDS) file or object
generated by radiator.</p>
<p><em>How to get GDS data ?</em>
Look into <code><a href="tidy_genomic_data.html">tidy_genomic_data</a></code>,
<code><a href="read_vcf.html">read_vcf</a></code> or
<code><a href="tidy_vcf.html">tidy_vcf</a></code>.</p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>(optional) The number of core used for parallel
execution during import.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>(optional, logical) When <code>verbose = TRUE</code>
the function is a little more chatty during execution.
Default: <code>verbose = TRUE</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>(optional) Advance mode that allows to pass further arguments
for fine-tuning the function. Also used for legacy arguments (see details or
special section)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>The function returns the GDS data inside the global environment.
In the directory generated with <code>detect_paralogs</code> in it, the function
writes a tibble with the summary and the HD plot.</p>
<p>The tibble as 1 line per SNPs/MARKERS and columns are:</p><ul><li><p>VARIANT_ID: the GDS SNP id, to have all the markers metadata use
<code><a href="extract_markers_metadata.html">extract_markers_metadata</a></code>.</p></li>
<li><p>NUMBER_HET: the number of heterozygote genotype.</p></li>
<li><p>NUMBER_MISSING: the number of missing genotype.</p></li>
<li><p>NUMBER_GENOTYPED: the number of individual genotyped for the marker.</p></li>
<li><p>TOTAL: the total number of sample for this marker.</p></li>
<li><p>DEPTH_REF: the sum of the REF allele depth/coverage.</p></li>
<li><p>DEPTH_ALT: the sum of the ALT allele depth/coverage.</p></li>
<li><p>MISSING_PROP: Missing genotype proportion.</p></li>
<li><p>HET_PROP: the proportion if heterozygote sample.</p></li>
<li><p>TOTAL_COUNTS: Total coverage/counts for the marker.</p></li>
<li><p>RATIO: the allele ratio.</p></li>
<li><p>Z_SCORE: the z-score (deviation in read ratio).</p></li>
</ul></div>
    <div id="assumptions-prerequisites">
    <h2>Assumptions/Prerequisites</h2>


<ol><li><p><strong>read the paper:</strong>
Read <a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12613" class="external-link">Garrett McKinney</a>
paper, twice. It's good.
Reference is below.</p></li>
<li><p><strong>Depth/Coverage information:</strong> The function requires read depth
information for each alleles (REF/ALT). This info will not be found in genepop
or structure file format. You usually need a VCF file for this.
<em>stacks, ANGSD, GATK, platypus, ipyrad, DArT, VCFTools, PLINK and freebayes</em>
are software producing the required output.</p></li>
<li><p><strong>Batch effect:</strong> You don't have any batch effect, e.g. with
sequencer id, library, sampling sites, tissue preservation,
person behind the wet-lab bench conducting the DNA extraction
(operating the pipette, etc.).</p></li>
<li><p><strong>Normalization:</strong> Any form of normalization, before or after
sequencing, as the potential to bias paralogs discovery.
This include: combining lanes or chips to increase individual coverage,
normalizing individuals FASTQ files, and of course normalizing read depth.</p></li>
<li><p><strong>Identity-by-Missingness:</strong> Absence of artifactual pattern
of missingness (<a href="https://github.com/thierrygosselin/grur" class="external-link">see missing visualization</a>)</p></li>
<li><p><strong>Low genotyping error rate:</strong> see <code><a href="detect_het_outliers.html">detect_het_outliers</a></code>
and <a href="https://github.com/eriqande/whoa" class="external-link">whoa</a>. You can also use:
<code><a href="filter_hwe.html">filter_hwe</a></code> with very low threshold to discard markers with
obvious genotyping errors.</p></li>
<li><p><strong>Low heterozygosity miscall rate:</strong> see <code><a href="detect_het_outliers.html">detect_het_outliers</a></code> and
<a href="https://github.com/eriqande/whoa" class="external-link">whoa</a>.</p></li>
<li><p><strong>Absence of pattern of heterozygosity driven by missingness:</strong>
see <code><a href="detect_mixed_genomes.html">detect_mixed_genomes</a></code>. Don't use this function if the
individual heterozygosity is not under control. If your data is not uniform
inside this function, don't expect mush of <code>detect_paralogs</code>.</p></li>
<li><p>Use unfiltered dataset for fun only.</p></li>
<li><p>Use this function preferably on filtered data, check
<code><a href="filter_rad.html">filter_rad</a></code>.</p></li>
<li><p><strong>Sex-biased sampling and sex markers</strong> will generate
false-positive with this function.
Use <code><a href="sexy_markers.html">sexy_markers</a></code> function before!</p></li>
</ol></div>
    <div id="references">
    <h2>References</h2>
    <p>McKinney GJ, Waples RK, Seeb LW, Seeb JE (2017)
Paralogs are revealed by proportion of heterozygotes and deviations in read
ratios in genotyping-by-sequencing data from natural populations.
Molecular Ecology Resources, 17, 656-669.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Thierry Gosselin <a href="mailto:thierrygosselin@icloud.com">thierrygosselin@icloud.com</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># require(httr)</span></span></span>
<span class="r-in"><span><span class="co"># using the Chinook VCF dataset in dryad, from McKinney et al. 2017 paper:</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/readBin.html" class="external-link">writeBin</a></span><span class="op">(</span>object <span class="op">=</span> <span class="fu">httr</span><span class="fu">::</span><span class="fu"><a href="https://httr.r-lib.org/reference/content.html" class="external-link">content</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu">httr</span><span class="fu">::</span><span class="fu"><a href="https://httr.r-lib.org/reference/GET.html" class="external-link">GET</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>url <span class="op">=</span> <span class="st">"https://datadryad.org/bitstream/handle/10255/dryad.123464/</span></span></span>
<span class="r-in"><span><span class="st">Chinook_sequenceReads.vcf?sequence=1"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>as <span class="op">=</span> <span class="st">"raw"</span><span class="op">)</span>, con <span class="op">=</span> <span class="st">"chinook.vcf"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Import VCF and run the function</span></span></span>
<span class="r-in"><span><span class="va">chinook.paralogs</span> <span class="op">&lt;-</span> <span class="fu">radiator</span><span class="fu">::</span><span class="fu"><a href="read_vcf.html">read_vcf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="st">"chinook.vcf"</span>, vcf.stats <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>    <span class="fu">radiator</span><span class="fu">::</span><span class="fu">detect_paralogs</span><span class="op">(</span>data <span class="op">=</span> <span class="va">.</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

