<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Compute pairwise genome similarity or distance between individuals
to highlight potential duplicate individuals — detect_duplicate_genomes • radiator</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Compute pairwise genome similarity or distance between individuals
to highlight potential duplicate individuals — detect_duplicate_genomes" />

<meta property="og:description" content="The function can compute two methods
to highligh potential duplicate individuals.
distance between individuals and/or
pairwise genome similarity
" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/thierrygosselin/radiator">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compute pairwise genome similarity or distance between individuals
to highlight potential duplicate individuals</h1>
    <small class="dont-index">Source: <a href='https://github.com/thierrygosselin/radiator/blob/master/R/detect_duplicate_genomes.R'><code>R/detect_duplicate_genomes.R</code></a></small>
    <div class="hidden name"><code>detect_duplicate_genomes.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>The function can compute two methods
to highligh potential duplicate individuals.</p><ol>
<li><p>distance between individuals and/or</p></li>
<li><p>pairwise genome similarity</p></li>
</ol>
    
    </div>

    <pre class="usage"><span class='fu'>detect_duplicate_genomes</span>(<span class='no'>data</span>, <span class='kw'>interactive.filter</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>detect.duplicate.genomes</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>dup.threshold</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>distance.method</span> <span class='kw'>=</span> <span class='st'>"manhattan"</span>, <span class='kw'>genome</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>threshold.common.markers</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>blacklist.duplicates</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>parallel.core</span> <span class='kw'>=</span> <span class='kw pkg'>parallel</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/parallel/topics/detectCores'>detectCores</a></span>() - <span class='fl'>1</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>data</th>
      <td><p>(4 options) A file or object generated by radiator:</p><ul>
<li><p>tidy data</p></li>
<li><p>Genomic Data Structure (GDS)</p></li>
</ul>

<p><em>How to get GDS and tidy data ?</em>
Look into <code><a href='tidy_genomic_data.html'>tidy_genomic_data</a></code>,
<code><a href='read_vcf.html'>read_vcf</a></code> or
<code><a href='tidy_vcf.html'>tidy_vcf</a></code>.</p></td>
    </tr>
    <tr>
      <th>interactive.filter</th>
      <td><p>(optional, logical) Do you want the filtering session to
be interactive. Figures of distribution are shown before asking for filtering
thresholds.
Default: <code>interactive.filter = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>detect.duplicate.genomes</th>
      <td><p>(optional, logical) For use inside radiator pipelines.
Default: <code>detect.duplicate.genomes = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>dup.threshold</th>
      <td><p>Default: <code>dup.threshold = 0</code>. Turn off filter.</p></td>
    </tr>
    <tr>
      <th>distance.method</th>
      <td><p>(character) Depending on input data, 2 different methods
are used (give similar results):</p><ul>
<li><p>gds data The calculation is fast it's <code><a href='https://www.rdocumentation.org/packages/SNPRelate/topics/snpgdsIBS'>SNPRelate::snpgdsIBS</a></code> under
the hood.</p></li>
<li><p>tidy data The distance measure uses <code><a href='https://www.rdocumentation.org/packages/amap/topics/dist'>amap::Dist</a></code>.
This must be one of "euclidean", "maximum", "manhattan", "canberra", "binary".</p></li>
</ul><p>Using <code>distance.method = NULL</code> will not run this method.
Default: <code>distance.method = "manhattan"</code>. This is very fast
compared to the genome similarity method. It uses allele counts and the codes
are tailored for biallelic and multiallelic markers.</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>(logical) Computes pairwise genome similarity in parallel.
The proportion of the shared genotypes is averaged across shared markers between
each pairwise comparison. This method makes filtering easier because the
threshold is more intuitive with the plots produced, but it's much longer
to run, even in parallel, so better to run overnight.
Default: <code>genome = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>threshold.common.markers</th>
      <td><p>(double, optional) When using the
pairwise genome similarity approach (<code>genome = TRUE</code>), using a threshold
will filter the pairwise comparison before generating the results. This is
usefull if the number of pairwise comparisons (n*(n-1)/2) is very large and
allows to reduce the number of false positive, when missing data is involved.
e.g. 2 samples might have 100<!-- % markers in common, but if they only have -->
30<!-- % of their genotyped markers in common, this is a poor fit. -->
Default: <code>threshold.common.markers = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>blacklist.duplicates</th>
      <td><p>(optional, logical)
With <code>blacklist.duplicates = TRUE</code>, after visualization,
the user is asked to enter a threshold to filter out duplicates.
With default, <code>blacklist.duplicates = FALSE</code>, the function as no
interaction with user.</p></td>
    </tr>
    <tr>
      <th>parallel.core</th>
      <td><p>(optional) The number of core used for parallel
execution during import.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>(optional, logical) When <code>verbose = TRUE</code>
the function is a little more chatty during execution.
Default: <code>verbose = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>(optional) Advance mode that allows to pass further arguments
for fine-tuning the function. Also used for legacy arguments (see details or
special section)</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list with potentially 8 objects:</p><ul>
<li><p><code>$distance </code>: results of the distance method</p></li>
<li><p><code>$distance.stats</code>: Summary statistics of the distance method</p></li>
<li><p><code>$pairwise.genome.similarity</code>: results of the genome method</p></li>
<li><p><code>$genome.stats</code>: Summary statistics of the genome method</p></li>
<li><p><code>$violin.plot.distance</code>: violin plot showing the distribution of pairwise distances</p></li>
<li><p><code>$manhattan.plot.distance</code>: same info different visual with manhattan plot</p></li>
<li><p><code>$violin.plot.genome</code>: violin plot showing the distribution of pairwise genome similarities</p></li>
<li><p><code>$manhattan.plot.genome</code>: same info different visual with manhattan plot</p></li>
<li><p><code>$blacklist.id.similar</code>: blacklisted duplicates</p></li>
</ul>

<p>Saved in the working directory:
individuals.pairwise.dist.tsv, individuals.pairwise.distance.stats.tsv,
individuals.pairwise.genome.similarity.tsv, individuals.pairwise.genome.stats.tsv,
blackliste.id.similar.tsv, blacklist.pairs.threshold.tsv</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Strategically, run the default first (<code>distance.method</code>,
no <code>genome</code>)</p>
<p><strong><code>distance.method</code> argument is fast, but...</strong></p>
<p>you don't know if the observed comparison (close or distant)
is influenced by missing values/the number of markers in common
between the pair compared. This is something that needs to be considered.
Be suspicious of a <em>distant outlier</em> from the same pop pairwise comparison,
and similarly, be suspicious of a <em>close outlier</em> from a different pop
pairwise comparisons.</p>
<p>If there is no outlier, don't bother running the function again with
(<code>genome = TRUE</code>).</p>
    <p><strong><code>genome = TRUE</code></strong></p>
<p>The function will run slower, but...
If you see outliers with the first run, take the time to run the function
with <code>genome = TRUE</code>. Because this option is much better at detecting
duplicated individuals and it also shows the impact of <strong>missingness</strong>
or the number of <strong>shared markers</strong> between comparisons.</p>
<p><em>Your outlier duo could well be the result of one of the individual having
an extremely low number genotypes...</em></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"># NOT RUN {
# First run and simplest way (if you have the tidy tibble):
dup <- radiator::detect_duplicate_genomes(data = "wombat_tidy.rad")

# This will use by default:
distance.method = "manhattan"
genome = FALSE
#parallel.core = all my CPUs - 1

#If you need a tidy tibble: use one of radiator \code{tidy_} function or
\code{radiator::tidy_genomic_data}


# To view the manhattan plot:
dup$manhattan.plot.distance

# to view the data stats
dup.data.stats <- dup$distance.stats

# to view the data
dup.data <- dup$distance

# Based on the look of the distribution using both manhattan and boxplot,
# I can filter the dataset to highlight potential duplicates.

# To run the distance (with euclidean distance instead of the default manhattan,
# and also carry the second analysis (with the genome method):
dup <- radiator::tidy_genomic_data(
    data = "wombat.vcf",
    strata = "strata_wombat.tsv") %>%
radiator::detect_duplicate_genomes(
    data = .,
    distance.method = "euclidean",
    genome = TRUE
)

# to view the data of the genome data
dup.data <- dup$pairwise.genome.similarity

# Based on the look of the distribution using both manhattan and boxplot,
# I can filter the dataset based on 98% of identical genotype proportion,
# to highlight potential duplicates:
dup.filtered <- dplyr::filter(.data = dup.data, PROP_IDENTICAL > 0.98)

# Get the list of duplicates id
dup.list.names <- tibble::tibble(
INDIVIDUALS = unique(c(dup.filtered$ID1, dup.filtered$ID2))
)
# }
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Thierry Gosselin <a href='mailto:thierrygosselin@icloud.com'>thierrygosselin@icloud.com</a></p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Thierry Gosselin, Floriaan Devloo-Delva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

