<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Detect heterozygotes outliers and estimate miscall rate — detect_het_outliers • radiator</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Detect heterozygotes outliers and estimate miscall rate — detect_het_outliers" />
<meta property="og:description" content="Explore departure from H-W equilibrium in bi-allelic RADseq data.
Highlight excess of homozygotes present in numeros RADseq studies.
The function estimate the genotyping error rate and heterozygote miscall rate.
The model focus on heterozygotes being
incorrectly called as homozygotes. See details below for more info." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/thierrygosselin/radiator">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Detect heterozygotes outliers and estimate miscall rate</h1>
    <small class="dont-index">Source: <a href='https://github.com/thierrygosselin/radiator/blob/master/R/detect_het_outliers.R'><code>R/detect_het_outliers.R</code></a></small>
    <div class="hidden name"><code>detect_het_outliers.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Explore departure from H-W equilibrium in bi-allelic RADseq data.
Highlight excess of homozygotes present in numeros RADseq studies.
The function estimate the genotyping error rate and heterozygote miscall rate.
The model focus on heterozygotes being
incorrectly called as homozygotes. See details below for more info.</p>
    </div>

    <pre class="usage"><span class='fu'>detect_het_outliers</span>(
  <span class='no'>data</span>,
  <span class='kw'>nreps</span> <span class='kw'>=</span> <span class='fl'>2000</span>,
  <span class='kw'>burn.in</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>strata</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>filename</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>parallel.core</span> <span class='kw'>=</span> <span class='kw pkg'>parallel</span><span class='kw ns'>::</span><span class='fu'><a href='https://rdrr.io/r/parallel/detectCores.html'>detectCores</a></span>() - <span class='fl'>1</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>data</th>
      <td><p>14 options for input (<strong>diploid data only</strong>): VCFs (SNPs or Haplotypes,
to make the vcf population ready),
plink (tped, bed), stacks haplotype file, genind (library(adegenet)),
genlight (library(adegenet)), gtypes (library(strataG)), genepop, DArT,
and a data frame in long/tidy or wide format. To verify that radiator detect
your file format use <code><a href='detect_genomic_format.html'>detect_genomic_format</a></code> (see example below).
Documented in <strong>Input genomic datasets</strong> of <code><a href='tidy_genomic_data.html'>tidy_genomic_data</a></code>.</p>
<p><strong>DArT and VCF data</strong>: <span class="pkg">radiator</span> was not meant to generate alleles
and genotypes if you are using a VCF file with no genotype
(only genotype likelihood: GL or PL).
Neither is <span class="pkg">radiator</span> able to magically generate a genind object
from a SilicoDArT dataset. Please look at the first few lines of your dataset
to understand it's limit before asking raditor to convert or filter your dataset.</p></td>
    </tr>
    <tr>
      <th>nreps</th>
      <td><p>(integer, optional) The number of MCMC sweeps to do.
Default: <code>nreps = 2000</code>.</p></td>
    </tr>
    <tr>
      <th>burn.in</th>
      <td><p>(integer, optional) The number of MCMC burn-in reps.
With default, during execution, you will be asked to enter the nuber of burn-in.
For this, a plot showing the heterozygote miscall rate for all
the MCMC sweeps will be printed. This plot will help pinpoint the
number of burn-in. The remaining MCMC sweeps will be used
to average the heterozygote miscall rate.
e.g. of common value <code>burn.in = 500</code>.
With default: <code>burn.in = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>strata</th>
      <td><p>(path or object) The strata file or object.
Additional documentation is available in <code><a href='read_strata.html'>read_strata</a></code>.
Use that function to whitelist/blacklist populations/individuals.
Option to set <code>pop.levels/pop.labels</code> is also available.</p></td>
    </tr>
    <tr>
      <th>filename</th>
      <td><p>(optional, character) If <code>!is.null(blacklist.id) ||
!is.null(pop.select)</code>, the modified strata is written by default in the
working directory with date and time appended to <code>strata_radiator_filtered</code>,
to make the file unique. If you plan on writing more than 1 strata file per minute,
use this argument to supply the unique filename.
Default: <code>filename = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>parallel.core</th>
      <td><p>(optional) The number of core used for parallel
execution during import.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>(optional) To pass further arguments for fine-tuning the function.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A folder generated automatically with date and time,
the file <code>het.summary.tsv</code> contains the summary statistics. The file
<code>markers.genotypes.boundaries.pdf</code> is the plot with boundaries.
The overall genotyping and heterozygotes miscall rate is writen in the file
<code>overall_error_rate.tsv</code>.
The function also returns a list inside the global environment with
8 objects:</p>
<ol>
<li><p>input the input data, cleaned if filters were used during import.</p></li>
<li><p>outlier.summary a list with a tibble and plot of genotypes frequencies
and boundaries (also written in the folder).</p></li>
<li><p>summary.alt.allele a tibble summarizing the number of markers with:</p><ul>
<li><p>no homozygote for the alternate allele (NO_HOM_ALT)</p></li>
<li><p>no heterozygote genotype (NO_HET)</p></li>
<li><p>one homozygote for the alternate allele(ONE_HOM_ALT)</p></li>
<li><p>one heterozygote genotype (ONE_HET)</p></li>
<li><p>one homozygote for the alternate allele only (ONE_HOM_ALT_ONLY)</p></li>
<li><p>one heterozygote genotype only (ONE_HET_ONLY)</p></li>
<li><p>one homozygote for the alternate allele and one heterozygote genotype only (ONE_HOM_ALT_ONE_HET_ONLY)</p></li>
</ul></li>
<li><p>m.nreps A tibble with the heterozygote miscall rate for each MCMC replicate</p></li>
<li><p>overall.genotyping.error.rate The overall genotyping error rate</p></li>
<li><p>overall.m The overall heterozygote miscall rate</p></li>
<li><p>simmed_genos The simulated genotypes</p></li>
</ol>

<p>The statistics are summarized per population and overall,
the grouping is found in the last column called <code>POP_ID</code>.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><strong>Before using the function:</strong></p><ol>
<li><p>Don't use raw RADseq data, this function will work best with filtered data</p></li>
<li><p>Remove duplicate <code><a href='detect_duplicate_genomes.html'>detect_duplicate_genomes</a></code>.</p></li>
<li><p>Remove mixed samples <code><a href='detect_mixed_genomes.html'>detect_mixed_genomes</a></code>.</p></li>
<li><p>Look at other filters in radiator package...</p></li>
</ol>

<p><strong>During import:</strong></p>
<p>By default the function will keep only polymorphic markers and markers common
between all populations. If you supply a tidy data frame or a <code>.rad</code> file,
the function skip all the filters, pop selection, etc. It will however scan and
remove monomorphic markers automatically.</p>
<p><strong>Keep track of the data:</strong></p>
<p>Use the argument filename to write the imported (and maybe further filtered)
tidy genomic data set inside the folder. The filename will be automatically
appended <code>.rad</code> to it. This file can be used again directly inside this
function and other radiator functions. See <code><a href='read_rad.html'>read_rad</a></code>.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='no'>het.prob</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>radiator</span><span class='kw ns'>::</span><span class='fu'>detect_het_outliers</span>(
<span class='kw'>data</span> <span class='kw'>=</span> <span class='st'>"tuna.vcf"</span>, <span class='kw'>strata</span> <span class='kw'>=</span> <span class='st'>"tuna.strata.tsv"</span>, <span class='kw'>nreps</span> <span class='kw'>=</span> <span class='fl'>2000</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Thierry Gosselin, Floriaan Devloo-Delva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


