<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Heterozygosity filter — filter_het • radiator</title><!-- favicons --><link rel="icon" type="image/png" sizes="48x48" href="../favicon-48x48.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Heterozygosity filter — filter_het"><meta property="og:description" content="Observed Heterozygosity based filtering.
The filter arguments of filter_het allows you to test rapidly
if departure from realistic expectations of heterozygosity statistics
are a problem in downstream analysis.
Highlight outliers individual's observed heterozygosity for a quick
diagnostic of mixed samples or poor polymorphism discovery. The statistic is
also contrasted with missing data to help differentiate problems.
(also computed in detect_mixed_genomes)
The observed heterozygosity in the dataset: an assembly artefact,
a genotyping problem, a problem of population groupings
or a reliable signal of biological polymorphism?
Detect assembly artifact or genotyping problem (e.g. under/over-splitting loci)
by looking at marker's observed heterozygosity statistics by population
or overall.
Use haplotype or snp level statistics. When the haplotype approach
is selected, consistencies of heterozygosity statistics along the read are
highlighted.
Interactive approach help by visualizing the data before making
a decision on thresholds.

"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/radiator" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Heterozygosity filter</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/radiator/blob/HEAD/R/filter_het.R" class="external-link"><code>R/filter_het.R</code></a></small>
    <div class="hidden name"><code>filter_het.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Observed Heterozygosity based filtering.
The filter arguments of <code>filter_het</code> allows you to test rapidly
if departure from realistic expectations of heterozygosity statistics
are a problem in downstream analysis.</p><ol><li><p>Highlight outliers individual's observed heterozygosity for a quick
diagnostic of mixed samples or poor polymorphism discovery. The statistic is
also contrasted with missing data to help differentiate problems.
(also computed in <a href="detect_mixed_genomes.html">detect_mixed_genomes</a>)</p></li>
<li><p>The observed heterozygosity in the dataset: an assembly artefact,
a genotyping problem, a problem of population groupings
or a reliable signal of biological polymorphism?
Detect assembly artifact or genotyping problem (e.g. under/over-splitting loci)
by looking at marker's observed heterozygosity statistics by population
or overall.</p></li>
<li><p>Use haplotype or snp level statistics. When the haplotype approach
is selected, consistencies of heterozygosity statistics along the read are
highlighted.</p></li>
<li><p>Interactive approach help by visualizing the data before making
a decision on thresholds.</p></li>
</ol></div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">filter_het</span><span class="op">(</span></span>
<span>  interactive.filter <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">data</span>,</span>
<span>  strata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ind.heterozygosity.threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  het.approach <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SNP"</span>, <span class="st">"overall"</span><span class="op">)</span>,</span>
<span>  het.threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  het.dif.threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  outlier.pop.threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  helper.tables <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  coverage.info <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  filename <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-interactive-filter">interactive.filter<a class="anchor" aria-label="anchor" href="#arg-interactive-filter"></a></dt>
<dd><p>(optional, logical) Do you want the filtering session to
be interactive. With the default, the user is asked to see figures of
distribution before making decisions for filtering with heterozygosity statistics.
Default: <code>interactive.filter == TRUE</code>.</p></dd>


<dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>14 options for input (<strong>diploid data only</strong>): VCFs (SNPs or Haplotypes,
to make the vcf population ready),
plink (tped, bed), stacks haplotype file, genind (library(adegenet)),
genlight (library(adegenet)), gtypes (library(strataG)), genepop, DArT,
and a data frame in long/tidy or wide format. To verify that radiator detect
your file format use <code><a href="detect_genomic_format.html">detect_genomic_format</a></code> (see example below).
Documented in <strong>Input genomic datasets</strong> of <code><a href="tidy_genomic_data.html">tidy_genomic_data</a></code>.</p>
<p><strong>DArT and VCF data</strong>: <span class="pkg">radiator</span> was not meant to generate alleles
and genotypes if you are using a VCF file with no genotype
(only genotype likelihood: GL or PL).
Neither is <span class="pkg">radiator</span> able to magically generate a genind object
from a SilicoDArT dataset. Please look at the first few lines of your dataset
to understand it's limit before asking raditor to convert or filter your dataset.</p></dd>


<dt id="arg-strata">strata<a class="anchor" aria-label="anchor" href="#arg-strata"></a></dt>
<dd><p>(optional)
The strata file is a tab delimited file with a minimum of 2 columns headers:
<code>INDIVIDUALS</code> and <code>STRATA</code>. Documented in <code><a href="read_strata.html">read_strata</a></code>.
DArT data: a third column <code>TARGET_ID</code> is required.
Documented on <code><a href="read_dart.html">read_dart</a></code>. Also use the strata read function to
blacklist individuals.
Default: <code>strata = NULL</code>.</p></dd>


<dt id="arg-ind-heterozygosity-threshold">ind.heterozygosity.threshold<a class="anchor" aria-label="anchor" href="#arg-ind-heterozygosity-threshold"></a></dt>
<dd><p>(string, double, optional)
Blacklist individuals based on observed heterozygosity (averaged across markers).</p>

<p>The string contains 2 thresholds values (min and max).
The values are proportions (0 to 1), where 0 turns off the min threshold and
1 turns off the max threshold.
Individuals with mean observed heterozygosity higher (&gt;) or lower (&lt;)
than the thresholds will be blacklisted.</p>
<p>Default: <code>ind.heterozygosity.threshold = NULL</code> will turn off completely
the filter and the function will only output the plots and table of heterozygosity.</p></dd>


<dt id="arg-het-approach">het.approach<a class="anchor" aria-label="anchor" href="#arg-het-approach"></a></dt>
<dd><p>(Character string). First value, <code>"SNP"</code> or
<code>"haplotype"</code>. The haplotype approach considers the statistic
consistency on the read (locus/haplotype). The major difference:
the haplotype approach results in blacklisting the entire locus/haplotype
with all the SNPs on the read.
With the SNP approach, SNPs are independently analyzed and blacklisted.
The second value, will use the statistics by population <code>"pop"</code> or
will consider the data overall, as 1 large group <code>"overall"</code>.
Default: <code>het.approach = c("SNP", "overall")</code>.</p></dd>


<dt id="arg-het-threshold">het.threshold<a class="anchor" aria-label="anchor" href="#arg-het-threshold"></a></dt>
<dd><p>Number Biallelic markers usually max 0.5. But departure
from this value is common. The higher the proportion threshold, the more relaxed
the filter is. With default there is no filtering.
Default: <code>het.threshold = 1</code>.</p></dd>


<dt id="arg-het-dif-threshold">het.dif.threshold<a class="anchor" aria-label="anchor" href="#arg-het-dif-threshold"></a></dt>
<dd><p>Number (0 - 1). For <code>het.approach = "haplotype"</code> only.
You can set a threshold for the difference in het along your read.
Set the number your willing to tolerate on the same read/haplotype.
e.g. if you have 2 SNP on a read/haplotype and on as a het of 0.9 and the other
0.1 and you set <code>het.dif.threshold = 0.3</code>, this markers will be blacklisted.
You should strive to have similar statistics along short read like RAD.
The higher the proportion threshold, the more relaxed the filter is.
With default, there is no filtering and all the range of differences are allowed.
Default: <code>het.dif.threshold = 1</code>.</p></dd>


<dt id="arg-outlier-pop-threshold">outlier.pop.threshold<a class="anchor" aria-label="anchor" href="#arg-outlier-pop-threshold"></a></dt>
<dd><p>(integer, optional) Useful to incorporate problematic
populations dragging down polymorphism discovery, but still wanted for analysis.
Use this threshold to allow variance in the number of populations passing
the thresholds described above.
e.g. with <code>outlier.pop.threshold = 2</code>,
you tolerate a maximum of 2 populations failing the
<code>het.threshold</code> and/or <code>het.dif.threshold</code>.
Manage outlier markers, individuals and populations
downstream with blacklists and whitelists produced by the function.
Default: <code>outlier.pop.threshold = 1</code>. See details for more info.</p></dd>


<dt id="arg-helper-tables">helper.tables<a class="anchor" aria-label="anchor" href="#arg-helper-tables"></a></dt>
<dd><p>(logical) Output tables that show
the number of markers blacklisted or whitelisted based on a series of
automatic thresholds to guide decisions. When <code>interactive.filter == TRUE</code>,
helper tables are written to the directory.
Default: <code>helper.tables = FALSE</code>.</p></dd>


<dt id="arg-coverage-info">coverage.info<a class="anchor" aria-label="anchor" href="#arg-coverage-info"></a></dt>
<dd><p>(optional, logical) Use
<code>coverage.info = TRUE</code>, if you want to visualize
the relationshio between locus coverage and locus observed heterozygosity
statistics. Coverage information is required (e.g. in a vcf file...).
Default: <code>coverage.info = FALSE</code>.</p></dd>


<dt id="arg-filename">filename<a class="anchor" aria-label="anchor" href="#arg-filename"></a></dt>
<dd><p>(optional) The function uses <code><a href="http://www.fstpackage.org/reference/write_fst.html" class="external-link">write.fst</a></code>,
to write the tidy data frame in
the folder created in the working directory. The file extension appended to
the <code>filename</code> provided is <code>.rad</code>.
Default: <code>filename = NULL</code>.</p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>(optional) The number of core used for parallel
execution during import.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>(optional) To pass further arguments for fine-tuning the function.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>The function returns inside the global environment a list with
15 objects, the objects names are found by using <code>names(your.list.name)</code>:</p>
<ol><li><p>filtered tidy data frame: <code>$tidy.filtered.het</code></p></li>
<li><p>whitelist of markers:<code>$whitelist.markers</code></p></li>
<li><p>the strata:<code>$strata</code></p></li>
<li><p>the filters parameters used:<code>$filters.parameters</code></p></li>
<li><p>the individual's heterozigosity:<code>$individual.heterozigosity</code></p></li>
<li><p>the heterozygosity statistics per populations and overall:<code>$heterozygosity.statistics</code></p></li>
<li><p>the blacklisted individuals based on the individual's heterozigosity:<code>$blacklist.ind.het</code></p></li>
<li><p>a list containing the helper tables:<code>$helper.table.het</code></p></li>
<li><p>the boxplot of individual observed heterozygosity:<code>$individual.heterozygosity.boxplot</code></p></li>
<li><p>the manhattan plot of individual heterozygosity:<code>$individual.heterozygosity.manhattan.plot</code></p></li>
<li><p>the boxplot of observed heterozygosity averaged across markers and pop:<code>$markers.pop.heterozygosity.boxplot</code></p></li>
<li><p>the density plot of observed heterozygosity averaged across markers and pop:<code>$markers.pop.heterozygosity.density.plot</code></p></li>
<li><p>the manhattan plot of observed heterozygosity averaged across markers and pop:<code>$markers.pop.heterozygosity.manhattan.plot</code></p></li>
<li><p>the relationship between the number of SNPs on a locus and locus observed heterozygosity statistics:<code>$snp.per.locus.het.plot</code></p></li>
<li><p>the relationship between locus coverage (read depth) and locus observed heterozygosity statistics:<code>$het.cov.plot</code></p></li>
<li><p>whitelist of markers:<code>$whitelist.markers</code></p></li>
<li><p>whitelist of markers:<code>$whitelist.markers</code></p></li>
</ol><p>In the working directory, output is conditional to <code>interactive.filter</code> argument</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><strong>Interactive version</strong></p>
<p>There are 4 steps in the interactive version to visualize and filter
the data based heterozygosity statistics:</p>
<p>Step 1. Individual's observed heterozygosity: outliers that might represent mixed samples
Step 2. Blacklist outliers based on a proportion threshold of mean observed heterozygosity
Step 3. Observed heterozygosity statistics per populations and overall
Step 4: Blacklist markers based on observed heterozygosity</p>


<p><strong>outlier.pop.threshold</strong></p>
<p>If your a regular radiator user, you've seen the <code>pop.num.threshold</code>.
<code>outlier.pop.threshold</code>, is different and requires more thinking,
because the number of populations genotyped potentially vary across markers,
which makes the use of <code>pop.num.threshold</code> less optimal.
e.g. If only 4 populations out of 10 are genotyped, for a marker you want
to keep, using <code>pop.num.threshold = 0.6</code> will lead to unwanted results
and inconsistensis.
It's easier to tolerate outliers with this new approach:
<code>outlier.pop.threshold = 2</code>.</p>

<p><strong>Individual observed heterozygosity (averaged across markers):</strong>
To help discard an individual based on his observed heterozygosity
(averaged across markers),
use the manhanttan plot to:</p><ol><li><p>contrast the individual with population and overall samples.</p></li>
<li><p>visualize the impact of missigness information (based on population or
overall number of markers) and the individual observed heterozygosity. The
larger the point, the more missing genotypes.</p></li>
</ol><p><strong>Outlier above average:</strong></p><ul><li><p>potentially represent two samples mixed together (action: blacklist), or...</p></li>
<li><p>a sample with more sequecing effort (point size small): did you merge your replicates fq files ? (action : keep and monitor)</p></li>
<li><p>a sample with poor sequencing effort (point size large) where the genotyped markers are
all heterozygotes, verify this with missingness (action: discard)</p></li>
</ul><p>In all cases, if there is no bias in the population sequencing effort,
the size of the point will usually be "average" based on the population or
overall number of markers.</p>

<p>You can visualize individual observed heterozygosity, choose thresholds and
then visualize, choose thresholds and filter markers based on observed
heterozygosity in one run with: <span class="pkg">radiator</span> <code>filter_het</code>.</p>
<p><strong>Outlier below average:</strong></p><ul><li><p>A point with a size larger than the population or overall average (= lots of missing):
the poor polymorphism discovery of the sample is probably the result of bad
DNA quality, a bias in sequencing effort, etc. (action: blacklist)</p></li>
<li><p>A point with a size that looks average (not much missing):
this sample requires more attention (action: blacklist) and conduct more tests.
e.g. for biallelic data, look for coverage imbalance between ALT/REF allele.
At this point you need to distinguish between an artifact of poor polymorphism discovery
or a biological reason (highly inbred individual, etc.).</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="plot_density_distribution_het.html">plot_density_distribution_het</a></p></div>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

