<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>MAC filter — filter_mac • radiator</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="MAC filter — filter_mac" />

<meta property="og:description" content="The Minor Allele Count function
remove/blacklist markers based on Minor/Alternate Allele Count (MAC). Use it
to remove noise, sequencing errors or low polymorphism markers. Some
analysis performs better with the full spectrum of allele frequency, so careful
with high threshold that inevitably results in biaises.
Filter targets: Marker's alternate allele(s)
Statistics: count of the allele per markers. This filter as NO concept
of strata/populations. It is computed globally." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/thierrygosselin/radiator">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>MAC filter</h1>
    <small class="dont-index">Source: <a href='https://github.com/thierrygosselin/radiator/blob/master/R/filter_mac.R'><code>R/filter_mac.R</code></a></small>
    <div class="hidden name"><code>filter_mac.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>The Minor Allele Count function
remove/blacklist markers based on Minor/Alternate Allele Count (MAC). Use it
to remove noise, sequencing errors or low polymorphism markers. Some
analysis performs better with the full spectrum of allele frequency, so careful
with high threshold that inevitably results in biaises.</p>
<p><strong>Filter targets</strong>: Marker's alternate allele(s)</p>
<p><strong>Statistics</strong>: count of the allele per markers. This filter as NO concept
of strata/populations. It is computed globally.</p>
    
    </div>

    <pre class="usage"><span class='fu'>filter_mac</span>(<span class='no'>data</span>, <span class='kw'>interactive.filter</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>filter.mac</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>filename</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>parallel.core</span> <span class='kw'>=</span> <span class='kw pkg'>parallel</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/parallel/topics/detectCores'>detectCores</a></span>() - <span class='fl'>1</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>data</th>
      <td><p>(4 options) A file or object generated by radiator:</p><ul>
<li><p>tidy data</p></li>
<li><p>Genomic Data Structure (GDS)</p></li>
</ul>

<p><em>How to get GDS and tidy data ?</em>
Look into <code><a href='tidy_genomic_data.html'>tidy_genomic_data</a></code>,
<code><a href='read_vcf.html'>read_vcf</a></code> or
<code><a href='tidy_vcf.html'>tidy_vcf</a></code>.</p></td>
    </tr>
    <tr>
      <th>interactive.filter</th>
      <td><p>(optional, logical) Do you want the filtering session to
be interactive. Figures of distribution are shown before asking for filtering
thresholds.
Default: <code>interactive.filter = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>filter.mac</th>
      <td><p>(optional, integer) The number of alternate (minor) allele
to keep a SNP (see details). e.g. <code>filter.mac = 3</code>. This argument as no
concept of locus, local or strata or pop. It's applied globally, by SNPs.
Default: <code>filter.mac = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>filename</th>
      <td><p>(optional, character) Write to folder the MAF filtered
tidy dataset. The name will be appended <code>.rad</code>. With default, the filtered
data is only in the global environment.
Default: <code>filename = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>parallel.core</th>
      <td><p>(optional) The number of core used for parallel
execution during import.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>(optional, logical) When <code>verbose = TRUE</code>
the function is a little more chatty during execution.
Default: <code>verbose = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>(optional) Advance mode that allows to pass further arguments
for fine-tuning the function. Also used for legacy arguments (see details or
special section)</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>With <code>interactive.filter = FALSE</code>, a list in the global environment,
with 7 objects:</p><ol>
<li><p>$tidy.filtered.mac</p></li>
<li><p>$whitelist.markers</p></li>
<li><p>$blacklist.markers</p></li>
<li><p>$mac.data</p></li>
<li><p>$filters.parameters</p></li>
</ol>

<p>With <code>interactive.filter = TRUE</code>, a list with 4 additionnal objects are generated.</p><ol>
<li><p>$distribution.mac.global</p></li>
<li><p>$distribution.mac.local</p></li>
<li><p>$mac.global.summary</p></li>
<li><p>$mac.helper.table</p></li>
</ol>

<p><strong>mac.helper.table:</strong></p>
<p>First and second variables (in columns) represents POP_ID and sample size (n).</p>
<p>The last 2 rows are the local MAF (suggested based on the lowest pop value)
and the TOTAL/GLOBAL observations.</p>
<p>Columns starting with ALT are the variable corresponding
to the number of alternative (ALT) allele (ranging from 1 to 20).
The observations in the ALT allele variable columns are the local (for the pop)
and global (last row) MAF of your dataset.
e.g. ALT_3 can potentially represent 3 heterozygote individuals with
the ALT allele or 1 homozygote individuals for the ALT allele and
1 heterozygote individual. And so on...</p>
    
    <h2 class="hasAnchor" id="mac-or-maf-"><a class="anchor" href="#mac-or-maf-"></a>MAC or MAF ?</h2>

    
    <p>Using count or frequency to remove a SNPs ? The preferred choice in radiator
as changed from frequency to count, because we think the filtering should not
alter the spectrum and this is only achieved if the same criteria is applied
for each SNP.</p>
    <p>Even small differences in missing data between RADseq markers
generates differences in MAF frequency thresholds applied.</p>
<p>Example with a datset consisting of N = 36 individuals and 3 SNPs
with varying level of missing genotypes:</p>
<ul>
<li><p><code>SNP number : number samples genotypes : REF/ALT counts</code></p></li>
<li><p>SNP1 : 36 : 69/3</p></li>
<li><p>SNP2 : 30 : 65/3</p></li>
<li><p>SNP3 : 24 : 45/3</p></li>
</ul>
    <p>Each SNPs have the same alternate allele count, corresponding to
2 individuals with the polymorphism: 1 homozygote + 1 heterozygote.
Applying a MAF threshold of
0.05 would mean that SNP3 would be blacklisted
(<code>24 * 2 * 0.05 = 2.4 alt alleles required to pass</code>).</p>
    <p><strong>Using count instead of frequency allows each RADseq markers,
with varying missing data, to be treated equally.</strong></p>
    
    <h2 class="hasAnchor" id="advance-mode"><a class="anchor" href="#advance-mode"></a>Advance mode</h2>

    
    <p><em>dots-dots-dots ...</em> allows to pass several arguments for fine-tuning the function:</p><ol>
<li><p><code>filter.common.markers</code> (optional, logical).
Default: <code>filter.common.markers = FALSE</code>,
Documented in <code><a href='filter_common_markers.html'>filter_common_markers</a></code>.</p></li>
<li><p><code>filter.monomorphic</code> (logical, optional) Should the monomorphic
markers present in the dataset be filtered out ?
Default: <code>filter.monomorphic = TRUE</code>.
Documented in <code><a href='filter_monomorphic.html'>filter_monomorphic</a></code>.</p></li>
<li><p><code>path.folder</code>: to write ouput in a specific path
(used internally in radiator).
Default: <code>path.folder = getwd()</code>.
If the supplied directory doesn't exist, it's created.</p></li>
</ol>
    
    <h2 class="hasAnchor" id="interactive-version"><a class="anchor" href="#interactive-version"></a>Interactive version</h2>

    
    <p>To help choose a threshold for the local and global MAF
use the interactive version.</p>
<p>2 steps in the interactive version:</p>
<p>Step 1. Visualization and helper table.</p>
<p>Step 2. Filtering markers based on MAC</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='co'># The minumum</span>
<span class='no'>mac</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>radiator</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/radiator/topics/filter_mac'>filter_mac</a></span>(<span class='kw'>data</span> <span class='kw'>=</span> <span class='no'>turtle.tidy.data</span>)
<span class='co'># This will use the default: interactive version,</span>
<span class='co'># a list is created and to view the filtered tidy data:</span>
<span class='no'>mac.tidy.data</span> <span class='kw'>&lt;-</span> <span class='no'>mac</span>$<span class='no'>tidy.filtered.mac</span>

<span class='co'># No user interaction</span>

<span class='co'># Using filter.mac = 4, is a good practice to remove mostly sequencing errors</span>
<span class='co'># and assembly artifacts, because it requires the markers to be genotyped in</span>
<span class='co'># 4 heterozygote individuals or 2 homozygote individuals for the alternate</span>
<span class='co'># allele or 2 heterozygote individuals and 1 homozygote individual for the</span>
<span class='co'># alternate allele. Overall, never less than 2 indiduals are required.</span>

<span class='no'>mac</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>radiator</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/radiator/topics/filter_mac'>filter_mac</a></span>(
        <span class='kw'>data</span> <span class='kw'>=</span> <span class='no'>turtle.gds</span>, <span class='co'># using gds object</span>
        <span class='kw'>filter.mac</span> <span class='kw'>=</span> <span class='fl'>4</span>,
        <span class='kw'>filename</span> <span class='kw'>=</span> <span class='st'>"turtle.mac"</span>)

<span class='co'># This will remove monomorphic markers (by default filter.monomorphic = TRUE)</span>
<span class='co'># The filtered data will be written in the directory under the name:</span>
<span class='co'># turtle.maf.rad</span>
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#mac-or-maf-">MAC or MAF ?</a></li>

      <li><a href="#advance-mode">Advance mode</a></li>

      <li><a href="#interactive-version">Interactive version</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Thierry Gosselin <a href='mailto:thierrygosselin@icloud.com'>thierrygosselin@icloud.com</a></p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Thierry Gosselin, Floriaan Devloo-Delva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

