<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Filter markers based on Hardy-Weinberg Equilibrium — filter_hwe • radiator</title><!-- favicons --><link rel="icon" type="image/png" sizes="48x48" href="../favicon-48x48.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Filter markers based on Hardy-Weinberg Equilibrium — filter_hwe"><meta property="og:description" content="Testing markers for Hardy-Weinberg proportions is a valuable tool
for the analysis and quality control of RADseq datasets.
HWE can highligh genotyping errors, presence of null alleles,
sequence duplication, copy number variation and other sequencing problems
related to read depth.
This function is designed for
bi-allelic markers and uses the exact test from the package
HardyWeinberg. The function is speedy because it uses the
C++ code developed by Christopher Chang and also available in PLINK.
The p-value generated by the function is the mid p-value. It's computed
as half the probability of the current sample + the probabilities of all
samples that are more extreme (see references below). Several output are
generated to help users filter the data (see details).
sampling sites vs well defined populations: be careful what strata
you're investigating and adjust filtering threshold accordinghly for
downstream analysis. If you're still at the populations discovery steps,
markers under HWD are totally normal.
Prior to HW filtering, I highly recommend removing outlier individuals,
filtering coverage and genotype likelihood (see details)."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/radiator" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Filter markers based on Hardy-Weinberg Equilibrium</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/radiator/blob/HEAD/R/filter_hwe.R" class="external-link"><code>R/filter_hwe.R</code></a></small>
    <div class="hidden name"><code>filter_hwe.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Testing markers for Hardy-Weinberg proportions is a valuable tool
for the analysis and quality control of RADseq datasets.
HWE can highligh genotyping errors, presence of null alleles,
sequence duplication, copy number variation and other sequencing problems
related to read depth.
This function is designed for
<strong>bi-allelic markers</strong> and uses the exact test from the package
HardyWeinberg. The function is speedy because it uses the
C++ code developed by Christopher Chang and also available in PLINK.
The p-value generated by the function is the mid p-value. It's computed
as half the probability of the current sample + the probabilities of all
samples that are more extreme (see references below). Several output are
generated to help users filter the data (see details).</p>
<p><strong>sampling sites vs well defined populations:</strong> be careful what strata
you're investigating and adjust filtering threshold accordinghly for
downstream analysis. If you're still at the populations discovery steps,
markers under HWD are totally normal.</p>
<p>Prior to HW filtering, I highly recommend removing outlier individuals,
filtering coverage and genotype likelihood (see details).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">filter_hwe</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  interactive.filter <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filter.hwe <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  strata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  hw.pop.threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  midp.threshold <span class="op">=</span> <span class="fl">4L</span>,</span>
<span>  filename <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>A tidy data frame object in the global environment or
a tidy data frame in wide or long format in the working directory.
<em>How to get a tidy data frame ?</em>
Look into <span class="pkg">radiator</span> <code><a href="tidy_genomic_data.html">tidy_genomic_data</a></code>.</p></dd>


<dt id="arg-interactive-filter">interactive.filter<a class="anchor" aria-label="anchor" href="#arg-interactive-filter"></a></dt>
<dd><p>(optional, logical) Do you want the filtering session to
be interactive. Figures of distribution are shown before asking for filtering
thresholds.
Default: <code>interactive.filter = TRUE</code>.</p></dd>


<dt id="arg-filter-hwe">filter.hwe<a class="anchor" aria-label="anchor" href="#arg-filter-hwe"></a></dt>
<dd><p>(optional, logical) Used inside radiator pipeline.
Default: <code>filter.hwe = TRUE</code>.</p></dd>


<dt id="arg-strata">strata<a class="anchor" aria-label="anchor" href="#arg-strata"></a></dt>
<dd><p>(optional)
The strata file is a tab delimited file with a minimum of 2 columns headers:
<code>INDIVIDUALS</code> and <code>STRATA</code>. Documented in <code><a href="read_strata.html">read_strata</a></code>.
DArT data: a third column <code>TARGET_ID</code> is required.
Documented on <code><a href="read_dart.html">read_dart</a></code>. Also use the strata read function to
blacklist individuals.
Default: <code>strata = NULL</code>.</p></dd>


<dt id="arg-hw-pop-threshold">hw.pop.threshold<a class="anchor" aria-label="anchor" href="#arg-hw-pop-threshold"></a></dt>
<dd><p>(integer, optional)
Remove markers that have a certain number of pops in Hardy-Weinberg
disequilibrium.
With default, all populations in dataset need to be in HWD before discarding
the marker.
Default: <code>hw.pop.threshold = NULL</code>.</p></dd>


<dt id="arg-midp-threshold">midp.threshold<a class="anchor" aria-label="anchor" href="#arg-midp-threshold"></a></dt>
<dd><p>(integer, optional)
By default the function generates blacklists/whitelists of markers and
filtered tidy datasets for the 5 mid p-value.
However, to get a final filtered object associated with the output of the
function, user need to choose one
of the 5 mid p-value:</p><ul><li><p><code>1</code> = 0.05 (*)</p></li>
<li><p><code>2</code> = 0.001 (**)</p></li>
<li><p><code>3</code> = 0.0001 (***)</p></li>
<li><p><code>4</code> = 0.0001 (****)</p></li>
<li><p><code>5</code> = 0.00001 (*****)</p></li>
</ul><p>.
With default, a very conservative mid p-value threshold is selected.
Default: <code>midp.threshold = 4</code>.</p></dd>


<dt id="arg-filename">filename<a class="anchor" aria-label="anchor" href="#arg-filename"></a></dt>
<dd><p>(optional) The function uses <code><a href="http://www.fstpackage.org/reference/write_fst.html" class="external-link">write.fst</a></code>,
to write the tidy data frame in
the folder created in the working directory. The file extension appended to
the <code>filename</code> provided is <code>.rad</code>.
Default: <code>filename = NULL</code>.</p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>(optional) The number of core used for parallel
execution during import.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>(optional, logical) When <code>verbose = TRUE</code>
the function is a little more chatty during execution.
Default: <code>verbose = TRUE</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>(optional) Advance mode that allows to pass further arguments
for fine-tuning the function. Also used for legacy arguments (see details or
special section)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list in the global environment objects:</p><ol><li><p>$path.folder: the path to the folder generated.</p></li>
<li><p>$hw.pop.threshold: the number of populations tolerated to be in HWD before
blacklisting the markers.</p></li>
<li><p>$plot.hwd.thresholds: useful figure that highlight the number of markers
blacklisted based on the number of populations in HWD and mid p-value thresholds.</p></li>
<li><p>$plot.tern: ternary plot of markers (currently unavailable until ggtern is updated).</p></li>
<li><p>$hw.manhattan: manhattan plot of markers in Hardy-Weinberg disequilibrium.</p></li>
<li><p>$hwe.pop.sum: a summary tibble with populations, number of markers in total,
number of markers monomorphic for the populations,
number of markers in Hardy-Weinberg Equilibrium (HWE),
number of markers in Hardy-Weinberg Dquilibrium (HWD) with all the different
mid p-values observed on the data.</p></li>
<li><p>$midp.threshold: the mid p-value threshold chosen for the final dataset (next)</p></li>
<li><p>$tidy.hw.filtered: the final filtered dataset (oter datasets <code>.rad</code>
are generated automatically by the function, check the folder)</p></li>
</ol><p>Written in the folder:</p><ol><li><p>genotypes.summary.tsv: A tibble with these columns:
<code>MARKERS, STRATA, HET, HOM_ALT, HOM_REF, MISSING, N,
FREQ_ALT, FREQ_REF, FREQ_HET, FREQ_HOM_REF_O, FREQ_HET_O, FREQ_HOM_ALT_O,
FREQ_HOM_REF_E, FREQ_HET_E, FREQ_HOM_ALT_E, N_HOM_REF_EXP,
N_HET_EXP, N_HOM_ALT_EXP, HOM_REF_Z_SCORE, HOM_HET_Z_SCORE,
HOM_ALT_Z_SCORE, READ_DEPTH</code></p></li>
<li><p>hw.pop.sum.tsv: a summary tibble with populations, number of markers in total,
number of markers monomorphic for the populations,
number of markers in Hardy-Weinberg Equilibrium (HWE),
number of markers in Hardy-Weinberg Dquilibrium (HWD) with all the different
mid p-values observed on the data.</p></li>
<li><p>hwd.helper.table.tsv: useful tibble that highlight the number of markers
blacklisted based on the number of populations in HWD and mid p-value thresholds.</p></li>
<li><p>hwd.plot.blacklist.markers.pdf: useful figure that highlight the number of markers
blacklisted based on the number of populations in HWD and mid p-value thresholds.</p></li>
<li><p>hwe.manhattan.plot.pdf: manhattan plot of markers in Hardy-Weinberg disequilibrium.</p></li>
<li><p>tidy.filtered.hwe.xxx.mid.p.value.xxx.hw.pop.threshold.rad: several
tidy dataset filtered with different mid p value and populations in HWD thresholds</p></li>
<li><p>whitelist.markers.hwe.xxx.mid.p.value.xxx.hw.pop.threshold.tsv: several
whitelist of  markers with different mid p value and populations in HWD thresholds</p></li>
<li><p>blacklist.markers.hwd.xxx.mid.p.value.xxx.hw.pop.threshold.tsv: several
blacklist of markers with different mid p value and populations in HWD thresholds</p></li>
</ol></div>
    <div id="details">
    <h2>Details</h2>
    <p><strong>Interactive version</strong></p>
<p>The user is asked to look at figures before choosing filter thresholds.</p>

<p><strong>HWE threshold</strong></p>
<p>I recommend starting with a low threshold.
Serious genotyping errors will generate extreme p-values (e.g. 1e-50),
which are detected by any reasonable configuration of this test,
while various life-history caracteristics will deflate/inflate
Hardy-Weinberg equilibrium.
Consequently, it's dangerous to choose a threshold that filters out too many
markers.</p>
<p><strong>strategies:</strong>
Disk space is cheap! Consequently, the function will automatically generate
several blacklists/whitelists of markers and
filtered tidy data (in the directory)
based on the <code>hw.pop.threshold</code> for 5 groups of mid p-values:</p><ul><li><p>1: MID_P_VALUE &lt;= 0.00001: *****</p></li>
<li><p>2: MID_P_VALUE &lt;= 0.0001: ****</p></li>
<li><p>3: MID_P_VALUE &lt;= 0.001: ***</p></li>
<li><p>4: MID_P_VALUE &lt;= 0.01: **</p></li>
<li><p>5: MID_P_VALUE &lt;= 0.05: *</p></li>
</ul><p>Test the sensitivity of downstream analysis and delete unwanted datasets.</p>
<p><strong>missing data</strong></p>
<p>The mid-p adjustment tends to bring the null rejection rate in line with the
nominal p-value, and also reduces the filter's tendency to favor retention
of SNPs with missing data (Graffelman and Moreno, 2013, Purcell et al., 2007).</p>
<p>If pattern of missing data is present in the dataset, or when missing data
accross markers vary by more than 0.10, you should not apply a single mid-p-value
threshold accross markers.</p>
<p><strong>Read depth, pooling lanes/chips and weird pattern of individual heterozygosity</strong></p>
<p>Because of read depth, heterozygote deficiency is usually observed in RADseq data,
but if sequencing lanes/chips were combined to generate individuals with more
coverage, the situation will likely be the reverse: heterozygote excess.
If lanes/chips pooling was used or if highly variable sequencing coverage is
observed between individuals and/or markers, there's a couple qc and filtering
steps you should do before conducting HWE filtering.</p>
<ul><li><p>run <span class="pkg">radiator</span> <code><a href="detect_mixed_genomes.html">detect_mixed_genomes</a></code> and
<code><a href="detect_het_outliers.html">detect_het_outliers</a></code> to highlight
outlier individuals with potential heterozygosity problems and get an idea of
the genotyping and heterozygote miscall rate</p></li>
<li><p>I recommend normalizing the data before <em>de novo</em> assembly, if this
is not possible...</p></li>
<li><p>use radiator::filter_rad or a more chirurgical approach to
coverage and genotype likelihood with radiator::filter_rad.</p></li>
</ul><p><strong>Permutation test</strong>
Hardy-Weinberg equilibrium refers to the statistical independence of alleles
within individuals. This independence can also be assessed by permutation test
inside the HardyWeinberg package of Jan Graffelman. To filter out markers with
genotyping problems the approach provided in this function is enough.</p>
<p><strong>Power test for HWE</strong>
Based on allele count/frequency and sample size. This function is longer to
generate for each markers and is on my todo list to include it in this filter.</p>

<p><strong>Markers under selection and genome scans:</strong></p>
<p>Scared of deleting those precious markers or that the filter might interfere
with genome scan analysis/detection ? Don't be. Your markers or analysis is no
good if it's done on bad data... Test the sensitivity of your downstream
analysis with the datasets generated with the different thresholds.</p>
    </div>
    <div id="note">
    <h2>Note</h2>
    <p><strong>Hardy-Weinberg assumptions (refresh):</strong></p><ol><li><p>Diploid organisms</p></li>
<li><p>Only reproduction sexual occurs</p></li>
<li><p>Generations are non-overlapping</p></li>
<li><p>Mating is random</p></li>
<li><p>Population size is infinitely large</p></li>
<li><p>Allele frequencies are equal in the sexes</p></li>
<li><p>Migration, Mutation and Selection are negligible</p></li>
</ol></div>
    <div id="references">
    <h2>References</h2>
    <p>Weir, B.S. (1996) Genetic data analysis II. Sinauer Associates,
Massachusetts. See Chapter3.</p>
<p>Wigginton, J.E., Cutler, D.J. and Abecasis, G.R. (2005)
A note on exact tests of Hardy-Weinberg equilibrium,
American Journal of Human Genetics (76) pp. 887-893.</p>
<p>Purcell et al. (2007) PLINK: A Toolset for Whole-Genome Association
and Population-Based Linkage Analysis.
American Journal of Human Genetics 81(3) pp. 559-575.</p>
<p>Graffelman, J. and Moreno, V. (2013) The mid p-value in exact
tests for Hardy-Weinberg equilibrium, Statistical Applications in Genetics
and Molecular Biology 12(4) pp. 433-448.</p>
<p>Graffelman J, Jain D, Weir B (2017)
A genome-wide study of Hardy-Weinberg equilibrium with next generation
sequence data. Human Genetics, 136, 727-741.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://www.r-project.org" class="external-link">HardyWeinberg</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://thierrygosselin.github.io/radiator/" class="external-link">radiator</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># for the interactive version (recommended)</span></span></span>
<span class="r-in"><span><span class="va">turtle.pop</span> <span class="op">&lt;-</span> <span class="fu">radiator</span><span class="fu">::</span><span class="fu">filter_hwe</span><span class="op">(</span></span></span>
<span class="r-in"><span>   data <span class="op">=</span> <span class="st">"turtle.vcf"</span>,</span></span>
<span class="r-in"><span>   strata <span class="op">=</span> <span class="st">"turtle.strata.tsv"</span>,</span></span>
<span class="r-in"><span>   filename <span class="op">=</span> <span class="st">"hwe.turtle"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

