<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>GBS/RADseq short and long distance linkage disequilibrium pruning — filter_ld • radiator</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="GBS/RADseq short and long distance linkage disequilibrium pruning — filter_ld" />

<meta property="og:description" content="SNP short and long distance linkage disequilibrium pruning.
What sets appart radiator LD pruning is the RADseq data tailored arguments:
minimize short linkage disequilibrium (LD):
5 values available for filter.short.ld argument (see below).
reduce long distance LD:
Long distance LD pruning is usually advised to avoid capturing the variance LD
in PCA analysis.
Use the argument filter.long.ld with values between 0.7 and 0.9 is a
good starting point. Ideally, you want to visualize the LD before choosing a threshold." />

<meta property="og:description" content="Strategically, run the function with filter.long.ld argument and
refilter the data using the outlier statistic
generated by the function (printed on the figure in the output) and using
long.ld.missing = TRUE. This advanced argument will choose the best SNP
based on missing data statistics, instead of choosing randomly one SNP
(see details).
" />

<meta property="og:description" content="This function is used internally in radiator
and might be of interest for users." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/thierrygosselin/radiator">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>GBS/RADseq short and long distance linkage disequilibrium pruning</h1>
    <small class="dont-index">Source: <a href='https://github.com/thierrygosselin/radiator/blob/master/R/filter_ld.R'><code>R/filter_ld.R</code></a></small>
    <div class="hidden name"><code>filter_ld.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>SNP short and long distance linkage disequilibrium pruning.</p>
<p>What sets appart radiator LD pruning is the RADseq data tailored arguments:</p><ul>
<li><p><strong>minimize short linkage disequilibrium (LD)</strong>:
5 values available for <code>filter.short.ld</code> argument (see below).</p></li>
<li><p><strong>reduce long distance LD</strong>:
Long distance LD pruning is usually advised to avoid capturing the variance LD
in PCA analysis.</p>
<p>Use the argument <code>filter.long.ld</code> with values between 0.7 and 0.9 is a
good starting point. Ideally, you want to visualize the LD before choosing a threshold.</p>

    <p>Strategically, run the function with <code>filter.long.ld</code> argument and
refilter the data using the outlier statistic
generated by the function (printed on the figure in the output) and using
<code>long.ld.missing = TRUE</code>. This advanced argument will choose the best SNP
based on missing data statistics, instead of choosing randomly one SNP
(see details).</p></li>
</ul>

    <p>This function is used internally in <a href='https://github.com/thierrygosselin/radiator'>radiator</a>
and might be of interest for users.</p>
    
    </div>

    <pre class="usage"><span class='fu'>filter_ld</span>(<span class='no'>data</span>, <span class='kw'>interactive.filter</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>filter.short.ld</span> <span class='kw'>=</span> <span class='st'>"mac"</span>,
  <span class='kw'>filter.long.ld</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>parallel.core</span> <span class='kw'>=</span> <span class='kw pkg'>parallel</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/parallel/topics/detectCores'>detectCores</a></span>() - <span class='fl'>1</span>,
  <span class='kw'>filename</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>data</th>
      <td><p>(4 options) A file or object generated by radiator:</p><ul>
<li><p>tidy data</p></li>
<li><p>Genomic Data Structure (GDS)</p></li>
</ul>

<p><em>How to get GDS and tidy data ?</em>
Look into <code><a href='tidy_genomic_data.html'>tidy_genomic_data</a></code>,
<code><a href='read_vcf.html'>read_vcf</a></code> or
<code><a href='tidy_vcf.html'>tidy_vcf</a></code>.</p></td>
    </tr>
    <tr>
      <th>interactive.filter</th>
      <td><p>(optional, logical) Do you want the filtering session to
be interactive. Figures of distribution are shown before asking for filtering
thresholds.
Default: <code>interactive.filter = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>filter.short.ld</th>
      <td><p>(character) 5 options (default: <code>filter.short.ld = "mac"</code>):</p><ol>
<li><p><code>filter.short.ld = "random"</code> for a random selection of 1 SNP on the read,</p></li>
<li><p><code>filter.short.ld = "first"</code> for the first one on the read...,</p></li>
<li><p><code>filter.short.ld = "last"</code> for the last SNP on the read and</p></li>
<li><p><code>filter.short.ld = "middle"</code> for locus with &gt; 2 SNPs/read the option to select at random
one SNP between the first and the last SNP on the read. If the locus as &lt;= 2
SNPs on the read, the first one is selected. Note that for that last option,
the numbers are reported.</p></li>
<li><p><code>filter.short.ld = "mac"</code> will select the SNP on the locus with the maximum global
Minor Allele Count (MAC).</p>
<p>Using <code>filter.short.ld = NULL</code>, skip this filter.</p></li>
</ol></td>
    </tr>
    <tr>
      <th>filter.long.ld</th>
      <td><p>(optional, double) The threshold to prune SNP based on
Long Distance Linkage Disequilibrium. The argument filter.long.ld is
the absolute value of measurement.
Default: <code>filter.long.ld = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>parallel.core</th>
      <td><p>(optional) The number of core used for parallel
execution during import.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></td>
    </tr>
    <tr>
      <th>filename</th>
      <td><p>(optional, character) File name prefix for file written in
the working directory.
Default: <code>filename = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>(optional, logical) When <code>verbose = TRUE</code>
the function is a little more chatty during execution.
Default: <code>verbose = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>(optional) Advance mode that allows to pass further arguments
for fine-tuning the function. Also used for legacy arguments (see details or
special section)</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list in the global environment, with these objects:</p><ol>
<li><p>$ld.summary: tibble with LD statistics used for the boxplot</p></li>
<li><p>$ld.boxplot: box plot of LD values</p></li>
<li><p>$whitelist.ld: whitelist of markers kept after filtering for LD.
The argument <code>filter.long.ld</code> must be used to generate the whitelist.</p></li>
<li><p>$blacklist.ld: blacklist of markers prunned during the filtering
for LD.
The argument <code>filter.long.ld</code> must be used to generate the blacklist.</p></li>
<li><p>$data: The filtered tidy dataset.</p></li>
<li><p>$gds: the path to the GDS file.</p></li>
</ol>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The function requires <a href='https://github.com/zhengxwen/SNPRelate'>SNPRelate</a>
(see example below on how to install).</p>
    <p><strong>Advance mode, using <em>dots-dots-dots</em></strong></p><ul>
<li><p><code>maf.data</code> (path) this argument is no longer supported.
It's a small cost in time in favour of making sure the MAC/MAF fits the actual
data.</p></li>
<li><p><code>long.ld.missing</code> (logical) With <code>long.ld.missing = TRUE</code>.
The function first generates long distance LD values between markers along
the same chromosome or scaffold with <em>SNPRelate::snpgdsLDMat</em>.
Based on the LD threshold (<code>filter.long.ld</code>) SNPs in LD will be pruned
based on missingness.
e.g. if 4 SNPs are in LD, the 1 SNP selected in
the end is base on genotyping rate/missingness. If this statistic is equal
between the SNPs in LD, 1 SNP is chosen randomly.</p>
<p>Using missigness add extra computational time. To speed the analysis when
missingness between markers is not an issue, use <code>long.ld.missing = FALSE</code>.
The function will use <em>SNPRelate::snpgdsLDpruning</em>
to prune the dataset. SNPs in LD are selected randomly.
Default: <code>long.ld.missing = FALSE</code>.</p></li>
<li><p><code>ld.method</code>: (optional, character) The values available are
<code>"composite"</code>, for LD composite measure, <code>"r"</code> for R coefficient
(by EM algorithm assuming HWE, it could be negative), <code>"r2"</code> for r^2,
<code>"dprime"</code> for D',
<code>"corr"</code> for correlation coefficient. The method corr and composite are
equivalent when SNPs are coded based on the presence of the alternate allele
(<code>0, 1, 2</code>).
Default: <code>ld.method = "r2"</code>.</p></li>
<li><p><code>ld.figures</code>: (logical) Generate long distance LD statistics and
figures.
Default: <code>ld.figures = TRUE</code></p></li>
<li><p><code>path.folder</code>: to write ouput in a specific path
(used internally in radiator). Default: <code>path.folder = getwd()</code>.
If the supplied directory doesn't exist, it's created.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Zheng X, Levine D, Shen J, Gogarten SM, Laurie C, Weir BS.
(2012) A high-performance computing toolset for relatedness and principal component
analysis of SNP data. Bioinformatics. 28: 3326-3328.
doi:10.1093/bioinformatics/bts606</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='https://github.com/zhengxwen/SNPRelate'>SNPRelate</a></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/library'>require</a></span>(<span class='no'>SNPRelate</span>)
<span class='co'>#To install SNPRelate:</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/install.packages'>install.packages</a></span>(<span class='st'>"BiocManager"</span>)
<span class='kw pkg'>BiocManager</span><span class='kw ns'>::</span><span class='fu'>install</span> (<span class='st'>"SNPRelate"</span>)
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/library'>library</a></span>(<span class='no'>radiator</span>)
<span class='no'>data</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>radiator</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/radiator/topics/read_vcf'>read_vcf</a></span>(<span class='kw'>data</span> <span class='kw'>=</span> <span class='st'>"my.vcf"</span>, <span class='kw'>strata</span> <span class='kw'>=</span> <span class='st'>"my.strata.tsv"</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># short distance LD, no long distance LD:</span>
<span class='no'>check.short.ld</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>radiator</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/radiator/topics/filter_ld'>filter_ld</a></span>(<span class='kw'>data</span> <span class='kw'>=</span> <span class='no'>data</span>, <span class='kw'>filter.short.ld</span> <span class='kw'>=</span> <span class='st'>"mac"</span>)

<span class='co'># short distance LD and long distance LD:</span>
<span class='no'>pruned.ld</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>radiator</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/radiator/topics/filter_ld'>filter_ld</a></span>(
        <span class='kw'>data</span> <span class='kw'>=</span> <span class='no'>data</span>,
        <span class='kw'>filter.short.ld</span> <span class='kw'>=</span> <span class='st'>"mac"</span>,
        <span class='kw'>filter.long.ld</span> <span class='kw'>=</span> <span class='fl'>0.8</span>)

<span class='co'># short distance LD and long distance LD, incorporating missing data:</span>
<span class='no'>pruned.ld</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>radiator</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/radiator/topics/filter_ld'>filter_ld</a></span>(
         <span class='kw'>data</span> <span class='kw'>=</span> <span class='no'>data</span>, <span class='co'># a GDS object generated by radiator</span>
         <span class='kw'>filter.short.ld</span> <span class='kw'>=</span> <span class='st'>"mac"</span>,
         <span class='kw'>filter.long.ld</span> <span class='kw'>=</span> <span class='fl'>0.8</span>,
         <span class='kw'>long.ld.missing</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)


<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Thierry Gosselin <a href='mailto:thierrygosselin@icloud.com'>thierrygosselin@icloud.com</a></p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Thierry Gosselin, Floriaan Devloo-Delva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

