<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>FH measure of IBDg — ibdg_fh • radiator</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="FH measure of IBDg — ibdg_fh" />

<meta property="og:description" content="FH is a proxy mesure of IBDg based on the excess in the observed
number of homozygous genotypes within an individual,
relative to the mean number of homozygous genotypes expected under random mating
(Keller et al., 2011; Kardos et al., 2015; Hedrick &amp;amp; Garcia-Dorado, 2016).
IBDg is the realized proportion of the individual genome
that is identical by descent by reference to the current population
under hypothetical random mating
(Keller et al., 2011; Kardos et al., 2015; Hedrick &amp;amp; Garcia-Dorado, 2016).
This function is using a modified version of the FH measure
(constructed using PLINK
-het option) described in (Keller et al., 2011; Kardos et al., 2015).
The novelties are:

population-wise: the individual's observed homozygosity is
contrasted against the expected homozygosity.
Two estimates of the expected homozygosity are provided based
on the population and/or the overall expected homozygosity
averaged across markers.
tailored for RADseq: instead of using the overall number
of markers, the population and the overall expected homozygosity
are averaged with the same markers the individual's are genotyped for.
This reduces the bias potentially introduced by comparing the individual's
observed homozygosity (computed from non-missing genotypes) with
an estimate computed with more markers found at the population or at the
overall level.
" />

<meta property="og:description" content="The FH measure is also computed in
stackr summary_haplotypes
function and grur
missing_visualization functions.
See theory below for the equations." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/getting_started.html">Getting started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/thierrygosselin/radiator">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>FH measure of IBDg</h1>
    <small class="dont-index">Source: <a href='https://github.com/thierrygosselin/radiator/blob/master/R/ibdg_fh.R'><code>R/ibdg_fh.R</code></a></small>
    <div class="hidden name"><code>ibdg_fh.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>FH is a proxy mesure of IBDg based on the excess in the observed
number of homozygous genotypes within an individual,
relative to the mean number of homozygous genotypes expected under random mating
(Keller et al., 2011; Kardos et al., 2015; Hedrick &amp; Garcia-Dorado, 2016).</p>
<p><strong>IBDg</strong> is the realized proportion of the individual genome
that is identical by descent by reference to the current population
under hypothetical random mating
(Keller et al., 2011; Kardos et al., 2015; Hedrick &amp; Garcia-Dorado, 2016).</p>
<p>This function is using a modified version of the FH measure
(constructed using <a href='http://pngu.mgh.harvard.edu/~purcell/plink/'>PLINK</a>
<code>-het</code> option) described in (Keller et al., 2011; Kardos et al., 2015).</p>
<p>The novelties are:</p>
<ul>
<li><p><strong>population-wise:</strong> the individual's observed homozygosity is
contrasted against the expected homozygosity.
Two estimates of the expected homozygosity are provided based
on the population and/or the overall expected homozygosity
averaged across markers.</p></li>
<li><p><strong>tailored for RADseq:</strong> instead of using the overall number
of markers, the population and the overall expected homozygosity
are averaged with the same markers the individual's are genotyped for.
This reduces the bias potentially introduced by comparing the individual's
observed homozygosity (computed from non-missing genotypes) with
an estimate computed with more markers found at the population or at the
overall level.</p></li>
</ul>

    <p>The FH measure is also computed in
<a href='https://github.com/thierrygosselin/stackr'>stackr</a> <em>summary_haplotypes</em>
function and <a href='https://github.com/thierrygosselin/grur'>grur</a>
<em>missing_visualization</em> functions.
See theory below for the equations.</p>
    
    </div>

    <pre class="usage"><span class='fu'>ibdg_fh</span>(<span class='no'>data</span>, <span class='kw'>path.folder</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>data</th>
      <td><p>(4 options) A file or object generated by radiator:</p><ul>
<li><p>tidy data</p></li>
<li><p>Genomic Data Structure (GDS)</p></li>
</ul>

<p><em>How to get GDS and tidy data ?</em>
Look into <code><a href='tidy_genomic_data.html'>tidy_genomic_data</a></code>,
<code><a href='read_vcf.html'>read_vcf</a></code> or
<code><a href='tidy_vcf.html'>tidy_vcf</a></code>.</p></td>
    </tr>
    <tr>
      <th>path.folder</th>
      <td><p>(path, optional) By default will print results in the working directory.
Default: <code>path.folder = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>(optional, logical) When <code>verbose = TRUE</code>
the function is a little more chatty during execution.
Default: <code>verbose = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>(optional) To pass further arguments for fine-tuning the function.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list is created with 6 objects (function call, tables, manhattan,
boxplot and distribution plot).
FH measure is on average negative when the parents are less related than
expected by random mating. The distribution <code>fh.distribution.plot</code>
should be centered around 0 in samples of non-inbred individuals.
The first table, <code>$fh</code>, gives the individual's value
while the second table, <code>$fh.stats</code>, show the population and overall averaged.</p>
    
    <h2 class="hasAnchor" id="theory"><a class="anchor" href="#theory"></a>Theory</h2>

    
    <p><strong>Modified FH:</strong>
$$F_{h_i} = \frac{\overline{Het}_{obs_{ij}} - \overline{Het}_{exp_j}}{\sum_{i}snp_{ij} - \overline{Het}_{exp_j}}$$</p>
<p><strong>Individual Observed Heterozygosity averaged across markers:</strong>
$$\overline{Het}_{obs_i} = \frac{\sum_iHet_{obs_i}}{\sum_i{snp_i}}$$</p>
<p><strong>Population expected Heterozygosity (under Hardy-Weinberg) and
tailored by averaging for each individual using his genotyped markers:</strong>
$$\overline{Het}_{exp_j} = \frac{\sum_jHet_{exp_j}}{\sum_j{snp_j}}$$</p>
    
    <h2 class="hasAnchor" id="advance-mode"><a class="anchor" href="#advance-mode"></a>Advance mode</h2>

    
    <p><em>dots-dots-dots ...</em> allows to pass several arguments for fine-tuning the function.
These arguments are described in <code><a href='tidy_genomic_data.html'>tidy_genomic_data</a></code>.</p><ul>
<li><p>filter.monomorphic = TRUE (default)</p></li>
<li><p>filter.common.markers = FALSE. The argument for common markers
between populations is set by default to maximize genome coverage of
individuals and populations.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Keller MC, Visscher PM, Goddard ME (2011)
Quantification of inbreeding due to distant ancestors and its detection
 using dense single nucleotide polymorphism data. Genetics, 189, 237–249.</p>
<p>Kardos M, Luikart G, Allendorf FW (2015)
Measuring individual inbreeding in the age of genomics: marker-based
measures are better than pedigrees. Heredity, 115, 63–72.</p>
<p>Hedrick PW, Garcia-Dorado A. (2016)
Understanding Inbreeding Depression, Purging, and Genetic Rescue.
Trends in Ecology and Evolution. 2016; 31: 940-952.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='co'># Using a  VCF file, the simplest for of the function:</span>
<span class='no'>fh</span> <span class='kw'>&lt;-</span> <span class='fu'>ibdg_fh</span>(<span class='kw'>data</span> <span class='kw'>=</span> <span class='st'>"sturgeon.gds"</span>)

<span class='co'># To see what's inside the list</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/names'>names</a></span>(<span class='no'>fh</span>)

<span class='co'># To view the boxplot:</span>
<span class='no'>fh</span>$<span class='no'>fh.boxplot</span>

<span class='co'># To view the distribution of FH values:</span>
<span class='no'>fh</span>$<span class='no'>fh.distribution.plot</span>
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#theory">Theory</a></li>

      <li><a href="#advance-mode">Advance mode</a></li>

      <li><a href="#references">References</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Thierry Gosselin <a href='mailto:thierrygosselin@icloud.com'>thierrygosselin@icloud.com</a></p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Thierry Gosselin, Floriaan Devloo-Delva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

