<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ radiator</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/radiator" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="basic">Basic <a href="#basic" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="read_strata.html">read_strata()</a></code> </p>
        </td>
        <td><p>read strata</p></td>
      </tr><tr><td>
          <p><code><a href="summary_strata.html">summary_strata()</a></code> </p>
        </td>
        <td><p>Summary of strata</p></td>
      </tr><tr><td>
          <p><code><a href="filter_rad.html">filter_rad()</a></code> </p>
        </td>
        <td><p>ONE FUNCTION TO RULE THEM ALL</p></td>
      </tr><tr><td>
          <p><code><a href="genomic_converter.html">genomic_converter()</a></code> </p>
        </td>
        <td><p>Conversion tool among several genomic formats</p></td>
      </tr><tr><td>
          <p><code><a href="radiator_pkg_install.html">radiator_pkg_install()</a></code> </p>
        </td>
        <td><p>radiator packages install helper</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="working-with-strata">Working with strata <a href="#working-with-strata" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="read_strata.html">read_strata()</a></code> </p>
        </td>
        <td><p>read strata</p></td>
      </tr><tr><td>
          <p><code><a href="summary_strata.html">summary_strata()</a></code> </p>
        </td>
        <td><p>Summary of strata</p></td>
      </tr><tr><td>
          <p><code><a href="join_strata.html">join_strata()</a></code> </p>
        </td>
        <td><p>Join the strata with the data</p></td>
      </tr><tr><td>
          <p><code><a href="individuals2strata.html">individuals2strata()</a></code> </p>
        </td>
        <td><p>Create a strata file from a list of individuals</p></td>
      </tr><tr><td>
          <p><code><a href="change_pop_names.html">change_pop_names()</a></code> </p>
        </td>
        <td><p>Transform into a factor the STRATA column, change names and reorder the levels</p></td>
      </tr><tr><td>
          <p><code><a href="generate_strata.html">generate_strata()</a></code> </p>
        </td>
        <td><p>Generate strata object from the data</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="read">Read <a href="#read" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="read_blacklist_genotypes.html">read_blacklist_genotypes()</a></code> </p>
        </td>
        <td><p>read blacklist of genotypes</p></td>
      </tr><tr><td>
          <p><code><a href="read_blacklist_id.html">read_blacklist_id()</a></code> </p>
        </td>
        <td><p>read_blacklist_id</p></td>
      </tr><tr><td>
          <p><code><a href="read_dart.html">read_dart()</a></code> </p>
        </td>
        <td><p>Read and tidy <a href="http://www.diversityarrays.com" class="external-link">DArT</a> output files.</p></td>
      </tr><tr><td>
          <p><code><a href="read_plink.html">read_plink()</a></code> </p>
        </td>
        <td><p>Reads PLINK tped and bed files</p></td>
      </tr><tr><td>
          <p><code><a href="read_rad.html">read_rad()</a></code> </p>
        </td>
        <td><p>Read radiator file ending <code>.gds, .rad, .gds.rad</code>.</p></td>
      </tr><tr><td>
          <p><code><a href="read_strata.html">read_strata()</a></code> </p>
        </td>
        <td><p>read strata</p></td>
      </tr><tr><td>
          <p><code><a href="read_vcf.html">read_vcf()</a></code> </p>
        </td>
        <td><p>Read VCF files and write a GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="read_whitelist.html">read_whitelist()</a></code> </p>
        </td>
        <td><p>read whitelist of markers</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="tidy">Tidy <a href="#tidy" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="tidy_genomic_data.html">tidy_genomic_data()</a></code> </p>
        </td>
        <td><p>Transform common genomic dataset format in a tidy data frame</p></td>
      </tr><tr><td>
          <p><code><a href="tidy_dart_metadata.html">tidy_dart_metadata()</a></code> </p>
        </td>
        <td><p>Import and tidy <a href="http://www.diversityarrays.com" class="external-link">DArT</a> metadata.</p></td>
      </tr><tr><td>
          <p><code><a href="tidy_fstat.html">tidy_fstat()</a></code> </p>
        </td>
        <td><p>fstat file to tidy dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="tidy_genepop.html">tidy_genepop()</a></code> </p>
        </td>
        <td><p>Import genepop file and convert to a tidy dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="tidy_genind.html">tidy_genind()</a></code> </p>
        </td>
        <td><p>Tidy a genind object to a tidy dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="tidy_genlight.html">tidy_genlight()</a></code> </p>
        </td>
        <td><p>Tidy a genlight object to a tidy dataframe and/or GDS object/file</p></td>
      </tr><tr><td>
          <p><code><a href="tidy_gtypes.html">tidy_gtypes()</a></code> </p>
        </td>
        <td><p>Tidy a gtypes object to a tidy dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="tidy_plink.html">tidy_plink()</a></code> </p>
        </td>
        <td><p>Tidy PLINK tped and bed files</p></td>
      </tr><tr><td>
          <p><code><a href="tidy_vcf.html">tidy_vcf()</a></code> </p>
        </td>
        <td><p>Tidy vcf file</p></td>
      </tr><tr><td>
          <p><code><a href="tidy_wide.html">tidy_wide()</a></code> </p>
        </td>
        <td><p>Read/Import a tidy genomic data frames.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="filter">Filter <a href="#filter" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="filter_blacklist_genotypes.html">filter_blacklist_genotypes()</a></code> </p>
        </td>
        <td><p>Filter dataset with blacklist of genotypes</p></td>
      </tr><tr><td>
          <p><code><a href="filter_common_markers.html">filter_common_markers()</a></code> </p>
        </td>
        <td><p>Filter common markers between strata</p></td>
      </tr><tr><td>
          <p><code><a href="filter_coverage.html">filter_coverage()</a></code> </p>
        </td>
        <td><p>Filter markers mean coverage</p></td>
      </tr><tr><td>
          <p><code><a href="filter_dart_reproducibility.html">filter_dart_reproducibility()</a></code> </p>
        </td>
        <td><p>Filter data based on DArT reproducibility statistics</p></td>
      </tr><tr><td>
          <p><code><a href="filter_fis.html">filter_fis()</a></code> </p>
        </td>
        <td><p>Fis filter</p></td>
      </tr><tr><td>
          <p><code><a href="filter_genotyping.html">filter_genotyping()</a></code> </p>
        </td>
        <td><p>Filter markers based on genotyping/missing rate</p></td>
      </tr><tr><td>
          <p><code><a href="filter_het.html">filter_het()</a></code> </p>
        </td>
        <td><p>Heterozygosity filter</p></td>
      </tr><tr><td>
          <p><code><a href="filter_hwe.html">filter_hwe()</a></code> </p>
        </td>
        <td><p>Filter markers based on Hardy-Weinberg Equilibrium</p></td>
      </tr><tr><td>
          <p><code><a href="filter_individuals.html">filter_individuals()</a></code> </p>
        </td>
        <td><p>Filter individuals based on genotyping/missingness rate,
heterozygosity and total coverage</p></td>
      </tr><tr><td>
          <p><code><a href="filter_ld.html">filter_ld()</a></code> </p>
        </td>
        <td><p>GBS/RADseq short and long distance linkage disequilibrium pruning</p></td>
      </tr><tr><td>
          <p><code><a href="filter_ma.html">filter_ma()</a></code> </p>
        </td>
        <td><p>MAC, MAF and MAD filter</p></td>
      </tr><tr><td>
          <p><code><a href="filter_monomorphic.html">filter_monomorphic()</a></code> </p>
        </td>
        <td><p>Filter monomorphic markers</p></td>
      </tr><tr><td>
          <p><code><a href="filter_rad.html">filter_rad()</a></code> </p>
        </td>
        <td><p>ONE FUNCTION TO RULE THEM ALL</p></td>
      </tr><tr><td>
          <p><code><a href="filter_snp_number.html">filter_snp_number()</a></code> </p>
        </td>
        <td><p>Filter SNP number per locus/read</p></td>
      </tr><tr><td>
          <p><code><a href="filter_snp_position_read.html">filter_snp_position_read()</a></code> </p>
        </td>
        <td><p>Filter markers/SNP based on their position on the read</p></td>
      </tr><tr><td>
          <p><code><a href="filter_whitelist.html">filter_whitelist()</a></code> </p>
        </td>
        <td><p>Filter dataset with whitelist of markers</p></td>
      </tr><tr><td>
          <p><code><a href="reset_filters.html">reset_filters()</a></code> </p>
        </td>
        <td><p>Reset filters (individuals and markers) in radiator GDS object.</p></td>
      </tr><tr><td>
          <p><code><a href="list_filters.html">list_filters()</a></code> </p>
        </td>
        <td><p>List current active filters (individuals and markers) in radiator GDS object.</p></td>
      </tr><tr><td>
          <p><code><a href="sync_gds.html">sync_gds()</a></code> </p>
        </td>
        <td><p>sync_gds</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="detect">Detect <a href="#detect" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="detect_all_missing.html">detect_all_missing()</a></code> </p>
        </td>
        <td><p>Detect markers with all missing genotypes</p></td>
      </tr><tr><td>
          <p><code><a href="detect_allele_problems.html">detect_allele_problems()</a></code> </p>
        </td>
        <td><p>Detect alternate allele problems</p></td>
      </tr><tr><td>
          <p><code><a href="detect_biallelic_markers.html">detect_biallelic_markers()</a></code> </p>
        </td>
        <td><p>Detect biallelic data</p></td>
      </tr><tr><td>
          <p><code><a href="detect_biallelic_problems.html">detect_biallelic_problems()</a></code> </p>
        </td>
        <td><p>Detect biallelic problems</p></td>
      </tr><tr><td>
          <p><code><a href="detect_duplicate_genomes.html">detect_duplicate_genomes()</a></code> </p>
        </td>
        <td><p>Compute pairwise genome similarity or distance between individuals
to highligh potential duplicate individuals</p></td>
      </tr><tr><td>
          <p><code><a href="detect_genomic_format.html">detect_genomic_format()</a></code> </p>
        </td>
        <td><p>Used internally in radiator to detect the file format</p></td>
      </tr><tr><td>
          <p><code><a href="detect_het_outliers.html">detect_het_outliers()</a></code> </p>
        </td>
        <td><p>Detect heterozygotes outliers and estimate miscall rate</p></td>
      </tr><tr><td>
          <p><code><a href="detect_microsatellites.html">detect_microsatellites()</a></code> </p>
        </td>
        <td><p>Detect microsatellites</p></td>
      </tr><tr><td>
          <p><code><a href="detect_mixed_genomes.html">detect_mixed_genomes()</a></code> </p>
        </td>
        <td><p>Detect mixed genomes</p></td>
      </tr><tr><td>
          <p><code><a href="detect_paralogs.html">detect_paralogs()</a></code> </p>
        </td>
        <td><p>Detect paralogs</p></td>
      </tr><tr><td>
          <p><code><a href="detect_ref_genome.html">detect_ref_genome()</a></code> </p>
        </td>
        <td><p>Detect reference genome</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="sex-markers">Sex markers <a href="#sex-markers" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="sexy_markers.html">sexy_markers()</a></code> </p>
        </td>
        <td><p>sexy_markers finds sex-linked markers and re-assigns sex</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="extract">Extract <a href="#extract" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="extract_coverage.html">check_coverage()</a></code> <code><a href="extract_coverage.html">extract_coverage()</a></code> </p>
        </td>
        <td><p>check_coverage</p></td>
      </tr><tr><td>
          <p><code><a href="extract_dart_target_id.html">extract_dart_target_id()</a></code> </p>
        </td>
        <td><p>Extract <a href="http://www.diversityarrays.com" class="external-link">DArT</a> target id</p></td>
      </tr><tr><td>
          <p><code><a href="extract_genotypes_metadata.html">extract_genotypes_metadata()</a></code> </p>
        </td>
        <td><p>extract_genotypes_metadata</p></td>
      </tr><tr><td>
          <p><code><a href="extract_individuals_metadata.html">extract_individuals_metadata()</a></code> </p>
        </td>
        <td><p>extract_individuals_metadata</p></td>
      </tr><tr><td>
          <p><code><a href="extract_individuals_vcf.html">extract_individuals_vcf()</a></code> </p>
        </td>
        <td><p>Extract individuals from vcf file</p></td>
      </tr><tr><td>
          <p><code><a href="extract_markers_metadata.html">extract_markers_metadata()</a></code> </p>
        </td>
        <td><p>extract_markers_metadata</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="summarize">Summarize <a href="#summarize" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="summary_gds.html">summary_gds()</a></code> </p>
        </td>
        <td><p>summary_gds</p></td>
      </tr><tr><td>
          <p><code><a href="summary_rad.html">summary_rad()</a></code> </p>
        </td>
        <td><p>Summary statistics for RADseq data</p></td>
      </tr><tr><td>
          <p><code><a href="summary_strata.html">summary_strata()</a></code> </p>
        </td>
        <td><p>Summary of strata</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="write">Write <a href="#write" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="write_arlequin.html">write_arlequin()</a></code> </p>
        </td>
        <td><p>Write an arlequin file from a tidy data frame</p></td>
      </tr><tr><td>
          <p><code><a href="write_bayescan.html">write_bayescan()</a></code> </p>
        </td>
        <td><p>Write a <a href="http://cmpg.unibe.ch/software/BayeScan/" class="external-link">BayeScan</a>
file from a tidy data frame</p></td>
      </tr><tr><td>
          <p><code><a href="write_betadiv.html">write_betadiv()</a></code> </p>
        </td>
        <td><p>Write a betadiv file from a tidy data frame</p></td>
      </tr><tr><td>
          <p><code><a href="write_colony.html">write_colony()</a></code> </p>
        </td>
        <td><p>Write a <code>COLONY</code> input file</p></td>
      </tr><tr><td>
          <p><code><a href="write_dadi.html">write_dadi()</a></code> </p>
        </td>
        <td><p>Write a <code>dadi</code> SNP input file from a tidy data frame.</p></td>
      </tr><tr><td>
          <p><code><a href="write_faststructure.html">write_faststructure()</a></code> </p>
        </td>
        <td><p>Write a faststructure file from a tidy data frame</p></td>
      </tr><tr><td>
          <p><code><a href="write_fineradstructure.html">write_fineradstructure()</a></code> </p>
        </td>
        <td><p>Write a fineRADstructure file from a tidy data frame</p></td>
      </tr><tr><td>
          <p><code><a href="write_gds.html">write_gds()</a></code> </p>
        </td>
        <td><p>Write a GDS object from a tidy data frame</p></td>
      </tr><tr><td>
          <p><code><a href="write_genepop.html">write_genepop()</a></code> </p>
        </td>
        <td><p>Write a genepop file</p></td>
      </tr><tr><td>
          <p><code><a href="write_genepopedit.html">write_genepopedit()</a></code> </p>
        </td>
        <td><p>Write a genepopedit flatten object</p></td>
      </tr><tr><td>
          <p><code><a href="write_genind.html">write_genind()</a></code> </p>
        </td>
        <td><p>Write a genind object from a tidy data frame or GDS file or object.</p></td>
      </tr><tr><td>
          <p><code><a href="write_genlight.html">write_genlight()</a></code> </p>
        </td>
        <td><p>Write a genlight object from a tidy data frame or GDS file or object.</p></td>
      </tr><tr><td>
          <p><code><a href="write_gsi_sim.html">write_gsi_sim()</a></code> </p>
        </td>
        <td><p>Write a gsi_sim file from a data frame (wide or long/tidy).</p></td>
      </tr><tr><td>
          <p><code><a href="write_gtypes.html">write_gtypes()</a></code> </p>
        </td>
        <td><p>Write a <a href="https://github.com/EricArcher/" class="external-link">strataG</a> object from
radiator GDS and tidy data file or object.</p></td>
      </tr><tr><td>
          <p><code><a href="write_hapmap.html">write_hapmap()</a></code> </p>
        </td>
        <td><p>Write a HapMap file from a tidy data frame</p></td>
      </tr><tr><td>
          <p><code><a href="write_hierfstat.html">write_hierfstat()</a></code> </p>
        </td>
        <td><p>Write a hierfstat file from a tidy data frame</p></td>
      </tr><tr><td>
          <p><code><a href="write_hzar.html">write_hzar()</a></code> </p>
        </td>
        <td><p>Write a HZAR file from a tidy data frame.</p></td>
      </tr><tr><td>
          <p><code><a href="write_ldna.html">write_ldna()</a></code> </p>
        </td>
        <td><p>Write a LDna object from a tidy data frame</p></td>
      </tr><tr><td>
          <p><code><a href="write_maverick.html">write_maverick()</a></code> </p>
        </td>
        <td><p>Write a maverick file from a tidy data frame</p></td>
      </tr><tr><td>
          <p><code><a href="write_pcadapt.html">write_pcadapt()</a></code> </p>
        </td>
        <td><p>Write a <a href="https://github.com/bcm-uga/pcadapt" class="external-link">pcadapt</a>
file from a tidy data frame</p></td>
      </tr><tr><td>
          <p><code><a href="write_plink.html">write_plink()</a></code> </p>
        </td>
        <td><p>Write a plink tped/tfam file from a tidy data frame</p></td>
      </tr><tr><td>
          <p><code><a href="write_rad.html">write_rad()</a></code> </p>
        </td>
        <td><p>Write tidy genomic data file or close GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="write_related.html">write_related()</a></code> </p>
        </td>
        <td><p>Write a related file from a tidy data frame</p></td>
      </tr><tr><td>
          <p><code><a href="write_rubias.html">write_rubias()</a></code> </p>
        </td>
        <td><p>Write a rubias object</p></td>
      </tr><tr><td>
          <p><code><a href="write_snprelate.html">write_snprelate()</a></code> </p>
        </td>
        <td><p>Write a SNPRelate object from a tidy data frame</p></td>
      </tr><tr><td>
          <p><code><a href="write_stockr.html">write_stockr()</a></code> </p>
        </td>
        <td><p>Write a stockR dataset from a tidy data frame or GDS file or object.</p></td>
      </tr><tr><td>
          <p><code><a href="write_structure.html">write_structure()</a></code> </p>
        </td>
        <td><p>Write a structure file from a tidy data frame</p></td>
      </tr><tr><td>
          <p><code><a href="write_vcf.html">write_vcf()</a></code> </p>
        </td>
        <td><p>Write a vcf file from a tidy data frame</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="utility-functions">Utility functions <a href="#utility-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Internal modules</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="allele_frequencies.html">allele_frequencies()</a></code> </p>
        </td>
        <td><p>Compute allele frequencies per markers and populations</p></td>
      </tr><tr><td>
          <p><code><a href="betas_estimator.html">betas_estimator()</a></code> </p>
        </td>
        <td><p>Estimate \(\beta\)s per population</p></td>
      </tr><tr><td>
          <p><code><a href="calibrate_alleles.html">calibrate_alleles()</a></code> </p>
        </td>
        <td><p>Calibrate REF and ALT alleles based on count</p></td>
      </tr><tr><td>
          <p><code><a href="change_pop_names.html">change_pop_names()</a></code> </p>
        </td>
        <td><p>Transform into a factor the STRATA column, change names and reorder the levels</p></td>
      </tr><tr><td>
          <p><code><a href="check_pop_levels.html">check_pop_levels()</a></code> </p>
        </td>
        <td><p>Check the use of pop.levels, pop.labels and pop.select arguments.</p></td>
      </tr><tr><td>
          <p><code><a href="clean_ind_names.html">clean_ind_names()</a></code> </p>
        </td>
        <td><p>Clean individual's names for radiator and other packages</p></td>
      </tr><tr><td>
          <p><code><a href="clean_markers_names.html">clean_markers_names()</a></code> </p>
        </td>
        <td><p>Clean marker's names for radiator and other packages</p></td>
      </tr><tr><td>
          <p><code><a href="clean_pop_names.html">clean_pop_names()</a></code> </p>
        </td>
        <td><p>Clean population's names for radiator and other packages</p></td>
      </tr><tr><td>
          <p><code><a href="ibdg_fh.html">ibdg_fh()</a></code> </p>
        </td>
        <td><p>FH measure of IBDg</p></td>
      </tr><tr><td>
          <p><code><a href="pi.html">pi()</a></code> </p>
        </td>
        <td><p>Nucleotide diversity</p></td>
      </tr><tr><td>
          <p><code><a href="private_alleles.html">private_alleles()</a></code> </p>
        </td>
        <td><p>Find private alleles</p></td>
      </tr><tr><td>
          <p><code><a href="private_haplotypes.html">private_haplotypes()</a></code> </p>
        </td>
        <td><p>private haplotypes</p></td>
      </tr><tr><td>
          <p><code><a href="separate_markers.html">separate_markers()</a></code> </p>
        </td>
        <td><p>Separate markers column into chrom, locus and pos</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="run-software">Run software <a href="#run-software" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="run_bayescan.html">run_bayescan()</a></code> </p>
        </td>
        <td><p>Run BayeScan</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

