<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Visual diagnostic of coverage imbalance — plot_coverage_imbalance_diagnostic • radiator</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Visual diagnostic of coverage imbalance — plot_coverage_imbalance_diagnostic"><meta property="og:description" content="GBS data and STACKS pipeline sometimes output REF and ALT
alleles that have coverage imbalance, i.e. the coverage is not equal
and skewed towards the REF or ALT alleles.
Thw density distribution figure of coverage imbalance between REF and ALT
alleles will highlight the problem in you data."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/radiator" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Visual diagnostic of coverage imbalance</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/radiator/blob/HEAD/R/visualization.R" class="external-link"><code>R/visualization.R</code></a></small>
    <div class="hidden name"><code>plot_coverage_imbalance_diagnostic.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>GBS data and STACKS pipeline sometimes output REF and ALT
alleles that have coverage imbalance, i.e. the coverage is not equal
and skewed towards the REF or ALT alleles.
Thw density distribution figure of coverage imbalance between REF and ALT
alleles will highlight the problem in you data.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_coverage_imbalance_diagnostic</span><span class="op">(</span></span>
<span>  <span class="va">tidy.vcf.file</span>,</span>
<span>  <span class="va">pop.levels</span>,</span>
<span>  <span class="va">read.depth.threshold</span>,</span>
<span>  <span class="va">aes.colour</span>,</span>
<span>  <span class="va">adjust.bin</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-tidy-vcf-file">tidy.vcf.file<a class="anchor" aria-label="anchor" href="#arg-tidy-vcf-file"></a></dt>
<dd><p>The tidy VCF file created with tidy_genomic_data.</p></dd>


<dt id="arg-pop-levels">pop.levels<a class="anchor" aria-label="anchor" href="#arg-pop-levels"></a></dt>
<dd><p>Character string defining your ordered populations.</p></dd>


<dt id="arg-read-depth-threshold">read.depth.threshold<a class="anchor" aria-label="anchor" href="#arg-read-depth-threshold"></a></dt>
<dd><p>Define the threshold you wish to analyse.</p></dd>


<dt id="arg-aes-colour">aes.colour<a class="anchor" aria-label="anchor" href="#arg-aes-colour"></a></dt>
<dd><p>GGPLOT2 aesthetics,
e.g. aes(y = ..count..).</p></dd>


<dt id="arg-adjust-bin">adjust.bin<a class="anchor" aria-label="anchor" href="#arg-adjust-bin"></a></dt>
<dd><p>Adjust GGPLOT2 bin size (0 to 1).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>4-plots highlighting the different combination of under
 or over the coverage threshold and mean genotype likelihood.
 Y- axis show the distribution of genotypes.
 X- axis show the coverage imbalance the
 Negative ratio (left x axis) : REF &gt; ALT.
 Positive ratio (right x axis) : ALT &gt; REF.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The figures shows the results of the of coverage threshold
selected and mean genotype likelihood.
You can test different threshold to inspect your data.
Ideally the lower left pannel of the 4-plot should be empty. If it is, this
shows that setting the threshold of the genotype likelihood filter
to the mean or close to it take care of the allelic coverage imbalance.
#' e.g. fig &lt;- plot_coverage_imbalance_diagnostic(
tidy.vcf.file, pop.levels, read.depth.threshold, aes.colour, adjust.bin)
Use ( fig + facet_grid(GROUP_GL ~ GROUP_COVERAGE)). The ratio is calculated :
(read depth ALT allele - read depth REF allele)/(read depth ALT allele + read depth REF allele).</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

