<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Read and tidy DArT output files. — read_dart • radiator</title><!-- favicons --><link rel="icon" type="image/png" sizes="48x48" href="../favicon-48x48.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Read and tidy DArT output files. — read_dart"><meta property="og:description" content="Used internally in radiator
and might be of interest for users. The function generate a GDS object/file
and optionally, a tidy dataset using
DArT files. Read the details section
to understand why it's better than dartR."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/radiator" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read and tidy <a href="http://www.diversityarrays.com" class="external-link">DArT</a> output files.</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/radiator/blob/HEAD/R/dart.R" class="external-link"><code>R/dart.R</code></a></small>
    <div class="hidden name"><code>read_dart.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Used internally in <a href="https://github.com/thierrygosselin/radiator" class="external-link">radiator</a>
and might be of interest for users. The function generate a GDS object/file
and optionally, a tidy dataset using
<a href="http://www.diversityarrays.com" class="external-link">DArT</a> files. Read the <em>details</em> section
to understand why it's better than dartR.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">read_dart</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  <span class="va">strata</span>,</span>
<span>  filename <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tidy.dart <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  calibrate.alleles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>(file) 6 files formats used by DArT are recognized by radiator.
Don't modify the DArT file, to do this, use the <code>strata</code> file/argument below.
The function can import files ending with <code>.csv</code> or <code>.tsv</code>.</p>
<ol><li><p><strong>1row</strong>: Genotypes are in 1 row and coded (0, 1, 2, -).
<code>0 for 2 reference alleles REF/REF</code>, <code>1 for 2 alternate alleles ALT/ALT</code>,
<code>2 for heterozygote REF/ALT</code>, <code>- for missing</code>.</p></li>
<li><p><strong>2rows</strong>: No genotypes. It's absence/presence, 0/1, of the REF and ALT alleles.
Sometimes called binary format.</p></li>
<li><p><strong>counts</strong>: No genotypes, It's counts/read depth for the REF and ALT alleles.
Sometimes just called count data. This should be the preferred file format,
because DArT output the coverage (read depth for each genotypes).</p></li>
<li><p><strong>silico.dart</strong>: SilicoDArT data. No genotypes, no REF or ALT alleles.
It's a file coded as absence/presence, 0/1, for the presence of sequence in
the clone id.</p></li>
<li><p><strong>silico.dart.counts</strong>: SilicoDArT data. No genotypes, no REF or ALT alleles.
It's a file coded as absence/presence, with counts for the presence of sequence in
the clone id.</p></li>
<li><p><strong>dart.vcf</strong>: For DArT VCFs, please use <code><a href="read_vcf.html">read_vcf</a></code>.</p></li>
</ol><p>If you encounter a problem, sent me your data so that I can update
the function.</p></dd>


<dt id="arg-strata">strata<a class="anchor" aria-label="anchor" href="#arg-strata"></a></dt>
<dd><p>A tab delimited file or object with a minimum of 3 columns headers:</p><ol><li><p><strong>TARGET_ID</strong>: generated by DArT, it's made of integers.
For <code>1row</code> and <code>2rows</code> the <code>TARGET_ID</code> is usually the sample
name submitted to DArT.</p>

<p>If this is new to you, take a look at this function:
<code><a href="extract_dart_target_id.html">extract_dart_target_id</a></code>, it's easier than opening your DArT file
in <code>MS EXCEL</code>. This will extract the TARGET_ID of your DArT file
you can then use it inside <code>MS EXCEL</code> and build the remaining columns.</p></li>
<li><p><strong>INDIVIDUALS</strong>: This is the time and place no name your sample correctly.</p></li>
<li><p><strong>STRATA</strong>: refers to any grouping of individuals. Usually,
your sampling sites or populations. Keep it simple, 3 or 4 letters, like
TAS for Tasmania, etc.</p>

<p>Silico DArT data is currently used to detect sex markers, so the <code>STRATA</code>
column should be filed with sex information: <code>M</code> or <code>F</code>.</p></li>
</ol></dd>


<dt id="arg-filename">filename<a class="anchor" aria-label="anchor" href="#arg-filename"></a></dt>
<dd><p>(optional) The function uses <code><a href="http://www.fstpackage.org/reference/write_fst.html" class="external-link">write.fst</a></code>,
to write the tidy data frame in
the working directory. The file extension appended to
the <code>filename</code> provided is <code>.rad</code>.
With default: <code>filename = NULL</code>, the tidy data frame is
in the global environment only (i.e. not written in the working directory...).</p></dd>


<dt id="arg-tidy-dart">tidy.dart<a class="anchor" aria-label="anchor" href="#arg-tidy-dart"></a></dt>
<dd><p>(logical, optional) Generate a tidy dataset.
Default:<code>tidy.dart = FALSE</code>.</p></dd>


<dt id="arg-calibrate-alleles">calibrate.alleles<a class="anchor" aria-label="anchor" href="#arg-calibrate-alleles"></a></dt>
<dd><p>(optional, logical)
Default: <code>calibrate.alleles = TRUE</code>.
Documented in <code><a href="calibrate_alleles.html">calibrate_alleles</a></code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>(optional, logical) When <code>verbose = TRUE</code>
the function is a little more chatty during execution.
Default: <code>verbose = TRUE</code>.</p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>(optional) The number of core used for parallel
execution during import.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>(optional) To pass further argument for fine-tuning the function.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A radiator GDS file and tidy dataframe with several columns depending on DArT file:
<code>silico.dart:</code> A tibble with 5 columns: <code>CLONE_ID, SEQUENCE, VALUE, INDIVIDUALS, STRATA</code>.
This object is also saved in the directory (file ending with .rad).</p>
<p>Common to <code>1row, 2rows and counts</code>: A GDS file is automatically generated.
To have a tidy tibble, the argument <code>tidy.dart = TRUE</code> must be used.</p>
<ol><li><p>VARIANT_ID: generated by radiator and correspond the markers in integer.</p></li>
<li><p>MARKERS: generated by radiator and correspond to CHROM + LOCUS + POS separated by 2 underscores.</p></li>
<li><p>CHROM: the chromosome info, for de novo: CHROM_1.</p></li>
<li><p>LOCUS: the locus info.</p></li>
<li><p>POS: the SNP id on the LOCUS.</p></li>
<li><p>COL: the position of the SNP on the short read.</p></li>
<li><p>REF: the reference allele.</p></li>
<li><p>ALT: the alternate allele.</p></li>
<li><p>INDIVIDUALS: the sample name.</p></li>
<li><p>STRATA/POP_ID: populations id of the sample.</p></li>
<li><p>GT_BIN: the genotype based on the number of alternate allele in the genotype
(the count/dosage of the alternate allele). <code>0, 1, 2, NA</code>.</p></li>
<li><p>REP_AVG: the reproducibility average, output specific of DArT.</p></li>
</ol><p>Other columns potentially in the tidy tibble:</p><ol><li><p>GT: the genotype in 6 digit format <em>à la genepop</em>.</p></li>
<li><p>GT_VCF: the genotype in VCF format <code>0/0, 0/1, 1/1, ./.</code>.</p></li>
<li><p>GT_VCF_NUC: the genotype in VCF format, but keeping the nucleotide information.
<code>A/A, A/T, T/T, ./.</code></p></li>
<li><p>AVG_COUNT_REF: the coverage for the reference allele, output specific of DArT.</p></li>
<li><p>AVG_COUNT_SNP: the coverage for the alternate allele, output specific of DArT.</p></li>
<li><p>READ_DEPTH: the number of reads used for the genotype (count data).</p></li>
<li><p>ALLELE_REF_DEPTH: the number of reads of the reference allele (count data).</p></li>
<li><p>ALLELE_ALT_DEPTH: the number of reads of the alternate allele (count data).</p></li>
</ol><p>Written in the working directory:</p><ul><li><p>The radiator GDS file</p></li>
<li><p>The DArT metadata information</p></li>
<li><p>The tidy DArT data</p></li>
<li><p>The strata file associated with this tidy dataset</p></li>
<li><p>The allele dictionary is a tibble with columns:
<code>MARKERS, CHROM, LOCUS, POS, REF, ALT</code>.</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>More details on what is happening under the hood when you import the DArT file
in R:</p><ol><li><p>The DArT file is imported</p><ul><li><p>DArT files should not be modify.</p></li>
<li><p>A lot of imports problems originates from files modifications,
a couple of common checks are done.</p></li>
<li><p>The format (1row, 2rows, counts, silico) is interpreted.</p></li>
<li><p>The number of target ids is checked.</p></li>
</ul></li>
<li><p>The strata file file is imported:</p><ul><li><p>This is the file that needs modifications.</p></li>
<li><p>This is here that you change the bad samples names.</p></li>
<li><p>Remove target ids (blacklist samples that you no longer want).</p></li>
<li><p>Change the order of the populations/sampling sites.</p></li>
<li><p>or alternatively, you can use the <code>pop.levels</code> argument
(see <em>dots-dots-dots ...</em> in the advance mode section below)</p></li>
</ul></li>
<li><p>The target ids between the DArT and strata files are verified and the files
are merged.</p></li>
<li><p>The data is inspected for duplicated names</p></li>
<li><p>DArT changed colnames in their files along the years, we tidy things:</p><ul><li><p>colnames in camelcase are changed to snakecase</p></li>
<li><p>ALLELE_SEQUENCE is changed to SEQUENCE</p></li>
<li><p>TRIMMED_SEQUENCE is changed to SEQUENCE</p></li>
<li><p>CLUSTER_CONSENSUS_SEQUENCE is changed to SEQUENCE</p></li>
<li><p>Genomic metadata are named and or re-named based on the Variant Call Format Specification:CHROM, LOCUS, POS, COL, REF, ALT</p></li>
</ul></li>
<li><p>With this function you have the option to tidy the DArT file:</p><ul><li><p>What's that ?
<a href="https://r4ds.hadley.nz/data-tidy.html" class="external-link">R for Data Science: explanation</a></p></li>
<li><p>It takes longer and you need more memory, but if you can allow it,
it's better for inspection and visualisation.</p></li>
<li><p>or you wait to filter the data and generate a tidy dataset with
<code><a href="tidy_genomic_data.html">tidy_genomic_data</a></code></p></li>
</ul></li>
<li><p>REF and ALT alleles are re-calibrated with <code><a href="calibrate_alleles.html">calibrate_alleles</a></code>:</p><ul><li><p>This is not optional</p></li>
<li><p>It takes longer than just reading the file like other software and packages, but
it's better.</p></li>
</ul></li>
</ol></div>
    <div id="advance-mode">
    <h2>Advance mode</h2>



<p><em>dots-dots-dots ...</em> allows to pass several arguments for fine-tuning the function:</p><ol><li><p><code>whitelist.markers</code>: detailed in <code><a href="filter_whitelist.html">filter_whitelist</a></code>.
Defautl: <code>whitelist.markers = NULL</code>.</p></li>
<li><p><code>missing.memory</code> (option, path)
This argument allows to erase genotypes that have bad statistics.
It's the path to a file <code>.rad</code> file that contains 3 columns:
<code>MARKERS, INDIVIDUALS, ERASE</code>. The file is produced by several radiator
functions. For DArT data, <code><a href="filter_rad.html">filter_rad</a></code> generate the file.
Defautl: <code>missing.memory = NULL</code>. Currently not used.</p></li>
<li><p><code>path.folder</code>: (optional, path) To write output in a specific folder.
Default: <code>path.folder = NULL</code>. The working directory is used.</p></li>
<li><p><code>pop.levels</code>: detailed in <code><a href="tidy_genomic_data.html">tidy_genomic_data</a></code>.</p></li>
</ol></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="extract_dart_target_id.html">extract_dart_target_id</a></code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Thierry Gosselin <a href="mailto:thierrygosselin@icloud.com">thierrygosselin@icloud.com</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">clownfish.dart.tidy</span> <span class="op">&lt;-</span> <span class="fu">radiator</span><span class="fu">::</span><span class="fu">read_dart</span><span class="op">(</span></span></span>
<span class="r-in"><span>    data <span class="op">=</span> <span class="st">"clownfish.dart.csv"</span>,</span></span>
<span class="r-in"><span>    strata <span class="op">=</span> <span class="st">"clownfish.strata.tsv"</span></span></span>
<span class="r-in"><span>    <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

