<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Tidy vcf file — tidy_vcf • radiator</title><!-- favicons --><link rel="icon" type="image/png" sizes="48x48" href="../favicon-48x48.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Tidy vcf file — tidy_vcf"><meta property="og:description" content="The function allows to tidy a VCF file.
Used internally in
radiator
and might be of interest for users.
It is highly recommended to use filter_rad to reduce
the number of markers. Advance options below are also available to
to manipulate and prune the dataset with blacklists and whitelists along
several other filtering options."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/radiator" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Tidy vcf file</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/radiator/blob/HEAD/R/vcf.R" class="external-link"><code>R/vcf.R</code></a></small>
    <div class="hidden name"><code>tidy_vcf.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The function allows to tidy a VCF file.</p>
<p>Used internally in
<a href="https://github.com/thierrygosselin/radiator" class="external-link">radiator</a>
and might be of interest for users.</p>
<p>It is highly recommended to use <code><a href="filter_rad.html">filter_rad</a></code> to reduce
the number of markers. Advance options below are also available to
to manipulate and prune the dataset with blacklists and whitelists along
several other filtering options.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">tidy_vcf</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  strata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filename <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>(VCF file, character string) The VCF SNPs are biallelic or haplotypes.
To make the VCF population-ready, the argument <code>strata</code> is required.</p></dd>


<dt id="arg-strata">strata<a class="anchor" aria-label="anchor" href="#arg-strata"></a></dt>
<dd><p>(optional)
The strata file is a tab delimited file with a minimum of 2 columns headers:
<code>INDIVIDUALS</code> and <code>STRATA</code>. Documented in <code><a href="read_strata.html">read_strata</a></code>.
DArT data: a third column <code>TARGET_ID</code> is required.
Documented on <code><a href="read_dart.html">read_dart</a></code>. Also use the strata read function to
blacklist individuals.
Default: <code>strata = NULL</code>.</p></dd>


<dt id="arg-filename">filename<a class="anchor" aria-label="anchor" href="#arg-filename"></a></dt>
<dd><p>(optional) The function uses <code><a href="http://www.fstpackage.org/reference/write_fst.html" class="external-link">write.fst</a></code>,
to write the tidy data frame in
the working directory. The file extension appended to
the <code>filename</code> provided is <code>.rad</code>.
With default: <code>filename = NULL</code>, the tidy data frame is
in the global environment only (i.e. not written in the working directory...).</p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>(optional) The number of core used for parallel
execution during import.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>(optional, logical) When <code>verbose = TRUE</code>
the function is a little more chatty during execution.
Default: <code>verbose = TRUE</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>(optional) To pass further argument for fine-tuning the tidying
(read below).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>The output in your global environment is a tidy data frame, the GDS file
generated is in the working directory under the name given during function execution.</p>
    </div>
    <div id="vcf-file-format">
    <h2>VCF file format</h2>



<p><strong>PLINK:</strong> radiator fills the <code>LOCUS</code> column of PLINK VCFs with
a unique integer based on the <code>CHROM</code> column
(<code>as.integer(factor(x = CHROM))</code>).
The <code>COL</code> column is filled with 1L for lack of bettern info on this.
Not what you need ? Open an issue on GitHub for a request.</p>
<p><strong>ipyrad:</strong> the pattern <code>locus_</code> in the <code>CHROM</code> column
is removed and used. The <code>COL</code> column is filled with the same value as
<code>POS</code>.</p>
<p><strong>GATK:</strong> Some VCF have an <code>ID</code> column filled with <code>.</code>,
the LOCUS information is all contained along the linkage group in the
<code>CHROM</code> column. To make it work with
<a href="https://github.com/thierrygosselin/radiator" class="external-link">radiator</a>,
the <code>ID</code> column is filled with the <code>POS</code> column info.
GATK with a mix of multi- and bi-allelic dataset won't generate VCF stats.</p>
<p><strong>platypus:</strong> Some VCF files don't have an ID filed with values,
here the same thing is done as GATK VCF files above.</p>
<p><strong>freebayes:</strong> Some VCF files don't have an ID filed with values,
here the same thing is done as GATK VCF files above.</p>
<p><strong>stacks:</strong> with <em>de novo</em> approaches, the CHROM column is
filled with "1", the LOCUS column correspond to the CHROM section in stacks VCF and
the COL column is POS -1. With a reference genome, the ID column in stacks VCF is
separated into "LOCUS", "COL", "STRANDS".</p>
<p><em>stacks problem: </em> current version as some intrinsic problem with
missing allele depth info, during the tidying process a message will
highlight the number of genotypes impacted by the problem. When possible, the
problem is corrected by adding the read depth info into the allele depth field.</p>
    </div>
    <div id="advance-mode-using-dots-dots-dots-">
    <h2>Advance mode, using <em>dots-dots-dots ...</em></h2>



<p>The arguments below are not available using code completion (e.g. with TAB),
consequently any misspelling will generate an error or be ignored.</p>
<p><em>dots-dots-dots ...</em> arguments names and values are reported and written
in the working directory.</p>
<p><strong>General arguments: </strong></p><ol><li><p><code>path.folder</code>: to write ouput in a specific path
(used internally in radiator). Default: <code>path.folder = getwd()</code>.
If the supplied directory doesn't exist, it's created.</p></li>
<li><p><code>random.seed</code>: (integer, optional) For reproducibility, set an integer
that will be used inside codes that uses randomness. With default,
a random number is generated, printed and written in the appropriate directory.
Random seed is recycled inside the function that will import the VCF file before
tidying.
Default: <code>random.seed = NULL</code>.</p></li>
</ol><p><strong>tidying arguments/behavior:</strong></p><ol><li><p><code>tidy.vcf:</code> (optional, logical)
Default: <code>tidy.vcf = TRUE</code>. But you can always stop the process after
the creation of the GDS file (equivalent of running <code><a href="read_vcf.html">read_vcf</a></code>).</p></li>
<li><p><code>tidy.check:</code> (optional, logical)
Default: <code>tidy.check = TRUE</code>. By default, the number of markers just before
tidying is checked. Tidying a VCF file with more than 20000 markers is
sub-optimal:</p><ul><li><p>a computer with lots of RAM is required</p></li>
<li><p>it's very slow to generate</p></li>
<li><p>it's very slow to run codes after</p></li>
<li><p>for most non model species this number of markers is not realistic...</p></li>
</ul><p>Consequently, the function execution is suspended and user are asked if they
still want to continue with the tidying or stop and keep the GDS file/object.</p>
<p>This behavior can be annoying, <em>if the user knows what he's doing</em>, to turn off
use: <code>tidy.check = FALSE</code>.</p></li>
<li><p><code>calibrate.alleles: </code> (optional, logical)
Default: <code>calibrate.alleles = FALSE</code>.
Documented in <code><a href="calibrate_alleles.html">calibrate_alleles</a></code>.</p></li>
<li><p><code>vcf.stats: </code> (optional, logical) Generates individuals and
markers statistics helpful for filtering.
These are very fast to generate and because computational
cost is minimal, even for huge VCFs, the default is <code>vcf.stats = TRUE</code>.</p></li>
<li><p><code>vcf.metadata: </code> (optional, logical or character string)
With <code>vcf.metadata = FALSE</code>, only the genotypes are kept (GT field)
in the tidy dataset.
With <code>vcf.metadata = TRUE</code>,
all the metadata contained in the <code>FORMAT</code> field will be kept in
the tidy data file. radiator is currently keeping and cleaning these metadata:
<code>"DP", "AD", "GL", "PL", "GQ", "HQ", "GOF", "NR", "NV", "CATG"</code>.
e.g. you only want AD and PL, <code>vcf.metadata = c("AD", "PL")</code>.
Need another metadata ? Submit a request on github...
Default: <code>vcf.metadata = TRUE</code>.</p></li>
</ol><p><strong>Filtering arguments:</strong></p><ol><li><p><code>blacklist.id: </code> (optional, character)
Default (<code>blacklist.id = NULL</code>).
Documented in <code><a href="tidy_genomic_data.html">tidy_genomic_data</a></code>.</p></li>
<li><p><code>filter.strands</code>: (optional, character)
 Default (<code>filter.strands = "blacklist"</code>).
 documented in <code><a href="read_vcf.html">read_vcf</a></code>.</p></li>
<li><p><code>whitelist.markers: </code>(optional, path)
Default: <code>whitelist.markers = NULL</code>.
Documented in <code><a href="filter_whitelist.html">filter_whitelist</a></code>.</p></li>
<li><p><code>filter.individuals.missing</code>: (double)
Default: <code>filter.individuals.missing = NULL</code>.
Documented in <code><a href="filter_individuals.html">filter_individuals</a></code>.</p></li>
<li><p><code>filter.monomorphic</code>: (logical)
Default: <code>filter.monomorphic = TRUE</code>.
Documented in <code><a href="filter_monomorphic.html">filter_monomorphic</a></code>.
Required package: <code>UpSetR</code>.</p></li>
<li><p><code>filter.common.markers</code>: (logical)
Default: <code>filter.common.markers = TRUE</code>.
Documented in <code><a href="filter_common_markers.html">filter_common_markers</a></code>.
Required package: <code>UpSetR</code>.</p></li>
<li><p><code>filter.ma</code>: (integer)
Default: <code>filter.ma = NULL</code>.
Documented in <code><a href="filter_ma.html">filter_ma</a></code>.</p></li>
<li><p><code>filter.coverage</code>: (logical)
Default: <code>filter.coverage = NULL</code>.
Documented in <code><a href="filter_coverage.html">filter_coverage</a></code>.</p></li>
<li><p><code>filter.genotyping</code>: (integer)
Default: <code>filter.genotyping = NULL</code>.
Documented in <code><a href="filter_genotyping.html">filter_genotyping</a></code>.</p></li>
<li><p><code>filter.snp.position.read: </code> (optional, character, integer)
Default: <code>filter.snp.position.read = NULL</code>.
Documented in <code><a href="filter_snp_position_read.html">filter_snp_position_read</a></code>.</p></li>
<li><p><code>filter.snp.number: </code> (optional, character, integer)
Default: <code>filter.snp.number = NULL</code>.
Documented in <code><a href="filter_snp_number.html">filter_snp_number</a></code>.</p></li>
<li><p><code>filter.short.ld</code>: (optional, character)
Default: <code>filter.short.ld = NULL</code>.
Documented in <code><a href="filter_ld.html">filter_ld</a></code>.</p></li>
<li><p><code>filter.long.ld</code>: (optional, character)
Default: <code>filter.long.ld = NULL</code>.
Documented in <code><a href="filter_ld.html">filter_ld</a></code>.
Required package: <code>SNPRelate</code>.</p></li>
<li><p><code>long.ld.missing</code>: Documented in <code><a href="filter_ld.html">filter_ld</a></code>.
Default: <code>long.ld.missing = FALSE</code>.</p></li>
<li><p><code>ld.method</code>: Documented in <code><a href="filter_ld.html">filter_ld</a></code>.
Default: <code>ld.method = "r2"</code>.</p></li>
</ol></div>
    <div id="references">
    <h2>References</h2>
    <p>Danecek P, Auton A, Abecasis G et al. (2011)
The variant call format and VCFtools.
Bioinformatics, 27, 2156-2158.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="read_vcf.html">read_vcf</a></code>,
<code><a href="tidy_genomic_data.html">tidy_genomic_data</a></code>,
<code><a href="filter_rad.html">filter_rad</a></code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Thierry Gosselin <a href="mailto:thierrygosselin@icloud.com">thierrygosselin@icloud.com</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># very basic with built-in defaults (not recommended):</span></span></span>
<span class="r-in"><span><span class="va">prep.data</span> <span class="op">&lt;-</span> <span class="fu">radiator</span><span class="fu">::</span><span class="fu">tidy_vcf</span><span class="op">(</span>data <span class="op">=</span> <span class="st">"populations.snps.vcf"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Using more arguments and filters (recommended):</span></span></span>
<span class="r-in"><span><span class="va">tidy.data</span> <span class="op">&lt;-</span> <span class="fu">radiator</span><span class="fu">::</span><span class="fu">tidy_vcf</span><span class="op">(</span></span></span>
<span class="r-in"><span>    data <span class="op">=</span> <span class="st">"populations.snps.vcf"</span>,</span></span>
<span class="r-in"><span>    strata <span class="op">=</span> <span class="st">"strata_salamander.tsv"</span>,</span></span>
<span class="r-in"><span>    filter.individuals.missing <span class="op">=</span> <span class="st">"outlier"</span>,</span></span>
<span class="r-in"><span>    filter.ma <span class="op">=</span> <span class="fl">4</span>,</span></span>
<span class="r-in"><span>    filter.genotyping <span class="op">=</span> <span class="fl">0.1</span>,</span></span>
<span class="r-in"><span>    filter.snp.position.read <span class="op">=</span> <span class="st">"outliers"</span>,</span></span>
<span class="r-in"><span>    filter.short.ld <span class="op">=</span> <span class="st">"mac"</span>,</span></span>
<span class="r-in"><span>    path.folder <span class="op">=</span> <span class="st">"salamander/prep_data"</span>,</span></span>
<span class="r-in"><span>    verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

