<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Transform common genomic dataset format in a tidy data frame — tidy_genomic_data • radiator</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Transform common genomic dataset format in a tidy data frame — tidy_genomic_data"><meta property="og:description" content="Transform genomic data set produced by massive parallel
sequencing pipeline (e.g.GBS/RADseq,
SNP chip, DArT, etc) into a tidy format.
The use of blacklist and whitelist along
several filtering options are available to prune the dataset.
Several arguments are available to make your data population-wise and easily
rename the pop id.
Used internally in radiator
and assigner
and might be of interest for users."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/radiator" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Transform common genomic dataset format in a tidy data frame</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/radiator/blob/HEAD/R/radiator_tidy.R" class="external-link"><code>R/radiator_tidy.R</code></a></small>
    <div class="hidden name"><code>tidy_genomic_data.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Transform genomic data set produced by massive parallel
sequencing pipeline (e.g.GBS/RADseq,
SNP chip, DArT, etc) into a tidy format.
The use of blacklist and whitelist along
several filtering options are available to prune the dataset.
Several arguments are available to make your data population-wise and easily
rename the pop id.
Used internally in <a href="https://github.com/thierrygosselin/radiator" class="external-link">radiator</a>
and <a href="https://github.com/thierrygosselin/assigner" class="external-link">assigner</a>
and might be of interest for users.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">tidy_genomic_data</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  strata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filename <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>14 options for input (<strong>diploid data only</strong>): VCFs (SNPs or Haplotypes,
to make the vcf population ready),
plink (tped, bed), stacks haplotype file, genind (library(adegenet)),
genlight (library(adegenet)), gtypes (library(strataG)), genepop, DArT,
and a data frame in long/tidy or wide format. To verify that radiator detect
your file format use <code><a href="detect_genomic_format.html">detect_genomic_format</a></code> (see example below).
Documented in <strong>Input genomic datasets</strong> of <code>tidy_genomic_data</code>.</p>
<p><strong>DArT and VCF data</strong>: <span class="pkg">radiator</span> was not meant to generate alleles
and genotypes if you are using a VCF file with no genotype
(only genotype likelihood: GL or PL).
Neither is <span class="pkg">radiator</span> able to magically generate a genind object
from a SilicoDArT dataset. Please look at the first few lines of your dataset
to understand it's limit before asking raditor to convert or filter your dataset.</p></dd>


<dt id="arg-strata">strata<a class="anchor" aria-label="anchor" href="#arg-strata"></a></dt>
<dd><p>(optional)
The strata file is a tab delimited file with a minimum of 2 columns headers:
<code>INDIVIDUALS</code> and <code>STRATA</code>. Documented in <code><a href="read_strata.html">read_strata</a></code>.
DArT data: a third column <code>TARGET_ID</code> is required.
Documented on <code><a href="read_dart.html">read_dart</a></code>. Also use the strata read function to
blacklist individuals.
Default: <code>strata = NULL</code>.</p></dd>


<dt id="arg-filename">filename<a class="anchor" aria-label="anchor" href="#arg-filename"></a></dt>
<dd><p>(optional) The function uses <code><a href="http://www.fstpackage.org/reference/write_fst.html" class="external-link">write.fst</a></code>,
to write the tidy data frame in
the working directory. The file extension appended to
the <code>filename</code> provided is <code>.rad</code>.
With default: <code>filename = NULL</code>, the tidy data frame is
in the global environment only (i.e. not written in the working directory...).</p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>(optional) The number of core used for parallel
execution during import.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>(optional, logical) When <code>verbose = TRUE</code>
the function is a little more chatty during execution.
Default: <code>verbose = TRUE</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>(optional) To pass further arguments for fine-tuning the function.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>The output in your global environment is a tidy data frame.
If <code>filename</code> is provided, the tidy data frame is also
written in the working directory with file extension <code>.rad</code>.
The file is written with the
<a href="https://github.com/fstpackage/fst" class="external-link">Lightning Fast Serialization of Data Frames for R</a> package.
To read the file back in R use <code><a href="http://www.fstpackage.org/reference/write_fst.html" class="external-link">read.fst</a></code>.</p>
    </div>
    <div id="input-genomic-datasets">
    <h2>Input genomic datasets</h2>


<ol><li><p>VCF files must end with <code>.vcf</code>: documented in <code><a href="tidy_vcf.html">tidy_vcf</a></code></p></li>
<li><p>PLINK files must end with <code>.tped</code> or <code>.bed</code>: documented in <code><a href="tidy_plink.html">tidy_plink</a></code></p></li>
<li><p>genind object from
<a href="https://github.com/thibautjombart/adegenet" class="external-link">adegenet</a>:
documented in <code><a href="tidy_genind.html">tidy_genind</a></code>.</p></li>
<li><p>genlight object from
<a href="https://github.com/thibautjombart/adegenet" class="external-link">adegenet</a>:
documented in <code><a href="tidy_genlight.html">tidy_genlight</a></code>.</p></li>
<li><p>gtypes object from
<a href="https://github.com/EricArcher/strataG" class="external-link">strataG</a>:
documented in <code><a href="tidy_gtypes.html">tidy_gtypes</a></code>.</p></li>
<li><p>dart data from <a href="http://www.diversityarrays.com" class="external-link">DArT</a>:
documented in <code><a href="read_dart.html">read_dart</a></code>.</p></li>
<li><p>genepop file must end with <code>.gen</code>, documented in <code><a href="tidy_genepop.html">tidy_genepop</a></code>.</p></li>
<li><p>fstat file must end with <code>.dat</code>, documented in <code><a href="tidy_fstat.html">tidy_fstat</a></code>.</p></li>
<li><p>haplotype file created in STACKS (e.g. <code>data = "batch_1.haplotypes.tsv"</code>).
To make the haplotype file population ready, you need the <code>strata</code> argument.</p></li>
<li><p>Data frames: documented in <code><a href="tidy_wide.html">tidy_wide</a></code></p></li>
</ol></div>
    <div id="advance-mode">
    <h2>Advance mode</h2>



<p><em>dots-dots-dots ...</em> allows to pass several arguments for fine-tuning the function:</p><ol><li><p><code>vcf.metadata</code> (optional, logical or string).
Default: <code>vcf.metadata = TRUE</code>. Documented in <code><a href="tidy_vcf.html">tidy_vcf</a></code>.</p></li>
<li><p><code>vcf.stats</code> (optional, logical).
Default: <code>vcf.stats = TRUE</code>.
Documented in <code><a href="tidy_vcf.html">tidy_vcf</a></code>.</p></li>
<li><p><code>whitelist.markers</code> (optional, path or object) To keep only markers in a whitelist.
Default <code>whitelist.markers = NULL</code>.
Documented in <code><a href="read_whitelist.html">read_whitelist</a></code>.</p></li>
<li><p><code>blacklist.id</code> (optional) Default: <code>blacklist.id = NULL</code>.
Ideally, managed in the strata file.
Documented in <code><a href="read_strata.html">read_strata</a></code> and <code><a href="read_blacklist_id.html">read_blacklist_id</a></code>.</p></li>
<li><p><code>filter.common.markers</code> (optional, logical).
Default: <code>filter.common.markers = TRUE</code>,
Documented in <code><a href="filter_common_markers.html">filter_common_markers</a></code>.</p></li>
<li><p><code>filter.monomorphic</code> (logical, optional) Should the monomorphic
markers present in the dataset be filtered out ?
Default: <code>filter.monomorphic = TRUE</code>.
Documented in <code><a href="filter_monomorphic.html">filter_monomorphic</a></code>.</p></li>
</ol></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="detect_genomic_format.html">detect_genomic_format</a></code> and <code><a href="genomic_converter.html">genomic_converter</a></code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Thierry Gosselin <a href="mailto:thierrygosselin@icloud.com">thierrygosselin@icloud.com</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co">#To verify your file is detected by radiator as the correct format:</span></span></span>
<span class="r-in"><span><span class="fu">radiator</span><span class="fu">::</span><span class="fu"><a href="detect_genomic_format.html">detect_genomic_format</a></span><span class="op">(</span>data <span class="op">=</span> <span class="st">"populations.snps.vcf"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># using VCF file as input</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/SeqArray" class="external-link">SeqArray</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tidy.vcf</span> <span class="op">&lt;-</span> <span class="fu">tidy_genomic_data</span><span class="op">(</span></span></span>
<span class="r-in"><span>   data <span class="op">=</span> <span class="st">"populations.snps.vcf"</span>, strata <span class="op">=</span> <span class="st">"strata.treefrog.tsv"</span>,</span></span>
<span class="r-in"><span>   whitelist.markers <span class="op">=</span> <span class="st">"whitelist.vcf.txt"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

