<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Import genepop file and convert to a tidy dataframe — tidy_genepop • radiator</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Import genepop file and convert to a tidy dataframe — tidy_genepop" />
<meta property="og:description" content="Used internally in radiator
and might be of interest for users.
The function tidy_genepop reads a file in the
genepop format
(see details and note for convention) and output a data frame in wide or long/tidy format.
To manipulate and prune the dataset prior to tidying, use the functions
tidy_genomic_data and
genomic_converter, that uses blacklist and whitelist along
several other filtering options." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/thierrygosselin/radiator">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Import genepop file and convert to a tidy dataframe</h1>
    <small class="dont-index">Source: <a href='https://github.com/thierrygosselin/radiator/blob/master/R/genepop.R'><code>R/genepop.R</code></a></small>
    <div class="hidden name"><code>tidy_genepop.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Used internally in <a href='https://github.com/thierrygosselin/radiator'>radiator</a>
and might be of interest for users.
The function <code>tidy_genepop</code> reads a file in the
<a href='https://genepop.curtin.edu.au/help_input.html'>genepop format</a>
(see details and note for convention) and output a data frame in wide or long/tidy format.</p>
<p>To manipulate and prune the dataset prior to tidying, use the functions
<code><a href='tidy_genomic_data.html'>tidy_genomic_data</a></code> and
<code><a href='genomic_converter.html'>genomic_converter</a></code>, that uses blacklist and whitelist along
several other filtering options.</p>
    </div>

    <pre class="usage"><span class='fu'>tidy_genepop</span><span class='op'>(</span><span class='va'>data</span>, strata <span class='op'>=</span> <span class='cn'>NULL</span>, tidy <span class='op'>=</span> <span class='cn'>TRUE</span>, filename <span class='op'>=</span> <span class='cn'>NULL</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>data</th>
      <td><p>A <a href='https://genepop.curtin.edu.au/help_input.html'>genepop</a>
filename with extension <code>.gen</code>.</p></td>
    </tr>
    <tr>
      <th>strata</th>
      <td><p>(optional) A tab delimited file with 2 columns. Header:
<code>INDIVIDUALS</code> and <code>STRATA</code>.
The <code>STRATA</code> column can be any hierarchical grouping.
To create a strata file see <code><a href='individuals2strata.html'>individuals2strata</a></code>.
Default: <code>strata = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>tidy</th>
      <td><p>(optional, logical) With <code>tidy = FALSE</code>,
the markers are the variables and the genotypes the observations (wide format).
With the default: <code>tidy = TRUE</code>, markers and genotypes are variables
with their own columns (long format).</p></td>
    </tr>
    <tr>
      <th>filename</th>
      <td><p>(optional) The file name for the tidy data frame
written to the working directory.
Default: <code>filename = NULL</code>, the tidy data is
in the global environment only (i.e. not written in the working directory).</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The output in your global environment is a wide or long/tidy data frame.
If <code>filename</code> is provided, the wide or long/tidy data frame is also
written to the working directory.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><a href='https://genepop.curtin.edu.au/help_input.html'>genepop format</a></p><ol>
<li><p><strong>First line:</strong> This line is used to store information about
your data, any characters are allowed.
This line is not kept inside <code>tidy_genepop</code>.</p></li>
<li><p><strong>Second line:</strong> 2 options i) wide: all the locus are stored on this same
line with comma (or a comma+space) separator;
ii) long: the name of the first locus
(the remaining locus are on separate and subsequent rows with the long format:
not recommended with genomic datasets with thousands of markers...).</p>
<p>The remaining lines are blocks of population and genotypes,
<a href='https://genepop.curtin.edu.au/help_input.html'>for genepop format examples</a>.</p></li>
<li><p><strong>population identifier:</strong> The population block are separated by
the word: <code>POP</code>, or <code>Pop</code> or <code>pop</code>. Flavors of the genepop
software uses the first or the last identifier of every sub-population,
in all output files, to name populations
<a href='https://genepop.curtin.edu.au/help_input.html'>(more info)</a>.
This is not very convenient for population naming and prone to errors,
this is where the <code>strata</code> argument inside <code>tidy_genepop</code>
(described above) becomes handy.</p></li>
<li><p><strong>individual identifier:</strong> After the population identifier
the individuals that belong to the same population are found subsequently on
separate lines. The genepop format specify you can use any character,
including a blank space or tab. Spaces are allowed in the identifier names.
You may leave it blank except for a comma if you wish. The comma between
the individual identifier and the list of genotypes is required.
For good naming habit however the function <code>tidy_genepop</code> will
replace <code>"_", ":"</code> with <code>"-"</code>, the comma <code>","</code> along any white
space characters defined as <code>"\t", "\n", "\f", "\r", "\p{Z}"</code> found in the individual
name will be trimmed.</p></li>
<li><p><strong>genotypes:</strong> For each locus, genotypes are separated by one or
more blank spaces or tab. 0101 indicates that this individual is homozygous
for the 01 allele at the first locus.
An alternative input format exists, where each allele is coded by three digits
(instead of two as described above). However, the total number of
different alleles, for each locus, should not be higher than 99.
Missing data is coded with zeros: <code>0000 or 000000</code>.</p></li>
</ol>

    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p><a href='https://genepop.curtin.edu.au/help_input.html'>genepop format notes:</a></p><ul>
<li><p>No constraint on blanks separating the various fields.</p></li>
<li><p>tabs or spaces allowed.</p></li>
<li><p>Loci names can appear on separate lines (long), or on one line (wide)
if separated by commas.</p></li>
<li><p>Individual identifier may have blanks but must end with a comma.</p></li>
<li><p>Alleles are numbered from 01 to 99 (or 001 to 999).
Consecutive numbers to designate alleles are not required.</p></li>
<li><p>Populations are defined by the position of the "Pop" separator.
To group various populations, just remove relevant "Pop" separators.</p></li>
<li><p>Missing data should be indicated as 00 (or 000) rather than blanks.
There are three possibilities for missing data :
no information (0000) or (000000), partial information
for first allele (1000) or (010000), partial information
for second allele (0010) or (000010).</p></li>
<li><p>The number of locus names should correspond to the number of genotypes
in each row.
If you remove one or several loci from your input file,
you should remove both their names and the corresponding genotypes.</p></li>
<li><p>No empty lines should be found within the file.</p></li>
<li><p>No more than one empty line should be present at the end of file.</p></li>
</ul>

    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Raymond M. &amp; Rousset F, (1995).
GENEPOP (version 1.2): population genetics software for exact tests
and ecumenicism.
J. Heredity, 86:248-249</p>
<p>Rousset F.
genepop'007: a complete re-implementation of the genepop software
for Windows and Linux.
Molecular Ecology Resources.
2008, 8: 103-106.
doi:10.1111/j.1471-8286.2007.01931.x</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='https://genepop.curtin.edu.au'>genepop</a></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Thierry Gosselin <a href='mailto:thierrygosselin@icloud.com'>thierrygosselin@icloud.com</a></p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='co'># We will use the genepop dataset provided with adegenet package</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>require</a></span><span class='op'>(</span><span class='st'><a href='https://github.com/thibautjombart/adegenet'>"adegenet"</a></span><span class='op'>)</span>

<span class='co'># The simplest form of the function:</span>
<span class='va'>nancycats.tidy</span> <span class='op'>&lt;-</span> <span class='fu'>radiator</span><span class='fu'>::</span><span class='fu'>tidy_genepop</span><span class='op'>(</span>
    data <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span>
        <span class='st'>"files/nancycats.gen"</span>,
        package <span class='op'>=</span> <span class='st'>"adegenet"</span>
        <span class='op'>)</span>
    <span class='op'>)</span>

<span class='co'># To output a data frame in wide format, with markers in separate columns:</span>
<span class='va'>nancycats.wide</span> <span class='op'>&lt;-</span> <span class='fu'>radiator</span><span class='fu'>::</span><span class='fu'>tidy_genepop</span><span class='op'>(</span>
    data <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span>
    <span class='st'>"files/nancycats.gen"</span>,
    package<span class='op'>=</span><span class='st'>"adegenet"</span>
<span class='op'>)</span>,
    tidy <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span>
<span class='op'>}</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Thierry Gosselin, Manuel Lamothe, Floriaan Devloo-Delva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


