<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Write a dadi SNP input file from a tidy data frame. — write_dadi • radiator</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Write a dadi SNP input file from a tidy data frame. — write_dadi"><meta property="og:description" content="This function will generate a dadi SNP input file using a
radiator tidy tibble.
Note that missing data can potentially bias demographic inference,
consequently you might want to check the impact of missing data by running
two datset in dadi: one with the missing data or several with varying thresholds
of missingness, and one with imputed genotypes
(check my other package grur for this)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/radiator" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Write a <code>dadi</code> SNP input file from a tidy data frame.</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/radiator/blob/HEAD/R/write_dadi.R" class="external-link"><code>R/write_dadi.R</code></a></small>
    <div class="hidden name"><code>write_dadi.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function will generate a <code>dadi</code> SNP input file using a
radiator tidy tibble.</p>
<p>Note that missing data can potentially bias demographic inference,
consequently you might want to check the impact of missing data by running
two datset in <code>dadi</code>: one with the missing data or several with varying thresholds
of missingness, and one with imputed genotypes
(check my other package <a href="https://thierrygosselin.github.io/grur/" class="external-link">grur</a> for this).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">write_dadi</span><span class="op">(</span>
  <span class="va">data</span>,
  fasta.ingroup <span class="op">=</span> <span class="cn">NULL</span>,
  fasta.outgroup <span class="op">=</span> <span class="cn">NULL</span>,
  sumstats.ingroup <span class="op">=</span> <span class="cn">NULL</span>,
  sumstats.outgroup <span class="op">=</span> <span class="cn">NULL</span>,
  dadi.input.filename <span class="op">=</span> <span class="cn">NULL</span>,
  calibrate.alleles <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>data</dt>
<dd><p>A tidy data frame object in the global environment or
a tidy data frame in wide or long format in the working directory.
<em>How to get a tidy data frame ?</em>
Look into <span class="pkg">radiator</span> <code><a href="tidy_genomic_data.html">tidy_genomic_data</a></code>.</p></dd>
<dt>fasta.ingroup</dt>
<dd><p>(optional) The fasta file, sequences for the ingroup.
Leave empty if no outgroup.
Default: <code>fasta.ingroup = NULL</code>.</p></dd>
<dt>fasta.outgroup</dt>
<dd><p>(optional) The fasta file, sequences for the outgroup.
Default: <code>fasta.outgroup = NULL</code>.</p></dd>
<dt>sumstats.ingroup</dt>
<dd><p>(optional) The sumstats output file from STACKS
when running STACKS for the ingroup fasta file.This file is
required to use with an outgroup. Leave empty if no outgroup.
Default: <code>sumstats.ingroup = NULL</code>.</p></dd>
<dt>sumstats.outgroup</dt>
<dd><p>(optional) The sumstats output file from STACKS
when running STACKS for the outgroup fasta file. This file is
required to use an outgroup.
Default: <code>sumstats.outgroup = NULL</code>.</p></dd>
<dt>dadi.input.filename</dt>
<dd><p>(optional) Name of the <code>dadi</code> SNP input file
written to the working directory. e.g. <code>dadi.file.tsv</code>.
Default use date and time to make the file. If used, the file extension
need to finish with <code>.tsv or .txt</code>.
Default: <code>dadi.input.filename = NULL</code>.</p></dd>
<dt>calibrate.alleles</dt>
<dd><p>(optional, logical) To re-calibrate REF an ALT alleles.
Will be done automatically to the dataset if the required genomic format is
not found. Please use if you have removed individuals.
Default: <code>calibrate.alleles = FALSE</code>.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>The function returns tibble and the dadi input file in the working
directory.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Gutenkunst RN, Hernandez RD, Williamson SH, Bustamante CD (2009)
Inferring the Joint Demographic History of Multiple Populations
from Multidimensional SNP Frequency Data (G McVean, Ed,).
PLoS genetics, 5, e1000695.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Thierry Gosselin <a href="mailto:thierrygosselin@icloud.com">thierrygosselin@icloud.com</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in"><span class="co"># without outgroup:</span></span>
<span class="r-in"><span class="va">dadi.data</span> <span class="op">&lt;-</span> <span class="fu">radiator</span><span class="fu">::</span><span class="fu">write_dadi</span><span class="op">(</span>data <span class="op">=</span> <span class="st">"my_tidy_dataset.rad"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># with outgroup and fasta generated by stacks:</span></span>
<span class="r-in"><span class="va">dadi.data</span> <span class="op">&lt;-</span> <span class="fu">radiator</span><span class="fu">::</span><span class="fu">write_dadi</span><span class="op">(</span></span>
<span class="r-in">   data <span class="op">=</span> <span class="st">"my_tidy_dataset.rad"</span>,</span>
<span class="r-in">   fasta.ingroup <span class="op">=</span> <span class="st">"batch_1.ingroup.fa"</span>,</span>
<span class="r-in">   fasta.outgroup <span class="op">=</span> <span class="st">"batch_1.outgroup.fa"</span>,</span>
<span class="r-in">   sumstats.ingroup <span class="op">=</span> <span class="st">"batch_1.sumstats.ingroup.tsv"</span>,</span>
<span class="r-in">   sumstats.outgroup <span class="op">=</span> <span class="st">"batch_1.sumstats.outgroup.tsv"</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer></div>

  


  

  </body></html>

