<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Write a COLONY input file — write_colony • radiator</title><!-- favicons --><link rel="icon" type="image/png" sizes="48x48" href="../favicon-48x48.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Write a COLONY input file — write_colony"><meta property="og:description" content="Write a COLONY input file."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/radiator" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Write a <code>COLONY</code> input file</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/radiator/blob/HEAD/R/write_colony.R" class="external-link"><code>R/write_colony.R</code></a></small>
    <div class="hidden name"><code>write_colony.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Write a <code>COLONY</code> input file.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">write_colony</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  strata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample.markers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pop.select <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  allele.freq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  inbreeding <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  mating.sys.males <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  mating.sys.females <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  clone <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  run.length <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  analysis <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  allelic.dropout <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  error.rate <span class="op">=</span> <span class="fl">0.02</span>,</span>
<span>  print.all.colony.opt <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  random.seed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>  filename <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>(4 options) A file or object generated by radiator:</p><ul><li><p>tidy data</p></li>
<li><p>Genomic Data Structure (GDS)</p></li>
</ul><p><em>How to get GDS and tidy data ?</em>
Look into <code><a href="tidy_genomic_data.html">tidy_genomic_data</a></code>,
<code><a href="read_vcf.html">read_vcf</a></code> or
<code><a href="tidy_vcf.html">tidy_vcf</a></code>.</p></dd>


<dt id="arg-strata">strata<a class="anchor" aria-label="anchor" href="#arg-strata"></a></dt>
<dd><p>(path or object) The strata file or object.
Additional documentation is available in <code><a href="read_strata.html">read_strata</a></code>.
Use that function to whitelist/blacklist populations/individuals.
Option to set <code>pop.levels/pop.labels</code> is also available.</p></dd>


<dt id="arg-sample-markers">sample.markers<a class="anchor" aria-label="anchor" href="#arg-sample-markers"></a></dt>
<dd><p>(number) <code>COLONY</code> can take a long time to run,
use a random subsample of your markers to speed test <code>COLONY</code>
e.g. <code>sample.markers = 500</code> to use only 500 randomly chosen markers.
Default: <code>sample.markers = NULL</code>, will use all markers.</p></dd>


<dt id="arg-pop-select">pop.select<a class="anchor" aria-label="anchor" href="#arg-pop-select"></a></dt>
<dd><p>(optional, string) Selected list of populations for
the analysis. e.g. <code>pop.select = c("QUE", "ONT")</code> to select <code>QUE</code>
and <code>ONT</code> population samples (out of 20 pops). If <code>pop.labels</code>
argument was used to rename the strata column, use the new names with
<code>pop.select</code>.
Default: <code>pop.select = NULL</code></p></dd>


<dt id="arg-allele-freq">allele.freq<a class="anchor" aria-label="anchor" href="#arg-allele-freq"></a></dt>
<dd><p>(optional, string) Allele frequency can be computed from
a select group.
e.g. <code>allele.freq = "QUE"</code> or <code>allele.freq = c("QUE", "ONT")</code>.
Using <code>allele.freq = "overall"</code> will use all the samples to compute the
allele frequency.
Default: <code>allele.freq = NULL</code>, will not compute allele frequency.</p></dd>


<dt id="arg-inbreeding">inbreeding<a class="anchor" aria-label="anchor" href="#arg-inbreeding"></a></dt>
<dd><p>(boolean) 0/1 no inbreeding/inbreeding.
Default: <code>inbreeding = 0</code></p></dd>


<dt id="arg-mating-sys-males">mating.sys.males<a class="anchor" aria-label="anchor" href="#arg-mating-sys-males"></a></dt>
<dd><p>(boolean) Mating system in males.
0/1 polygyny/monogyny.
Default: <code>mating.sys.males = 0</code>.</p></dd>


<dt id="arg-mating-sys-females">mating.sys.females<a class="anchor" aria-label="anchor" href="#arg-mating-sys-females"></a></dt>
<dd><p>(boolean) Mating system in females.
0/1 polygyny/monogyny.
Default: <code>mating.sys.females = 0</code>.</p></dd>


<dt id="arg-clone">clone<a class="anchor" aria-label="anchor" href="#arg-clone"></a></dt>
<dd><p>(boolean) Should clones and duplicated individuals be inferred.
0/1, yes/no. Default: <code>clone = 0</code>.</p></dd>


<dt id="arg-run-length">run.length<a class="anchor" aria-label="anchor" href="#arg-run-length"></a></dt>
<dd><p>(integer) Length of run. 1 (short), 2 (medium), 3 (long),
4 (very long). Start with short or medium run and consider longer run if your
estimates probability are not stable or really good.
Default: <code>run.length = 2</code>.</p></dd>


<dt id="arg-analysis">analysis<a class="anchor" aria-label="anchor" href="#arg-analysis"></a></dt>
<dd><p>(integer) Analysis method.
0 (Pairwise-Likelihood Score), 1 (Full Likelihood),
2 (combined Pairwise-Likelihood Score and Full Likelihood).
Default: <code>analysis = 1</code>.</p></dd>


<dt id="arg-allelic-dropout">allelic.dropout<a class="anchor" aria-label="anchor" href="#arg-allelic-dropout"></a></dt>
<dd><p>Locus allelic dropout rate.
Default : <code>allelic.dropout = 0</code>.</p></dd>


<dt id="arg-error-rate">error.rate<a class="anchor" aria-label="anchor" href="#arg-error-rate"></a></dt>
<dd><p>Locus error rate.
Default:<code>error.rate = 0.02</code>.</p></dd>


<dt id="arg-print-all-colony-opt">print.all.colony.opt<a class="anchor" aria-label="anchor" href="#arg-print-all-colony-opt"></a></dt>
<dd><p>(logical) Should all <code>COLONY</code> options be printed in the file.</p>
<p><strong>This require manual curation, for the file to work directly with <code>COLONY</code></strong>.
Default = <code>print.all.colony.opt = FALSE</code>.</p></dd>


<dt id="arg-random-seed">random.seed<a class="anchor" aria-label="anchor" href="#arg-random-seed"></a></dt>
<dd><p>(integer, optional) For reproducibility, set an integer
that will be used inside the function that requires randomness. With default,
a random number is generated and printed in the appropriate output.
Default: <code>random.seed = NULL</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>(optional, logical) When <code>verbose = TRUE</code>
the function is a little more chatty during execution.
Default: <code>verbose = TRUE</code>.</p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>(optional) The number of core used for parallel
execution during import.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></dd>


<dt id="arg-filename">filename<a class="anchor" aria-label="anchor" href="#arg-filename"></a></dt>
<dd><p>Name of the acronym for filenaming in the working directory.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>(optional) Advance mode that allows to pass further arguments
for fine-tuning the function. Also used for legacy arguments (see details or
special section)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code>COLONY</code> file in your working directory (2 if you selected imputations arguments...)</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><strong>It is highly recommended to read (twice!) the user guide distributed with
<code>COLONY</code> to find out the details for input and output of the software.</strong></p>
<p>Not all options are provided here.</p>
<p>But to ease the process, all the required options to properly run <code>COLONY</code>
will be printed in the file written in your working directory.
Change the values accordingly and wisely.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Jones OR, Wang J (2010) COLONY: a program for parentage and
sibship inference from multilocus genotype data.
Molecular Ecology Resources, 10, 551–555.</p>
<p>Wang J (2012) Computationally Efficient Sibship and
Parentage Assignment from Multilocus Marker Data. Genetics, 191, 183–194.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code>COLONY</code> is available on Jinliang Wang web site
<a href="https://www.zsl.org/science/software/colony" class="external-link">https://www.zsl.org/science/software/colony</a></p>

<p><a href="https://thierrygosselin.github.io/radiator/articles/rad_genomics_computer_setup.html#colony" class="external-link">colony installation instructions</a></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Thierry Gosselin <a href="mailto:thierrygosselin@icloud.com">thierrygosselin@icloud.com</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># Simplest way to run the function with a tidy dataset:</span></span></span>
<span class="r-in"><span><span class="va">colony.file</span> <span class="op">&lt;-</span> <span class="fu">radiator</span><span class="fu">::</span><span class="fu">write_colony</span><span class="op">(</span>data <span class="op">=</span> <span class="st">"turtle.data.rad"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

