# remove NOTE about no visible binding for global variable during R CMD check --
if (getRversion() >= "2.15.1") {
  utils::globalVariables(
    c("ID", "CloneID", "SnpPosition", "CallRate", "AvgCountRef", "AvgCountSnp",
      "RepAvg", "NOT_USEFUL", "SNP", "CALL_RATE", "AVG_COUNT_REF",
      "AVG_COUNT_SNP", "REP_AVG", "NEW_ID", "SNP_N", "ALLELE_NAME", "ALLELE_NUMBER",
      "ALLELES_COUNT", "GENOTYPED_PROP", "MISSING_IND_PROP", "AlleleID", "HET_NUMBER",
      "HET_PERCENT", "HET_PROP", "everything", "DP", "AD", "vcf.headers",
      "GT_VCF", "INDIVIDUALS2", "ALLELE_REF_DEPTH",
      "ALLELE_ALT_DEPTH", "GT_BIN", "GT_HAPLO", "REF_NEW", "REF_ALT_CHANGE",
      "GT_VCF_A1", "GT_VCF_A2", "MAF", "Allele1", "Allele2", "POP", "IN_GROUP",
      "OUT_GROUP", "ID.FILTER", "ANCESTRAL", "SEQUENCES", "GARBAGE", "SNP_READ_POS",
      "FASTA_REF", "BP", "Chr", "Locus", "Locus ID", "Col", "PP", "ALLELE_GROUP",
      "PROBLEM", "IND_LEVEL_POLYMORPHISM", "HOM", "HET", "N_GENOT", "DIPLO",
      "FREQ_ALLELES", "HOM_E", "HOM_O", "FH", "HET_O", "HET_E", "PI", "pi",
      "MONOMORPHIC", "POLYMORPHIC", "CONSENSUS", "PARALOGS", "Seg Dist",
      "REF.x", "ALT.x", "REF.y", "ALT.y", "BLACKLIST", "ALLELE_COVERAGE_RATIO",
      "..scaled..", "GENOTYPE_LIKELIHOOD_GROUP", "GL_MAX", "GL_MIN", "VALUE",
      "GL_DIFF", "ALLELE_ALT_DEPTH_NEW", "ALLELE_REF_DEPTH_NEW", "ALT_NEW",
      "CHANGE", "n.al.pop", "n.al.tot", "TOTAL_READ", "Missingness",
      "MISSING_GENOTYPE", "INDIVIDUALS_NUMBER", "PERC", "Axis.1", "Axis.2", "V1",
      "Axis.3", "Axis.4", "ids", "GQ", "PL", "ARTIFACTS", "GENOTYPED",
      "IND_THRESHOLD", "N_IND", "PERCENT", "POP_GENOTYPED", "PROB_POP", "LOCUS",
      "HET_DIFF", "HET_O", "HET_MAX", "GT", "GL", "INDIVIDUALS", "POP_ID", "N",
      "HET_O", "HOM_O", "HET_E", "HOM_E", "FREQ_ALT", "FREQ_REF",
      "GLOBAL_MAF", "PP", "PQ", "QQ", "AND", "OR", "filters.parameters",
      "het.global.summary", "het.global.threshold",
      "plot.distribution.het.local", "violinplot.het.global",
      "violinplot.het.local", "HET_GROUP", "HET_MIN",
      "HET_THRESHOLD", "POP_THRESHOLD", "PROP", "0.4", "0.9","DIF_0.1", "DIF_0.9",
      "DIF_OUTLIERS", "DIF_THRESHOLD", "HET_DIF", "HET_MEAN", "MAX_0.4",
      "MAX_0.9", "MAX_DIF_OUTLIERS", "MAX_DIF_THRESHOLD", "MAX_OUTLIERS",
      "MAX_THRESHOLD", "OUTLIERS", "THRESHOLD", "WHITELIST", "het.dif.threshold",
      "ALT2", "REF2", "#CHROM", "A1", "A2", "ALLELES", "ALLELES_GROUP", "ALT",
      "COL1", "COL3", "COL4", "COUNT", "ERASE", "FILTER", "FORMAT", "FORMAT_ID",
      "INDIVIDUALS_ALLELES", "INFO", "KEEPER", "MAF_GLOBAL", "MAF_LOCAL",
      "MARKERS", "MARKERS_ALLELES", "MARKER_NUMBER", "MAX_COUNT_MARKERS",
      "NUMBER", "QUAL", "READ_DEPTH", "REF", "STRATA", "TOTAL",
      "path.folder.step1", "path.folder.step2", "sum.pop", "EIGENVALUES",
      "MISSING_GENOTYPE_PROP", "A1_A2", "ALLELE_COPIES", "READ_DEPTH_MEAN",
      "READ_DEPTH_MIN", "READ_DEPTH_MAX", "ALLELE_REF_DEPTH", "ALLELE_REF_DEPTH_MEAN",
      "ALLELE_REF_DEPTH_MIN", "ALLELE_REF_DEPTH_MAX", "ALLELE_ALT_DEPTH",
      "ALLELE_ALT_DEPTH_MEAN", "ALLELE_ALT_DEPTH_MIN", "ALLELE_ALT_DEPTH_MAX",
      "GL", "GL_MEAN", "ALLELE_SUM", "FIS", "FIS_MAX", "FIS_MIN", "FIS_DIFF",
      "FREQ", "HOM_ALT",
      "fh", "HOM_REF", "BETAI", "HB", "HW", "NN", "NN_C", "N_POP", "N_POP_C",
      "DIFFERENT", "DISTANCE", "ID1", "ID2", "IDENTICAL", "IDENTICAL_GT",
      "PAIRWISE", "POP_COMP", "IDENTICAL_PROP","IDENTICAL_PROP_TOTAL",
      "MARKERS_MISSING", "TOTAL_MARKERS_GENOTYPED", "X1",
      "HET_RANGE", "MISSING_GROUP", "MISSING_PROP", "MISSING_PROP_OVERALL",
      "MISSING_PROP_POP", ".", "ALLELE", "ALLELE1", "ALLELE2", "ALLELES_FREQ",
      "Catalog ID", "Cnt", "GENOTYPE", "GROUP", "HAPLOTYPES", "MARKERS.ALLELE_GROUP",
      "POLYMORPHISM", "ALT_ALLELE_NUMBER", "PROB_HET", "PROB_HOM_ALT", "PROB_HOM_REF",
      "NAPL", "DISTANCE_RELATIVE", "MARKERS_COMMON", "id.pairwise", "input.prep",
      "CHROM_LOCUS", "NEW_MARKERS", "POP_ID_N", "INDIVIDUALS_N", "GT_N", "GT_IMP",
      "INDIVIDUALS_REP", "KEEP", "GT_VCF_NUC", "INTEGERS", "GOF", "NR", "NV",
      "A1_NUC", "A2_NUC", "RANK", "MAF_P", "MAF_L", "BAD_ID", "strata",
      "SPLIT_VEC", "ALLELES_COUNT_OVERALL", "N_INV", "HS", "GIS", "PI_NEI",
      "FQ_FILES", "PARAMETER", "SPLIT", "mtime", "REF_ALT", "ARTIFACT",
      "HETEROZYGOSITY", "HOMOZYGOSITY", "LOCUS_NUMBER", "BLACKLIST_ARTIFACT",
      "HAPLOTYPE", "MAX_NUMBER_SNP_LOCUS", "MEAN_NUMBER_SNP_LOCUS",
      "NUMBER_LOCUS_4SNP", "SNP_LOCUS", "SNP_NUMBER", "BLACKLISTED_FLAG",
      "BLACKLIST_USTACKS", "FILTERED", "FOR_CATALOG",
      "HAPLOTYPE_NUMBER", "LOCUS_TOTAL", "LUMBERJACKSTACK_FLAG", "SEQUENCE",
      "SEQ_TYPE", "ALLELE_N", "BAYESCAN_MARKERS", "BAYESCAN_POP", "COVERAGE",
      "GENE_N", "SNP_NUM", "ALPHA", "FST", "FST_GROUP", "ITERATIONS", "LOG10_PO",
      "LOG10_Q", "POST_PROB", "PO_GROUP", "Q_VALUE", "SELECTION", "ACCURACY",
      "SELECTION_TYPE_ON_LOCUS", "Distance", "R", "A", "NEW_GT", "Markers", "READS",
      "CONSENSUS_MAX", "REF_OLD", "ORIG_GT", "GT_1", "GT_2", "TEMP", "VCF_A1",
      "VCF_A2", "IMBALANCE_PERCENT", "LOW_COVERAGE_PERCENT", "BLACKLIST_PSTACKS", "n",
      "COL", "COL_TYPE", "DELETE", "OR_POP_THRESHOLD", "POLYMORPHIC.x",
      "POLYMORPHIC.y", "EXPECTED", "FREQ_HET_E", "FREQ_HET_O", "FREQ_HOM_ALT_E",
      "FREQ_HOM_ALT_O", "FREQ_HOM_REF_E", "FREQ_HOM_REF_O", "GENOTYPES",
      "N_HET_EXP", "N_HOM_ALT_EXP", "N_HOM_REF_EXP", "OBSERVED",
      "DISCARD", "NUMBERS", "TARGET_ID", "iter", "m", "PRIVATE_ALLELES",
      "GENOTYPED_MAX", "PAIRS", "NN_G", "N_GLOBAL", "N_LOCAL", "PP_G", "QQ_G",
      "DIFF", "HAPLOTYPES_NEW", "NUC", "ALT_GLOBAL", "ALT_LOCAL", "GLOBAL",
      "LOCAL", "OVERALL", "HQ", "CHROM", "POS", "GENOTYPED_MIN", "METHOD",
      "ID1_STRATA", "ID2_STRATA", "MARKERS_COMMON_PROP", "MARKERS_MISSING",
      "MARKERS_TOTAL", "PROP_IDENTICAL", "CLONEID", "ALLELE_RATIO_LIK",
      "BLACKLIST_GENOTYPES_COMMENT", "DATA", "GENOTYPED_PROP_INDIVIDUALS",
      "GENOTYPED_PROP_MARKERS", "GL_IQR", "IQR", "MODEL", "NUMBER_INDIVIDUALS",
      "OUTLIERS_COVERAGE", "OUTLIERS_N", "Q25", "Q75", "MAX", "MEDIAN", "MIN",
      "MISSING", "OUTLIERS_HIGH", "OUTLIERS_HIGH_N", "OUTLIERS_LOW",
      "OUTLIERS_LOW_N", "OUTLIERS_TOTAL", "MEAN", "n.pop", "READ_DEPTH_SCALED",
      "**", "***", "****", "*****","AA", "AB", "BB", "COMMENTS", "FILTERS", "GROUPINGS",
      "HWE", "MID_P_VALUE", "MONO", "N_POP_HWD", "SIGNIFICANCE", "X", "p",
      "VALUES", "TOTAL_READ_COUNTS", "aswer.opt", "markers.meta", "vcf.connection",
      "ALT_COUNT", "INDIVIDUALS_VCF", "MAC", "MAC_FILTER", "REF_COUNT",
      "SNP_PER_LOCUS_MAC", "SNP_POS_READ_IQR", "SNP_POS_READ_OUTLIERS",
      "SNP_POS_READ_Q75", "VARIANT_ID", "genotypes", "NEW_POP", "NEW_INDIVIDUALS",
      "biallelic", "LD", "x", "SNPRELATE", "LONG_DISTANCE_PRUNING",
      "COVERAGE_MEAN", "FILTER_COVERAGE", "FILTER_LONG_LD",
      "FILTER_SHORT_LD", "FILTER_STRANDS", "GENOTYPED_THRESHOLD", "MARKERS_A",
      "MARKERS_B", "MISSING_POP", "NUMBER_MARKERS", "SNP_PER_LOCUS", "id.select",
      "individuals.missing", "keep.both.strands", "n.ind", "snp.select.no.maf",
      "common.markers", "FILTER_INDIVIDUALS_MISSING", "tidy.data", "num.variant",
      "STRANDS", "COVERAGE_TOTAL", "COVERAGE_MEAN", "NUMBER_SNP",
      "NUMBER_SNP_BLACKLISTED", "INDIVIDUALS_CLEAN", "MAC_GLOBAL", "LIST",
      "WHITELISTED_MARKERS", "input", "ALT_CHECK", "R_DEPTH", "A_DEPTH",
      "data.tidy", "CHECK", "parameters", "force.stats", "COVERAGE_LOW",
      "COVERAGE_HIGH", "REPRODUCIBILITY", "blacklist.id", "blacklist.genotype",
      "AM", "ARGUMENTS", "AT", "BLACKLISTED_MARKERS", "CHROM_TICK",
      "LD_SUBSAMPLE", "LD_THRESHOLD", "RM", "RT", "STATS",
      "blacklist.markers.snp.number", "blacklist.snp.number.markers", "data",
      "data.temp", "detect.mixed.genomes", "detect.duplicate.genomes",
      "dev.off", "dp", "file.date",
      "filter.common.markers", "filter.coverage", "filter.genotyping",
      "filter.individuals.coverage.total", "filter.individuals.heterozygosity",
      "filter.individuals.missing", "filter.long.ld", "filter.ma", "ma.stats",
      "filter.monomorphic", "filter.reproducibility", "filter.short.ld",
      "filter.snp.number", "filter.snp.position.read", "filter.strands",
      "filters.parameters.path", "first.data.info",
      "filter.hwe", "midp.threshold", "hw.pop.threshold",
      "imputation.method", "hierarchical.levels", "keep.gds", "ld.figures",
      "ld.method", "long.ld.missing", "max.marker", "missing.memory",
      "monomorphic.out", "num.tree", "pdf", "pop.id.levels", "pop.labels",
      "pop.levels", "pop.select", "radiator.dots", "result", "snp.ld",
      "unfiltered.data", "vcf.metadata", "whitelist.common.markers",
      "whitelist.markers", "whitelist.markers.coverage",
      "whitelist.markers.genotyping", "whitelist.markers.mac",
      "whitelist.markers.snp.position.read", "write.tidy", "ibs", "info",
      "vcf.stats", "keep.allele.names", "blacklist.genotypes",
      "gt", "gt.vcf.nuc", "gt.vcf",
      "internal", "ALT_DEPTH_MEAN", "ALT_DEPTH_TOTAL", "REF_DEPTH_MEAN",
      "REF_DEPTH_TOTAL", "MARKERS1", "SWITCH", "TARGET_ID1", "VARIANT_ID.x",
      "VARIANT_ID.y", "variant.id", "problem.filename", "FILTER_VCF", "fig.upsetr",
      "MARKERS_NEW", "C3", "C4", "C5", "C6", "ONE_RATIO_REF", "ONE_RATIO_SNP",
      "rdepth", "GENETIC_SEX", "M", "MEAN_HET_F", "MEAN_HET_M",
      "MEAN_READ_DEPTH_F", "MEAN_READ_DEPTH_M", "PRESENCE_ABSENCE_F",
      "PRESENCE_ABSENCE_M", "QR_RESIDUALS", "RATIO", "Species", "U",
      "coverage.thresholds", "population", "residuals", "species", "tau", "yw",
      "MEAN_RD", "MISSINGNESS", "dup.threshold", "ALLELE_SEQUENCE",
      "CLONE_ID", "CLUSTER_CONSENSUS_SEQUENCE", "TRIMMED_SEQUENCE",
      "INDIVIDUALS1", "SEX_MARKERS", "VARIANT_ID1", "COLLECTION",
      "NA_character", "REPUNIT", "SAMPLE_TYPE", "SampleNum", "repunit",
      "sample_type", "CALIBRATE", "DEPTH_ALT", "DEPTH_REF", "HET_PROP",
      "NUMBER_GENOTYPED", "NUMBER_HET", "NUMBER_MISSING", "TOTAL_COUNTS",
      "Z_SCORE",
      "MARKERS.x", "GT_BIN.x", "TARGET_ID.x", "TARGET_ID.y", "folder.name", "mis.threshold.data",
      "mis.threshold.silicodata", "threshold.y.markers", "threshold.y.silico.markers",
      "sex.id.input", "het.qr.input", "threshold.x.markers.qr",
      "zoom.data", "threshold.x.markers.RD", "zoom.silicodata",
      "threshold.x.markers.RD.silico", "VISUAL_SEX", "GENETIC_SEX.x",
      "GENETIC_SEX.y", "MEAN_GT", "QCcode", "alleles", "assayLSID", "center",
      "panelLSID", "pop.info", "protLSID", "strand", "MICROSATELLITES",
      "INDIVIDUALS_PLINK", "ind.heterozygosity.threshold", "INDIVIDUALS...3",
      "INDIVIDUALS...7", "MARKERS...2", "MARKERS...6",
      "VARIANT_ID...1", "VARIANT_ID...4", "VARIANT_ID...5",
      "THRESHOLDS", "THRESHOLD_OUTLIERS_HIGH",
      "THRESHOLD_OUTLIERS_LOW", "A_SUM", "N_ALLELES",
      "filter.individuals.coverage.iqr",
      "filter.individuals.coverage.median", "COMMON_META", "where", "everything",
      "M_SEQ", "ID_SEQ", "1", "2", "ORIG_GT_BIN", "STRATA_SEQ", "IDTEMP", ".data",
      "EQUAL_COUNTS", "strataG", "HET_EXP", "id", "stratum", "DOS_ALT",
      "ALLELES_ALT_A2", "ALLELES_REF_A1", "NUC_ALT_A2", "NUC_REF_A1",
      "MAD_GLOBAL_CORR", "MAC_GLOBAL_CORR", "MAF_GLOBAL_DEPTH_CORR",
      "TOTAL_COUNT", "m.info", "MAF_GLOBAL_COUNT_CORR", "MAF_GLOBAL_DEPTH_CORR",
      "MAD_GLOBAL", "i.info", "ID1_POP", "ID2_POP", "VARIANT_DART", "MIX_RISK",
      "COVERAGE_GROUP", "BAD", "BAD_HET", "BAD_FEELING", "GDS",
      "ID_SEQ...2", "ID_SEQ...5", "to", "from", "old.format", "LABELS"
    )
  )
}
